// @avl-services/ng-services/esm5/lib/theme.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
var ThemeService = /** @class */ (function () {
    function ThemeService() {
        this.themeChangedSubject = new Subject();
        this.themeChanged = this.themeChangedSubject.asObservable();
        this.changeTheme();
    }
    Object.defineProperty(ThemeService.prototype, "isDarkTheme", {
        get: function () {
            if (this._isDarkTheme == null) {
                this._isDarkTheme = localStorage.getItem('theme') === 'dark' || false;
            }
            return this._isDarkTheme;
        },
        set: function (newValue) {
            this._isDarkTheme = newValue;
            localStorage.setItem('theme', newValue === true ? 'dark' : 'light');
            this.themeChangedSubject.next(this.getCurrentTheme());
            this.changeTheme();
        },
        enumerable: true,
        configurable: true
    });
    ThemeService.prototype.getCurrentTheme = function () {
        return this.isDarkTheme ? 'dark' : 'light';
    };
    ThemeService.prototype.changeTheme = function () {
        document.body.setAttribute('theme', this.getCurrentTheme());
    };
    ThemeService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThemeService_Factory() { return new ThemeService(); }, token: ThemeService, providedIn: "root" });
ThemeService.ɵfac = function ThemeService_Factory(t) { return new (t || ThemeService)(); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThemeService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
    return ThemeService;
}());
export { ThemeService };
