// @avl-services/ng-services/esm5/lib/translation-management.service.js
import { __decorate } from "tslib";
/// <reference types="@avl-controls/core" />
import { Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { NgxTranslateTranslationService } from '@avl-controls/ng-binders';
import '@avl-controls/core/dist/avl-localization-service/avl-localization-service';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "@avl-controls/ng-binders";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-translate/core';
import * as ɵngcc2 from '@avl-controls/ng-binders';
var TranslationManagementService = /** @class */ (function () {
    function TranslationManagementService(translateService, translationService) {
        this.translateService = translateService;
        this.translationService = translationService;
        this.languages = [
            { code: 'en', label: 'English' },
            { code: 'hr', label: 'Croatian' },
        ];
    }
    TranslationManagementService.prototype.initializeTranslations = function () {
        var _this = this;
        return function () {
            AvlLocalizationService.translationService = _this.translationService;
            var lang = localStorage.getItem('lang');
            if (lang == null) {
                lang = _this.languages[0].code;
                localStorage.setItem('lang', lang);
            }
            return _this.translateService.use(lang).toPromise();
        };
    };
    Object.defineProperty(TranslationManagementService.prototype, "language", {
        get: function () {
            return localStorage.getItem('lang');
        },
        enumerable: true,
        configurable: true
    });
    TranslationManagementService.prototype.changeLanguage = function (lang) {
        var oldLang = localStorage.getItem('lang');
        if (!oldLang || oldLang !== lang) {
            localStorage.setItem('lang', lang);
            location.reload();
        }
    };
    TranslationManagementService.ctorParameters = function () { return [
        { type: TranslateService },
        { type: NgxTranslateTranslationService }
    ]; };
    TranslationManagementService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TranslationManagementService_Factory() { return new TranslationManagementService(i0.ɵɵinject(i1.TranslateService), i0.ɵɵinject(i2.NgxTranslateTranslationService)); }, token: TranslationManagementService, providedIn: "root" });
TranslationManagementService.ɵfac = function TranslationManagementService_Factory(t) { return new (t || TranslationManagementService)(ɵngcc0.ɵɵinject(ɵngcc1.TranslateService), ɵngcc0.ɵɵinject(ɵngcc2.NgxTranslateTranslationService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TranslationManagementService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.TranslateService }, { type: ɵngcc2.NgxTranslateTranslationService }]; }, null); })();
    return TranslationManagementService;
}());
export { TranslationManagementService };
