// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/animate-button.directive.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, Input, HostBinding } from '@angular/core';
import { MARGIN } from '../avl-notification-container.consts';
let AnimateButton = class AnimateButton {
    constructor() {
        this.getInfoHeight = () => {
            return this.infoStackHeight > this.infoStackIndicatorHeight
                ? this.infoStackHeight
                : this.infoStackIndicatorHeight;
        };
        this.getWarningHeight = () => {
            return this.warningStackHeight > this.warningStackIndicatorHeight
                ? this.warningStackHeight
                : this.warningStackIndicatorHeight;
        };
        this.getOffsetHeight = (height) => {
            return height ? height + MARGIN * 2 : 0;
        };
    }
    ngOnChanges(changes) {
        const infoStackOffsetHeight = this.getOffsetHeight(this.getInfoHeight());
        const warningStackOffsetHeight = this.getOffsetHeight(this.getWarningHeight());
        this.styleTop = infoStackOffsetHeight + warningStackOffsetHeight + 'px';
    }
};
AnimateButton.ɵfac = function AnimateButton_Factory(t) { return new (t || AnimateButton)(); };
AnimateButton.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AnimateButton, selectors: [["", "animateButton", ""]], hostVars: 2, hostBindings: function AnimateButton_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("top", ctx.styleTop);
    } }, inputs: { increment: "increment", infoStackHeight: "infoStackHeight", warningStackHeight: "warningStackHeight", infoStackIndicatorHeight: "infoStackIndicatorHeight", warningStackIndicatorHeight: "warningStackIndicatorHeight" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
__decorate([
    Input(),
    __metadata("design:type", Number)
], AnimateButton.prototype, "increment", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], AnimateButton.prototype, "infoStackHeight", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], AnimateButton.prototype, "warningStackHeight", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], AnimateButton.prototype, "infoStackIndicatorHeight", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], AnimateButton.prototype, "warningStackIndicatorHeight", void 0);
__decorate([
    HostBinding('style.top'),
    __metadata("design:type", String)
], AnimateButton.prototype, "styleTop", void 0);
AnimateButton = __decorate([ __metadata("design:paramtypes", [])
], AnimateButton);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AnimateButton, [{
        type: Directive,
        args: [{ selector: '[animateButton]' }]
    }], function () { return []; }, { styleTop: [{
            type: HostBinding,
            args: ['style.top']
        }], increment: [{
            type: Input
        }], infoStackHeight: [{
            type: Input
        }], warningStackHeight: [{
            type: Input
        }], infoStackIndicatorHeight: [{
            type: Input
        }], warningStackIndicatorHeight: [{
            type: Input
        }] }); })();
export { AnimateButton };
