// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/animate-snackbar.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import { animate, style } from '@angular/animations';
import { ANIMATION_DURATION, MARGIN } from '../avl-notification-container.consts';
import * as i0 from "@angular/core";
import * as i1 from "@angular/animations";
/** @hidden */
export class AnimateSnackbar {
    constructor(builder, el) {
        this.builder = builder;
        /** @hidden */
        this.defaultClass = true;
        this.getAnimation = (animationName, initialStyle) => {
            const elementWidth = this.element.nativeElement.offsetWidth - 3 * MARGIN;
            const inStackStyle = style({
                marginTop: MARGIN + 'px',
                marginLeft: MARGIN + 'px',
                marginRight: MARGIN + 'px',
                width: elementWidth + 'px'
            });
            switch (animationName) {
                case 4 /* FromStack */:
                    return [animate(ANIMATION_DURATION + 'ms ease-in', inStackStyle), initialStyle];
                case 3 /* ToStack */: {
                    return [initialStyle, animate(ANIMATION_DURATION + 'ms ease-in', inStackStyle)];
                }
                default:
                    return [initialStyle, animate('1ms ease-in', initialStyle)];
            }
        };
        this.element = el;
    }
    ngOnChanges(changes) {
        if (changes.animateSnackbar) {
            const isInStack = changes.animateSnackbar.currentValue === 3 /* ToStack */;
            if (isInStack) {
                this.move();
            }
            else if (changes.animateSnackbar.previousValue === 3 /* ToStack */) {
                this.move();
            }
        }
    }
    show(metadata) {
        if (this.player) {
            this.player.destroy();
        }
        if (metadata) {
            const factory = this.builder.build(metadata);
            const player = factory.create(this.element.nativeElement);
            player.play();
        }
    }
    move() {
        const animationMetadata = this.animate();
        this.show(animationMetadata);
    }
    animate() {
        const initialStyle = style({
            marginTop: '0',
            width: '100%',
            marginLeft: '0',
            marginRight: '0'
        });
        return this.getAnimation(this.animateSnackbar, initialStyle);
    }
}
AnimateSnackbar.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AnimateSnackbar, deps: [{ token: i1.AnimationBuilder }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
AnimateSnackbar.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.3", type: AnimateSnackbar, selector: "[animateSnackbar]", inputs: { animateSnackbar: "animateSnackbar", positionAnimation: "positionAnimation", notification: "notification", reopen: "reopen" }, host: { properties: { "class.in-stack": "this.inStack", "class.no-animation": "this.noAnimation", "class.snackbar-container": "this.defaultClass" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AnimateSnackbar, decorators: [{
            type: Directive,
            args: [{ selector: '[animateSnackbar]' }]
        }], ctorParameters: function () { return [{ type: i1.AnimationBuilder }, { type: i0.ElementRef }]; }, propDecorators: { animateSnackbar: [{
                type: Input
            }], positionAnimation: [{
                type: Input
            }], notification: [{
                type: Input,
                args: ['notification']
            }], reopen: [{
                type: Input,
                args: ['reopen']
            }], inStack: [{
                type: HostBinding,
                args: ['class.in-stack']
            }], noAnimation: [{
                type: HostBinding,
                args: ['class.no-animation']
            }], defaultClass: [{
                type: HostBinding,
                args: ['class.snackbar-container']
            }] } });
