// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/animate-button.directive.js
import { Directive, Input, HostBinding } from '@angular/core';
import { MARGIN } from '../avl-notification-container.consts';
import * as i0 from "@angular/core";
export class AnimateButton {
    constructor() {
        this.getInfoHeight = () => {
            return this.infoStackHeight > this.infoStackIndicatorHeight
                ? this.infoStackHeight
                : this.infoStackIndicatorHeight;
        };
        this.getWarningHeight = () => {
            return this.warningStackHeight > this.warningStackIndicatorHeight
                ? this.warningStackHeight
                : this.warningStackIndicatorHeight;
        };
        this.getOffsetHeight = (height) => {
            return height ? height + MARGIN * 2 : 0;
        };
    }
    ngOnChanges(changes) {
        const infoStackOffsetHeight = this.getOffsetHeight(this.getInfoHeight());
        const warningStackOffsetHeight = this.getOffsetHeight(this.getWarningHeight());
        this.styleTop = infoStackOffsetHeight + warningStackOffsetHeight + 'px';
    }
}
AnimateButton.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AnimateButton, deps: [], target: i0.ɵɵFactoryTarget.Directive });
AnimateButton.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.3", type: AnimateButton, selector: "[animateButton]", inputs: { increment: "increment", infoStackHeight: "infoStackHeight", warningStackHeight: "warningStackHeight", infoStackIndicatorHeight: "infoStackIndicatorHeight", warningStackIndicatorHeight: "warningStackIndicatorHeight" }, host: { properties: { "style.top": "this.styleTop" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AnimateButton, decorators: [{
            type: Directive,
            args: [{ selector: '[animateButton]' }]
        }], ctorParameters: function () { return []; }, propDecorators: { increment: [{
                type: Input
            }], infoStackHeight: [{
                type: Input
            }], warningStackHeight: [{
                type: Input
            }], infoStackIndicatorHeight: [{
                type: Input
            }], warningStackIndicatorHeight: [{
                type: Input
            }], styleTop: [{
                type: HostBinding,
                args: ['style.top']
            }] } });
