// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/show-hide-snackbar.directive.js
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**@hidden */
export class SnackbarMethodDescription {
    constructor(method, animate) {
        this.method = method;
        this.animate = animate;
    }
}
export class ShowHideSnackbarDirective {
    constructor(el) {
        this.element = el;
    }
    ngOnChanges(changes) {
        var _a, _b, _c, _d;
        if (changes.methodDescription) {
            if (((_a = changes.methodDescription.currentValue) === null || _a === void 0 ? void 0 : _a.method) === 'show') {
                setTimeout(() => {
                    var _a, _b;
                    return (_a = this.element.nativeElement) === null || _a === void 0 ? void 0 : _a.showSnackbar('relative', !((_b = changes.methodDescription.currentValue) === null || _b === void 0 ? void 0 : _b.animate));
                }, 0);
            }
            else if (((_b = changes.methodDescription.currentValue) === null || _b === void 0 ? void 0 : _b.method) === 'hide') {
                (_c = this.element.nativeElement) === null || _c === void 0 ? void 0 : _c.hideSnackbar((_d = changes.methodDescription.currentValue) === null || _d === void 0 ? void 0 : _d.animate);
            }
        }
    }
}
ShowHideSnackbarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: ShowHideSnackbarDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
ShowHideSnackbarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.3", type: ShowHideSnackbarDirective, selector: "[showHideSnackbar]", inputs: { methodDescription: "methodDescription" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: ShowHideSnackbarDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[showHideSnackbar]' }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { methodDescription: [{
                type: Input
            }] } });
