// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/show-hide-snackbar.directive.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, Input, SimpleChanges, OnChanges, ElementRef } from '@angular/core';
/**@hidden */
export class SnackbarMethodDescription {
    constructor(method, animate) {
        this.method = method;
        this.animate = animate;
    }
}
let ShowHideSnackbarDirective = class ShowHideSnackbarDirective {
    constructor(el) {
        this.element = el;
    }
    ngOnChanges(changes) {
        var _a, _b, _c, _d;
        if (changes.methodDescription) {
            if (((_a = changes.methodDescription.currentValue) === null || _a === void 0 ? void 0 : _a.method) === 'show') {
                setTimeout(() => { var _a, _b; return (_a = this.element.nativeElement) === null || _a === void 0 ? void 0 : _a.showSnackbar('relative', !((_b = changes.methodDescription.currentValue) === null || _b === void 0 ? void 0 : _b.animate)); }, 0);
            }
            else if (((_b = changes.methodDescription.currentValue) === null || _b === void 0 ? void 0 : _b.method) === 'hide') {
                (_c = this.element.nativeElement) === null || _c === void 0 ? void 0 : _c.hideSnackbar((_d = changes.methodDescription.currentValue) === null || _d === void 0 ? void 0 : _d.animate);
            }
        }
    }
};
ShowHideSnackbarDirective.ɵfac = function ShowHideSnackbarDirective_Factory(t) { return new (t || ShowHideSnackbarDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ShowHideSnackbarDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ShowHideSnackbarDirective, selectors: [["", "showHideSnackbar", ""]], inputs: { methodDescription: "methodDescription" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
ShowHideSnackbarDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input(),
    __metadata("design:type", SnackbarMethodDescription)
], ShowHideSnackbarDirective.prototype, "methodDescription", void 0);
ShowHideSnackbarDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], ShowHideSnackbarDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ShowHideSnackbarDirective, [{
        type: Directive,
        args: [{ selector: '[showHideSnackbar]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { methodDescription: [{
            type: Input
        }] }); })();
export { ShowHideSnackbarDirective };
