// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/snackbar-listener.directive.js
import { Directive, HostListener, Output, EventEmitter } from '@angular/core';
import * as i0 from "@angular/core";
export class SnackbarListener {
    constructor() {
        /** @hidden */
        this.notificationClicked = new EventEmitter();
        /** @hidden */
        this.buttonClicked = new EventEmitter();
        /** @hidden */
        this.notificationDismissed = new EventEmitter();
    }
    onNotificationClicked(event) {
        this.notificationClicked.emit(event);
    }
    onButtonClicked(event) {
        if (event instanceof Event) {
            this.buttonClicked.emit(event);
        }
    }
    onNotificationDismissed(event) {
        if (event instanceof Event) {
            this.notificationDismissed.emit(event);
        }
    }
}
SnackbarListener.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: SnackbarListener, deps: [], target: i0.ɵɵFactoryTarget.Directive });
SnackbarListener.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.3", type: SnackbarListener, selector: "[snackbarListener]", outputs: { notificationClicked: "notificationClicked", buttonClicked: "buttonClicked", notificationDismissed: "notificationDismissed" }, host: { listeners: { "notification-clicked": "onNotificationClicked($event)", "button-clicked": "onButtonClicked($event)", "notification-dismissed": "onNotificationDismissed($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: SnackbarListener, decorators: [{
            type: Directive,
            args: [{ selector: '[snackbarListener]' }]
        }], propDecorators: { notificationClicked: [{
                type: Output
            }], buttonClicked: [{
                type: Output
            }], notificationDismissed: [{
                type: Output
            }], onNotificationClicked: [{
                type: HostListener,
                args: ['notification-clicked', ['$event']]
            }], onButtonClicked: [{
                type: HostListener,
                args: ['button-clicked', ['$event']]
            }], onNotificationDismissed: [{
                type: HostListener,
                args: ['notification-dismissed', ['$event']]
            }] } });
