// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-container/directives/snackbar-listener.directive.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, HostListener, Output, EventEmitter } from '@angular/core';
let SnackbarListener = class SnackbarListener {
    constructor() {
        /** @hidden */
        this.notificationClicked = new EventEmitter();
        /** @hidden */
        this.buttonClicked = new EventEmitter();
        /** @hidden */
        this.notificationDismissed = new EventEmitter();
    }
    onNotificationClicked(event) {
        this.notificationClicked.emit(event);
    }
    onButtonClicked(event) {
        if (event instanceof Event) {
            this.buttonClicked.emit(event);
        }
    }
    onNotificationDismissed(event) {
        if (event instanceof Event) {
            this.notificationDismissed.emit(event);
        }
    }
};
SnackbarListener.ɵfac = function SnackbarListener_Factory(t) { return new (t || SnackbarListener)(); };
SnackbarListener.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: SnackbarListener, selectors: [["", "snackbarListener", ""]], hostBindings: function SnackbarListener_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("notification-clicked", function SnackbarListener_notification_clicked_HostBindingHandler($event) { return ctx.onNotificationClicked($event); })("button-clicked", function SnackbarListener_button_clicked_HostBindingHandler($event) { return ctx.onButtonClicked($event); })("notification-dismissed", function SnackbarListener_notification_dismissed_HostBindingHandler($event) { return ctx.onNotificationDismissed($event); });
    } }, outputs: { notificationClicked: "notificationClicked", buttonClicked: "buttonClicked", notificationDismissed: "notificationDismissed" } });
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], SnackbarListener.prototype, "notificationClicked", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], SnackbarListener.prototype, "buttonClicked", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], SnackbarListener.prototype, "notificationDismissed", void 0);
__decorate([
    HostListener('notification-clicked', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], SnackbarListener.prototype, "onNotificationClicked", null);
__decorate([
    HostListener('button-clicked', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], SnackbarListener.prototype, "onButtonClicked", null);
__decorate([
    HostListener('notification-dismissed', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], SnackbarListener.prototype, "onNotificationDismissed", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SnackbarListener, [{
        type: Directive,
        args: [{ selector: '[snackbarListener]' }]
    }], function () { return []; }, { notificationClicked: [{
            type: Output
        }], buttonClicked: [{
            type: Output
        }], notificationDismissed: [{
            type: Output
        }], onNotificationClicked: [{
            type: HostListener,
            args: ['notification-clicked', ['$event']]
        }], onButtonClicked: [{
            type: HostListener,
            args: ['button-clicked', ['$event']]
        }], onNotificationDismissed: [{
            type: HostListener,
            args: ['notification-dismissed', ['$event']]
        }] }); })();
export { SnackbarListener };
