// @avl-ng-controls/navigation/esm2015/lib/components/avl-side-nav/avl-side-nav.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@angular/common';

function AvlSideNavComponent_ng_template_5_a_0_avl_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-icon");
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵattribute("icon", item_r3.icon);
} }
function AvlSideNavComponent_ng_template_5_a_0_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 21);
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("src", item_r3.iconUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
function AvlSideNavComponent_ng_template_5_a_0_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 15);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_a_0_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(2); return ctx_r12.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_a_0_avl_icon_2_Template, 1, 1, "avl-icon", 17);
    ɵngcc0.ɵɵtemplate(3, AvlSideNavComponent_ng_template_5_a_0_img_3_Template, 1, 1, "img", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 19);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "div", 20);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("selected", item_r3 === ctx_r4.selectedPage);
    ɵngcc0.ɵɵpropertyInterpolate("routerLink", item_r3.route);
    ɵngcc0.ɵɵproperty("queryParams", ctx_r4.getParams(item_r3.params));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.iconUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r3.label);
} }
function AvlSideNavComponent_ng_template_5_div_1_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 27);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_1_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(3); return ctx_r17.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const paragraph_r16 = ctx.$implicit;
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r15 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("selected", paragraph_r16.id === ctx_r15.selectedParagraph);
    ɵngcc0.ɵɵpropertyInterpolate("href", ctx_r15.trimSlash(item_r3.route + ctx_r15.getUrlParams(item_r3.params) + "#" + paragraph_r16.id), ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(paragraph_r16.label);
} }
function AvlSideNavComponent_ng_template_5_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_1_a_2_Template, 3, 4, "a", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 24);
    ɵngcc0.ɵɵelement(4, "div", 25);
    ɵngcc0.ɵɵelement(5, "div", 26);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@expandCollapse", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", item_r3.paragraphs);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r5.getScrubberStyle());
} }
function AvlSideNavComponent_ng_template_5_div_2_avl_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-icon");
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵattribute("icon", item_r3.icon);
} }
function AvlSideNavComponent_ng_template_5_div_2_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 21);
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("src", item_r3.iconUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
function AvlSideNavComponent_ng_template_5_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r27 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_2_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r27); const item_r3 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r25 = ɵngcc0.ɵɵnextContext(); return ctx_r25.groupClicked(item_r3); });
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_2_avl_icon_2_Template, 1, 1, "avl-icon", 17);
    ɵngcc0.ɵɵtemplate(3, AvlSideNavComponent_ng_template_5_div_2_img_3_Template, 1, 1, "img", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 29);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "avl-icon", 30);
    ɵngcc0.ɵɵelement(7, "div", 20);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("expanded", item_r3.expanded)("selected", !ctx_r6.open && item_r3 === ctx_r6.selectedGroup);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.iconUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r3.label);
} }
function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r35 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 34);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r35); const ctx_r34 = ɵngcc0.ɵɵnextContext(5); return ctx_r34.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const paragraph_r33 = ctx.$implicit;
    const subItem_r30 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r32 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("selected", paragraph_r33.id === ctx_r32.selectedParagraph);
    ɵngcc0.ɵɵpropertyInterpolate3("href", "", subItem_r30.route, "", ctx_r32.getUrlParams(item_r3.params), "#", paragraph_r33.id, "", ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(paragraph_r33.label);
} }
function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_a_2_Template, 3, 6, "a", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 24);
    ɵngcc0.ɵɵelement(4, "div", 25);
    ɵngcc0.ɵɵelement(5, "div", 26);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const subItem_r30 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r31 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("@expandCollapse", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", subItem_r30.paragraphs);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r31.getScrubberStyle());
} }
function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    const _r40 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 32);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r40); const ctx_r39 = ɵngcc0.ɵɵnextContext(3); return ctx_r39.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "div", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_Template, 6, 3, "div", 12);
} if (rf & 2) {
    const subItem_r30 = ctx.$implicit;
    const ctx_r29 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassProp("selected", subItem_r30 === ctx_r29.selectedPage);
    ɵngcc0.ɵɵpropertyInterpolate("routerLink", subItem_r30.route);
    ɵngcc0.ɵɵproperty("queryParams", ctx_r29.getParams(subItem_r30.params));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(subItem_r30.label);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", subItem_r30.paragraphs && subItem_r30.paragraphs.length > 0 && subItem_r30 === ctx_r29.selectedPage);
} }
function AvlSideNavComponent_ng_template_5_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r42 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 31);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_3_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r42); const ctx_r41 = ɵngcc0.ɵɵnextContext(2); return ctx_r41.toggleOpen(); });
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_3_ng_template_2_Template, 5, 6, "ng-template", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("@expandCollapse", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", item_r3.subItems);
} }
function AvlSideNavComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, AvlSideNavComponent_ng_template_5_a_0_Template, 7, 7, "a", 11);
    ɵngcc0.ɵɵtemplate(1, AvlSideNavComponent_ng_template_5_div_1_Template, 6, 3, "div", 12);
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_2_Template, 8, 7, "div", 13);
    ɵngcc0.ɵɵtemplate(3, AvlSideNavComponent_ng_template_5_div_3_Template, 3, 2, "div", 14);
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", !item_r3.subItems);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.open && item_r3.paragraphs && item_r3.paragraphs.length > 0 && item_r3 === ctx_r2.selectedPage);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.subItems);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.open && item_r3.subItems && item_r3.subItems.length > 0 && item_r3.expanded);
} }
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Output, EventEmitter, Input, HostBinding } from '@angular/core';
import { trigger, style, animate, transition } from '@angular/animations';
import { Router, ActivationEnd, ActivatedRoute } from '@angular/router';
import { filter } from 'rxjs/operators';
import '@avl-controls/core/dist/avl-icon/avl-icon';
const rowHeightPx = 48;
let AvlSideNavComponent = class AvlSideNavComponent {
    constructor(router, activeRoute) {
        this.router = router;
        this.activeRoute = activeRoute;
        this.openValue = false;
        /**
         * Emitted when the open property changes.
         */
        this.openChange = new EventEmitter();
        activeRoute.queryParams.subscribe(p => {
            this.params = p;
        });
        this.subscription = this.router.events
            .pipe(filter(event => event instanceof ActivationEnd))
            .subscribe((event) => {
            this.selectItems();
        });
    }
    /**
     * Can be set to open or close the side nav.
     * Returns the current state of the side nav.
     */
    get open() {
        return this.openValue;
    }
    set open(val) {
        this.openValue = val;
        this.openChange.next(val);
    }
    /**
     * @hidden
     */
    toggleOpen() {
        if (!this.manuallyOpened) {
            this.open = !this.open;
        }
    }
    /**
     * @hidden
     */
    arrowToggleOpen() {
        this.manuallyOpened = !this.open;
        this.open = !this.open;
    }
    /**
     * @hidden
     */
    mobileClose() {
        if (Math.max(document.documentElement.clientWidth, window.innerWidth) < 960) {
            this.open = false;
        }
    }
    /**
     * @hidden
     */
    groupClicked(item) {
        if (!this.open) {
            this.open = true;
            if (!item.expanded) {
                item.expanded = true;
            }
        }
        else {
            item.expanded = !item.expanded;
        }
    }
    /**
     * @hidden
     */
    getScrubberStyle() {
        return {
            height: this.selectedParagraph ? '14px' : '0px',
            top: `${this.selectedParagraphIndex * rowHeightPx}px`
        };
    }
    /**
    * @hidden
    */
    trimSlash(str) {
        if (str[0] === '/') {
            str = str.substring(1, str.length);
        }
        if (str[str.length - 1] === '/') {
            str = str.substring(0, str.length - 1);
        }
        return str;
    }
    selectItems() {
        if (!this.menuItems) {
            return;
        }
        const fragment = this.router.parseUrl(this.router.url).fragment;
        const url = this.router.url.split('#')[0];
        const queue = this.menuItems.slice();
        let currentGroup;
        this.selectedGroup = undefined;
        this.selectedPage = undefined;
        this.selectedParagraph = undefined;
        let fallbackSelectedPage;
        while (queue.length) {
            const item = queue.pop();
            if (isGroup(item)) {
                currentGroup = item;
                item.subItems.map(subitem => { queue.push(subitem); });
            }
            else {
                if (fallbackSelectedPage == null && this.trimSlash(url).startsWith(this.trimSlash(item.route))) {
                    fallbackSelectedPage = item;
                }
                if (item.route && this.trimSlash(url).endsWith(this.trimSlash(item.route + this.getUrlParams(item.params)))) {
                    if (this.selectedPage && this.selectedPage.route && this.selectedPage.route.length > item.route.length) {
                        continue;
                    }
                    if (item.params && item.params.length > 0) {
                        const object = new Object();
                        item.params.forEach(i => {
                            object[i.name] = i.value;
                        });
                        if (JSON.stringify(this.params) === JSON.stringify(object)) {
                            this.selectedPage = item;
                        }
                    }
                    else {
                        this.selectedPage = item;
                    }
                    if (isTopLevelPage(item)) {
                        this.selectedGroup = undefined;
                    }
                    else {
                        this.selectedGroup = currentGroup;
                    }
                    if (item.paragraphs && item.paragraphs.length > 0) {
                        const paragraphIndex = item.paragraphs.findIndex((p) => p.id === fragment);
                        if (fragment && item.paragraphs && paragraphIndex !== -1) {
                            this.selectedParagraph = fragment;
                            this.selectedParagraphIndex = paragraphIndex;
                        }
                    }
                }
            }
        }
        if (this.selectedPage == null) {
            this.selectedPage = fallbackSelectedPage;
        }
    }
    /**
    * @hidden
    */
    ngOnChanges(changes) {
        if (changes['menuItems']) {
            this.selectItems();
        }
    }
    /**
    * @hidden
    */
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    /**
    * @hidden
    */
    getParams(params) {
        if (params && params.length > 0) {
            const object = {};
            params.forEach(param => {
                object[param.name] = param.value;
            });
            return object;
        }
        else {
            return null;
        }
    }
    /**
    * @hidden
    */
    getUrlParams(params) {
        if (params && params.length > 0) {
            let url = '?';
            params.forEach(param => {
                url += param.name + '=' + param.value + '&';
            });
            return url.slice(0, -1);
        }
        else {
            return '';
        }
    }
};
AvlSideNavComponent.ɵfac = function AvlSideNavComponent_Factory(t) { return new (t || AvlSideNavComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute)); };
AvlSideNavComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlSideNavComponent, selectors: [["avl-side-nav"]], hostVars: 6, hostBindings: function AvlSideNavComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("open", ctx.open)("responsive", ctx.responsive)("mobile", ctx.mobile);
    } }, inputs: { open: "open", menuItems: "menuItems", responsive: "responsive", mobile: "mobile" }, outputs: { openChange: "openChange" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 11, vars: 5, consts: [[1, "outer-container", "toolbar-spacing"], ["outerContainer", ""], [1, "item-scroll"], [1, "inner-container"], ["innerContainer", ""], ["ngFor", "", 3, "ngForOf"], [1, "arrow-spacer"], ["id", "Layer_1", "data-name", "Layer 1", "xmlns", "http://www.w3.org/2000/svg", "width", "24", "height", "24", "viewBox", "0 0 24 24", 1, "open-close-arrow", 3, "click"], ["transform", "translate(405 102) rotate(-180)"], ["d", "M-972.955,492l6.774-6,1.581,1.4-5.194,4.6,5.194,4.6-1.581,1.4Zm-5.645,0,6.773-6,1.581,1.4-5.193,4.6,5.193,4.6-1.581,1.4Z", "transform", "translate(1364.6 -402)"], [1, "scrim", 3, "click"], ["tabindex", "0", "class", "item top-level-page", 3, "routerLink", "queryParams", "selected", "click", 4, "ngIf"], ["class", "child-container", 4, "ngIf"], ["tabindex", "0", "class", "item group", 3, "expanded", "selected", "click", 4, "ngIf"], ["class", "child-container", 3, "click", 4, "ngIf"], ["tabindex", "0", 1, "item", "top-level-page", 3, "routerLink", "queryParams", "click"], [1, "image-div"], [4, "ngIf"], [3, "src", 4, "ngIf"], [1, "label-div"], [1, "item-overlay"], [3, "src"], [1, "child-container"], ["class", "item top-level-paragraph", 3, "href", "selected", "click", 4, "ngFor", "ngForOf"], [1, "page-progress-bar"], [1, "page-progress-bar-background"], [1, "page-progress-bar-scrubber", 3, "ngStyle"], [1, "item", "top-level-paragraph", 3, "href", "click"], ["tabindex", "0", 1, "item", "group", 3, "click"], [1, "label-div", "group"], ["icon", "arrow-drop-down", 1, "group-arrow"], [1, "child-container", 3, "click"], [1, "item", "subpage", 3, "routerLink", "queryParams", "click"], ["class", "item subpage-paragraph", 3, "href", "selected", "click", 4, "ngFor", "ngForOf"], [1, "item", "subpage-paragraph", 3, "href", "click"]], template: function AvlSideNavComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3, 4);
        ɵngcc0.ɵɵtemplate(5, AvlSideNavComponent_ng_template_5_Template, 4, 4, "ng-template", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(6, "div", 6);
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(7, "svg", 7);
        ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_Template__svg_svg_click_7_listener() { return ctx.arrowToggleOpen(); });
        ɵngcc0.ɵɵelementStart(8, "g", 8);
        ɵngcc0.ɵɵelement(9, "path", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵnamespaceHTML();
        ɵngcc0.ɵɵelementStart(10, "div", 10);
        ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_Template_div_click_10_listener() { return ctx.open = false; });
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("closed", !ctx.open);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.menuItems);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵclassProp("closed", !ctx.open);
    } }, directives: [ɵngcc2.NgForOf, ɵngcc2.NgIf, ɵngcc1.RouterLinkWithHref, ɵngcc2.NgStyle], styles: ["[_nghost-%COMP%]{font-family:var(--avl-font-family);height:100%;display:block;--text-color:var(--avl-side-nav-text-color, var(--avl-emphasis-high-color));--icons-color:var(--avl-side-nav-icons-color, var(--avl-emphasis-medium-color));--background-color:var(--avl-side-nav-background-color, var(--avl-side-nav-color));--item-icon-width:72px;--side-nav-mini-width:var(--item-icon-width);--side-nav-open-width:var(--avl-side-nav-open-width, 256px);--item-height:48px;box-shadow:var(--avl-box-shadow-12dp);z-index:3}.outer-container[_ngcontent-%COMP%]{position:relative;box-sizing:content-box;height:calc(100% - var(--avl-app-bar-height,56px));background-color:var(--background-color);width:var(--side-nav-open-width);overflow:hidden}.inner-container[_ngcontent-%COMP%]{width:var(--side-nav-open-width)}.toolbar-spacing[_ngcontent-%COMP%]{padding-top:var(--avl-app-bar-height,56px)}.open-close-arrow[_ngcontent-%COMP%]{position:absolute;bottom:12px;right:24px;font-size:24px;fill:var(--icons-color);cursor:pointer;-webkit-user-select:none;-moz-user-select:none;user-select:none;width:24px}.open-close-arrow[_ngcontent-%COMP%]:not(.closed){transform:scale(-1,1)}.outer-container.closed[_ngcontent-%COMP%]{width:var(--side-nav-mini-width);-webkit-animation-name:closing;animation-name:closing;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:195ms}.arrow-spacer[_ngcontent-%COMP%]{height:48px}.item-scroll[_ngcontent-%COMP%]{height:calc(100% - 48px);overflow-y:auto;overflow-x:hidden}.item[_ngcontent-%COMP%]{display:flex;align-items:center;cursor:pointer;color:var(--text-color);height:var(--item-height);position:relative;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;isolation:isolate;outline:0;white-space:pre-wrap;padding-right:24px}.item[_ngcontent-%COMP%]:active{color:var(--text-color)}.top-level-paragraph[_ngcontent-%COMP%]{padding-left:88px}.subpage[_ngcontent-%COMP%]{padding-left:72px}.subpage-paragraph[_ngcontent-%COMP%]{padding-left:88px}.label-div[_ngcontent-%COMP%]{z-index:10;font-family:Roboto;max-width:160px;overflow:hidden;text-overflow:ellipsis;max-height:40px;word-break:break-word;-webkit-box-orient:vertical;display:-webkit-box;-webkit-line-clamp:2}.group.expanded[_ngcontent-%COMP%] > .label-div.group[_ngcontent-%COMP%]{font-family:Roboto-Medium}.label-div.group[_ngcontent-%COMP%]{max-width:calc(100% - 112px)}.image-div[_ngcontent-%COMP%]{color:var(--avl-emphasis-medium-color);display:flex;align-items:center;justify-content:center;width:var(--item-icon-width);height:var(--item-height);margin-right:0;z-index:9}img[_ngcontent-%COMP%]{height:auto;width:auto;max-height:24px;max-width:24px}avl-icon[_ngcontent-%COMP%]{margin-bottom:0;font-size:24px;color:var(--avl-emphasis-medium-color)}.group-arrow[_ngcontent-%COMP%]{position:absolute;right:24px}.group.expanded[_ngcontent-%COMP%] > .group-arrow[_ngcontent-%COMP%]{transform:rotate(180deg)}.item-overlay[_ngcontent-%COMP%]{box-sizing:border-box;position:absolute;height:100%;width:100%;left:0;z-index:0}.item.selected[_ngcontent-%COMP%] > .item-overlay[_ngcontent-%COMP%]{background-color:var(--avl-side-nav-product-color,var(--avl-product-focus-color))}.item[_ngcontent-%COMP%]:hover:not(.selected) > .item-overlay[_ngcontent-%COMP%]{background-color:var(--avl-side-nav-product-color-hover,var(--avl-product-hover-color))}.item.selected[_ngcontent-%COMP%]{font-weight:500}.child-container[_ngcontent-%COMP%]{position:relative;display:flex;overflow:hidden;flex-direction:column-reverse}.page-progress-bar[_ngcontent-%COMP%]{position:absolute;left:72px;top:11px;bottom:11px;width:2px;overflow:hidden}.page-progress-bar-background[_ngcontent-%COMP%]{background-color:var(--text-color);opacity:.2;width:100%;height:100%}.page-progress-bar-scrubber[_ngcontent-%COMP%]{position:absolute;width:100%;background-color:var(--text-color);opacity:1;transition:height .2s,top .2s;margin-top:6px;margin-bottom:2px}@-webkit-keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@-webkit-keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}@keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}.outer-container[_ngcontent-%COMP%]:not(.closed){-webkit-animation-name:opening;animation-name:opening;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:225ms}@media (max-width:960px){.responsive[_nghost-%COMP%]{position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}.responsive[slot=left-side][_nghost-%COMP%]{left:0}.responsive[slot=right-side][_nghost-%COMP%]{right:0}.responsive[_nghost-%COMP%]:not(.open){box-shadow:none}.responsive[_nghost-%COMP%]   .item-scroll[_ngcontent-%COMP%]{height:100%}.responsive[_nghost-%COMP%]   .open-close-arrow[_ngcontent-%COMP%]{display:none}.responsive[_nghost-%COMP%]   .arrow-spacer[_ngcontent-%COMP%]{display:none;visibility:hidden}.responsive[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{transition:250ms;transition-duration:250px}.responsive[slot=left-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;left:0}.responsive.open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}.responsive[slot=left-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{left:var(--side-nav-open-width)}.responsive[slot=right-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{right:var(--side-nav-open-width)}.responsive[slot=right-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;right:0}}.mobile[_nghost-%COMP%]{position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}.mobile[slot=left-side][_nghost-%COMP%]{left:0}.mobile[slot=right-side][_nghost-%COMP%]{right:0}.mobile[_nghost-%COMP%]:not(.open){box-shadow:none}.mobile[_nghost-%COMP%]   .item-scroll[_ngcontent-%COMP%]{height:100%}.mobile[_nghost-%COMP%]   .open-close-arrow[_ngcontent-%COMP%]{display:none}.mobile[_nghost-%COMP%]   .arrow-spacer[_ngcontent-%COMP%]{display:none;visibility:hidden}.mobile[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{transition:250ms;transition-duration:250px}.mobile[slot=left-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;left:0}.mobile.open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}.mobile[slot=left-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{left:var(--side-nav-open-width)}.mobile[slot=right-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{right:var(--side-nav-open-width)}.mobile[slot=right-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;right:0}"], data: { animation: [
            trigger('expandCollapse', [
                transition(':enter', [
                    style({ 'max-height': '0px' }),
                    animate('225ms ease-in', style({ transform: 'translateY(0%)', 'max-height': '200px' }))
                ]),
                transition(':leave', [
                    style({ 'max-height': '200px' }),
                    animate('180ms ease-in', style({ 'max-height': '0px' }))
                ])
            ])
        ] } });
AvlSideNavComponent.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute }
];
__decorate([
    Input(),
    HostBinding('class.open'),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], AvlSideNavComponent.prototype, "open", null);
__decorate([
    Input(),
    __metadata("design:type", Array)
], AvlSideNavComponent.prototype, "menuItems", void 0);
__decorate([
    Input(),
    HostBinding('class.responsive'),
    __metadata("design:type", Boolean)
], AvlSideNavComponent.prototype, "responsive", void 0);
__decorate([
    Input(),
    HostBinding('class.mobile'),
    __metadata("design:type", Boolean)
], AvlSideNavComponent.prototype, "mobile", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlSideNavComponent.prototype, "openChange", void 0);
AvlSideNavComponent = __decorate([ __metadata("design:paramtypes", [Router,
        ActivatedRoute])
], AvlSideNavComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSideNavComponent, [{
        type: Component,
        args: [{
                selector: 'avl-side-nav',
                template: "<div #outerContainer [class.closed]=\"!open\" class=\"outer-container toolbar-spacing\">\r\n  <div class=\"item-scroll\">\r\n    <div #innerContainer class=\"inner-container\">\r\n      <ng-template ngFor let-item [ngForOf]=\"menuItems\">\r\n        <!-- Top level page -->\r\n        <a *ngIf=\"!item.subItems\" tabindex=\"0\" class=\"item top-level-page\" routerLink={{item.route}} [queryParams]=\"getParams(item.params)\" [class.selected]=\"item === selectedPage\" (click)=\"mobileClose()\">\r\n          <div class=\"image-div\">\r\n            <avl-icon *ngIf=\"item.icon\" attr.icon=\"{{item.icon}}\"></avl-icon>\r\n            <img *ngIf=\"item.iconUrl\" [src]=\"item.iconUrl\">\r\n          </div>\r\n          <div class=\"label-div\">{{item.label}}</div>\r\n          <div class=\"item-overlay\"></div>\r\n        </a>\r\n        <!-- Paragraphs of a top level page -->\r\n        <div class=\"child-container\" *ngIf=\"open && item.paragraphs && item.paragraphs.length > 0 && item === selectedPage\" [@expandCollapse]>\r\n          <!-- div for preventing reversing of the items by flex-direction: column-reverse -->\r\n          <div>\r\n            <a *ngFor=\"let paragraph of item.paragraphs\" class=\"item top-level-paragraph\"\r\n              href=\"{{trimSlash(item.route + getUrlParams(item.params) + '#' + paragraph.id)}}\" [class.selected]=\"paragraph.id === selectedParagraph\" (click)=\"mobileClose()\">\r\n              <div class=\"label-div\">{{paragraph.label}}</div>\r\n            </a>\r\n          </div>\r\n          <div class=\"page-progress-bar\">\r\n            <div class=\"page-progress-bar-background\"></div>\r\n            <div class=\"page-progress-bar-scrubber\" [ngStyle]=\"getScrubberStyle()\"></div>\r\n          </div>\r\n        </div>\r\n        <!-- Group -->\r\n        <div *ngIf=\"item.subItems\" tabindex=\"0\" class=\"item group\" [class.expanded]=\"item.expanded\" (click)=\"groupClicked(item)\"\r\n          [class.selected]=\"!open && item === selectedGroup\">\r\n          <div class=\"image-div\">\r\n            <avl-icon *ngIf=\"item.icon\" attr.icon=\"{{item.icon}}\"></avl-icon>\r\n            <img *ngIf=\"item.iconUrl\" [src]=\"item.iconUrl\">\r\n          </div>\r\n          <div class=\"label-div group\">{{item.label}}</div>\r\n          <avl-icon class=\"group-arrow\" icon=\"arrow-drop-down\"></avl-icon>\r\n          <div class=\"item-overlay\"></div>\r\n        </div>\r\n        <!-- Subpages -->\r\n        <div class=\"child-container\" *ngIf=\"open && item.subItems && item.subItems.length > 0 && item.expanded\" [@expandCollapse] (click)=\"toggleOpen()\">\r\n          <!-- div for preventing reversing of the items by flex-direction: column-reverse -->\r\n          <div>\r\n            <ng-template ngFor let-subItem [ngForOf]=\"item.subItems\">\r\n              <a class=\"item subpage\" routerLink={{subItem.route}} [queryParams]=\"getParams(subItem.params)\" [class.selected]=\"subItem === selectedPage\" (click)=\"mobileClose()\">\r\n                <div class=\"label-div\">{{subItem.label}}</div>\r\n                <div class=\"item-overlay\"></div>\r\n              </a>\r\n              <!-- Paragraphs of a subpage -->\r\n              <div class=\"child-container\" *ngIf=\"subItem.paragraphs && subItem.paragraphs.length > 0 && subItem === selectedPage\" [@expandCollapse]>\r\n                <!-- div for preventing reversing of the items by flex-direction: column-reverse -->\r\n                <div>\r\n                  <a *ngFor=\"let paragraph of subItem.paragraphs\" class=\"item subpage-paragraph\" href=\"{{subItem.route}}{{getUrlParams(item.params)}}#{{paragraph.id}}\"\r\n                    [class.selected]=\"paragraph.id === selectedParagraph\" (click)=\"mobileClose()\">\r\n                    <div class=\"label-div\">{{paragraph.label}}</div>\r\n                  </a>\r\n                </div>\r\n                <div class=\"page-progress-bar\">\r\n                  <div class=\"page-progress-bar-background\"></div>\r\n                  <div class=\"page-progress-bar-scrubber\" [ngStyle]=\"getScrubberStyle()\"></div>\r\n                </div>\r\n              </div>\r\n            </ng-template>\r\n          </div>\r\n        </div>\r\n      </ng-template>\r\n    </div>\r\n  </div>\r\n  <div class=\"arrow-spacer\"></div>\r\n\r\n  <svg (click)=\"arrowToggleOpen()\" [class.closed]=\"!open\" class=\"open-close-arrow\"  id=\"Layer_1\" data-name=\"Layer 1\" xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\" viewBox=\"0 0 24 24\">\r\n    <g transform=\"translate(405 102) rotate(-180)\">\r\n      <path d=\"M-972.955,492l6.774-6,1.581,1.4-5.194,4.6,5.194,4.6-1.581,1.4Zm-5.645,0,6.773-6,1.581,1.4-5.193,4.6,5.193,4.6-1.581,1.4Z\" transform=\"translate(1364.6 -402)\"/>\r\n    </g>\r\n  </svg>\r\n\r\n</div>\r\n<div class=\"scrim\" (click)=\"open = false;\"></div>",
                animations: [
                    trigger('expandCollapse', [
                        transition(':enter', [
                            style({ 'max-height': '0px' }),
                            animate('225ms ease-in', style({ transform: 'translateY(0%)', 'max-height': '200px' }))
                        ]),
                        transition(':leave', [
                            style({ 'max-height': '200px' }),
                            animate('180ms ease-in', style({ 'max-height': '0px' }))
                        ])
                    ])
                ],
                styles: [":host{font-family:var(--avl-font-family);height:100%;display:block;--text-color:var(--avl-side-nav-text-color, var(--avl-emphasis-high-color));--icons-color:var(--avl-side-nav-icons-color, var(--avl-emphasis-medium-color));--background-color:var(--avl-side-nav-background-color, var(--avl-side-nav-color));--item-icon-width:72px;--side-nav-mini-width:var(--item-icon-width);--side-nav-open-width:var(--avl-side-nav-open-width, 256px);--item-height:48px;box-shadow:var(--avl-box-shadow-12dp);z-index:3}.outer-container{position:relative;box-sizing:content-box;height:calc(100% - var(--avl-app-bar-height,56px));background-color:var(--background-color);width:var(--side-nav-open-width);overflow:hidden}.inner-container{width:var(--side-nav-open-width)}.toolbar-spacing{padding-top:var(--avl-app-bar-height,56px)}.open-close-arrow{position:absolute;bottom:12px;right:24px;font-size:24px;fill:var(--icons-color);cursor:pointer;-webkit-user-select:none;-moz-user-select:none;user-select:none;width:24px}.open-close-arrow:not(.closed){transform:scale(-1,1)}.outer-container.closed{width:var(--side-nav-mini-width);-webkit-animation-name:closing;animation-name:closing;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:195ms}.arrow-spacer{height:48px}.item-scroll{height:calc(100% - 48px);overflow-y:auto;overflow-x:hidden}.item{display:flex;align-items:center;cursor:pointer;color:var(--text-color);height:var(--item-height);position:relative;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;isolation:isolate;outline:0;white-space:pre-wrap;padding-right:24px}.item:active{color:var(--text-color)}.top-level-paragraph{padding-left:88px}.subpage{padding-left:72px}.subpage-paragraph{padding-left:88px}.label-div{z-index:10;font-family:Roboto;max-width:160px;overflow:hidden;text-overflow:ellipsis;max-height:40px;word-break:break-word;-webkit-box-orient:vertical;display:-webkit-box;-webkit-line-clamp:2}.group.expanded>.label-div.group{font-family:Roboto-Medium}.label-div.group{max-width:calc(100% - 112px)}.image-div{color:var(--avl-emphasis-medium-color);display:flex;align-items:center;justify-content:center;width:var(--item-icon-width);height:var(--item-height);margin-right:0;z-index:9}img{height:auto;width:auto;max-height:24px;max-width:24px}avl-icon{margin-bottom:0;font-size:24px;color:var(--avl-emphasis-medium-color)}.group-arrow{position:absolute;right:24px}.group.expanded>.group-arrow{transform:rotate(180deg)}.item-overlay{box-sizing:border-box;position:absolute;height:100%;width:100%;left:0;z-index:0}.item.selected>.item-overlay{background-color:var(--avl-side-nav-product-color,var(--avl-product-focus-color))}.item:hover:not(.selected)>.item-overlay{background-color:var(--avl-side-nav-product-color-hover,var(--avl-product-hover-color))}.item.selected{font-weight:500}.child-container{position:relative;display:flex;overflow:hidden;flex-direction:column-reverse}.page-progress-bar{position:absolute;left:72px;top:11px;bottom:11px;width:2px;overflow:hidden}.page-progress-bar-background{background-color:var(--text-color);opacity:.2;width:100%;height:100%}.page-progress-bar-scrubber{position:absolute;width:100%;background-color:var(--text-color);opacity:1;transition:height .2s,top .2s;margin-top:6px;margin-bottom:2px}@-webkit-keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@-webkit-keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}@keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}.outer-container:not(.closed){-webkit-animation-name:opening;animation-name:opening;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:225ms}@media (max-width:960px){:host(.responsive){position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}:host(.responsive[slot=left-side]){left:0}:host(.responsive[slot=right-side]){right:0}:host(.responsive:not(.open)){box-shadow:none}:host(.responsive) .item-scroll{height:100%}:host(.responsive) .open-close-arrow{display:none}:host(.responsive) .arrow-spacer{display:none;visibility:hidden}:host(.responsive) .scrim{transition:250ms;transition-duration:250px}:host(.responsive[slot=left-side]:not(.open)) .scrim{visibility:hidden;left:0}:host(.responsive.open) .scrim{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}:host(.responsive[slot=left-side].open) .scrim{left:var(--side-nav-open-width)}:host(.responsive[slot=right-side].open) .scrim{right:var(--side-nav-open-width)}:host(.responsive[slot=right-side]:not(.open)) .scrim{visibility:hidden;right:0}}:host(.mobile){position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}:host(.mobile[slot=left-side]){left:0}:host(.mobile[slot=right-side]){right:0}:host(.mobile:not(.open)){box-shadow:none}:host(.mobile) .item-scroll{height:100%}:host(.mobile) .open-close-arrow{display:none}:host(.mobile) .arrow-spacer{display:none;visibility:hidden}:host(.mobile) .scrim{transition:250ms;transition-duration:250px}:host(.mobile[slot=left-side]:not(.open)) .scrim{visibility:hidden;left:0}:host(.mobile.open) .scrim{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}:host(.mobile[slot=left-side].open) .scrim{left:var(--side-nav-open-width)}:host(.mobile[slot=right-side].open) .scrim{right:var(--side-nav-open-width)}:host(.mobile[slot=right-side]:not(.open)) .scrim{visibility:hidden;right:0}"]
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc1.ActivatedRoute }]; }, { openChange: [{
            type: Output
        }], open: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.open']
        }], menuItems: [{
            type: Input
        }], responsive: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.responsive']
        }], mobile: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.mobile']
        }] }); })();
export { AvlSideNavComponent };
/**
* @hidden
*/
function isGroup(item) {
    return item.subItems !== undefined && item.subItems.length > 0;
}
/**
* @hidden
*/
function isTopLevelPage(item) {
    return !isGroup(item) && (item.icon !== undefined || item.iconUrl !== undefined);
}
