// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-menu/avl-notification-menu.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@avl-controls/ng-binders';
import * as ɵngcc3 from '@ngx-translate/core';

const _c0 = ["list"];
function AvlNotificationMenuComponent_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" (", ctx_r0.numberOfUnreadNotifications, ") ");
} }
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { NOTIFICATIONS_OVERFLOW_MENU, TYPE_ICON_MAPPINGS } from './avl-notification-menu.consts';
let AvlNotificationMenuComponent = class AvlNotificationMenuComponent {
    constructor() {
        this.open = false;
        this.openChange = new EventEmitter();
        this.openDetails = new EventEmitter();
        this.notificationStateChanged = new EventEmitter();
        this.markAllAsRead = new EventEmitter();
        this.markAllAsUnread = new EventEmitter();
        this.overflowMenuItems = [];
        this.notificationList = [];
        this.numberOfUnreadNotifications = 0;
        this.overflowMenuHandlers = {
            markAllAsRead: () => {
                this.markAllAsRead.emit();
                this.numberOfUnreadNotifications = 0;
            },
            markAllAsUnread: () => {
                this.markAllAsUnread.emit();
                this.numberOfUnreadNotifications = this.notificationList.length;
            },
            undefined: () => { }
        };
        this.translateLabels();
    }
    set notifications(notifications) {
        this.notificationList = this.prepareNotifications(notifications);
        this.numberOfUnreadNotifications = notifications.filter(notif => !notif.read).length;
    }
    ngAfterViewInit() {
        this.list.nativeElement.revertedList = true;
    }
    onNotificationStateChanged(event) {
        const notificationMenuItem = Object.assign({}, event.detail.value);
        const notification = Object.assign(Object.assign({}, notificationMenuItem), { read: notificationMenuItem.iconToggleValue });
        notification.read ? this.numberOfUnreadNotifications-- : this.numberOfUnreadNotifications++;
        this.notificationStateChanged.emit(notification);
        this.list.nativeElement.removeFocus();
    }
    onItemClicked(item) {
        this.overflowMenuHandlers[item.action]();
    }
    onOpenChanged(event) {
        if (!event.detail.value) {
            this.openChange.emit(false);
        }
        else {
            this.notificationList = [...this.notificationList];
        }
    }
    onOverflowMenuClicked() {
        this.list.nativeElement.removeFocus();
    }
    onDetailsClicked() {
        this.open = false;
        this.openDetails.emit();
        this.openChange.emit(false);
    }
    prepareNotifications(notifications) {
        return notifications.map(notification => (Object.assign(Object.assign({}, notification), { label: notification.messageText, sublabel: this.formatDateTime(notification.timestamp), icon: TYPE_ICON_MAPPINGS[notification.type], iconToggleUnchecked: 'mail', iconToggleChecked: 'communication:mail-outline', iconToggleValue: notification.read })));
    }
    formatDateTime(date) {
        return date === null
            ? ''
            : `${AvlLocalizationService.formatDateTime(date, {
                format: 'date-short'
            })} ${AvlLocalizationService.formatDateTime(date, { format: 'time-long' })}`;
    }
    translateLabels() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!AvlLocalizationService || !AvlLocalizationService.translationService) {
                return;
            }
            this.overflowMenuItems = yield this.translateMenu(NOTIFICATIONS_OVERFLOW_MENU);
        });
    }
    translateMenu(menu) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield Promise.all(menu.map((group) => __awaiter(this, void 0, void 0, function* () { return yield this.translateMenuGroup(group); })));
        });
    }
    translateMenuGroup(group) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                items: yield Promise.all(group.items.map((item) => __awaiter(this, void 0, void 0, function* () { return yield this.translateMenuItem(item); })))
            };
        });
    }
    translateMenuItem(item) {
        return __awaiter(this, void 0, void 0, function* () {
            const translateService = AvlLocalizationService.translationService;
            return Object.assign(Object.assign({}, item), { label: yield translateService.translate(item.label), sublabel: item.sublabel ? yield translateService.translate(item.sublabel) : null });
        });
    }
};
AvlNotificationMenuComponent.ɵfac = function AvlNotificationMenuComponent_Factory(t) { return new (t || AvlNotificationMenuComponent)(); };
AvlNotificationMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlNotificationMenuComponent, selectors: [["avl-notification-menu"]], viewQuery: function AvlNotificationMenuComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.list = _t.first);
    } }, inputs: { open: "open", notifications: "notifications", trigger: "trigger" }, outputs: { openChange: "openChange", openDetails: "openDetails", notificationStateChanged: "notificationStateChanged", markAllAsRead: "markAllAsRead", markAllAsUnread: "markAllAsUnread" }, decls: 15, vars: 12, consts: [[3, "noCloseOnInsideClick", "open", "trigger", "open-changed"], ["id", "header"], ["id", "headerText"], [4, "ngIf"], [3, "items", "itemClicked"], ["icon", "more-vert", "slot", "trigger", 3, "click"], [3, "rows", "row-changed"], ["list", ""], ["id", "footer"], [3, "click"]], template: function AvlNotificationMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "avl-popup", 0);
        ɵngcc0.ɵɵlistener("open-changed", function AvlNotificationMenuComponent_Template_avl_popup_open_changed_0_listener($event) { return ctx.onOpenChanged($event); });
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "span");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, AvlNotificationMenuComponent_span_6_Template, 2, 1, "span", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "avl-overflow-menu", 4);
        ɵngcc0.ɵɵlistener("itemClicked", function AvlNotificationMenuComponent_Template_avl_overflow_menu_itemClicked_7_listener($event) { return ctx.onItemClicked($event); });
        ɵngcc0.ɵɵelementStart(8, "avl-icon-button", 5);
        ɵngcc0.ɵɵlistener("click", function AvlNotificationMenuComponent_Template_avl_icon_button_click_8_listener() { return ctx.onOverflowMenuClicked(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "avl-list", 6, 7);
        ɵngcc0.ɵɵlistener("row-changed", function AvlNotificationMenuComponent_Template_avl_list_row_changed_9_listener($event) { return ctx.onNotificationStateChanged($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 8);
        ɵngcc0.ɵɵelementStart(12, "avl-button", 9);
        ɵngcc0.ɵɵlistener("click", function AvlNotificationMenuComponent_Template_avl_button_click_12_listener() { return ctx.onDetailsClicked(); });
        ɵngcc0.ɵɵtext(13);
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("noCloseOnInsideClick", true)("open", ctx.open)("trigger", ctx.trigger);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 8, "VizBox.NotificationMenu.Notifications"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.numberOfUnreadNotifications);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("items", ctx.overflowMenuItems);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("rows", ctx.notificationList);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(14, 10, "VizBox.NotificationMenu.Details"));
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.AvlOverflowMenuDirective, ɵngcc2.AvlInteractionControlDirective, ɵngcc2.AvlIconButtonDirective, ɵngcc2.AvlMultiSelectControlDirective, ɵngcc2.AvlSingleSelectControlDirective, ɵngcc2.AvlListDirective], pipes: [ɵngcc3.TranslatePipe], styles: ["avl-popup[_ngcontent-%COMP%]{width:40%}avl-list[_ngcontent-%COMP%]{width:100%;height:250px;padding-top:8px;--avl-item-padding-left:16px;--avl-item-padding-right:16px;--notification-item-height:48px}#header[_ngcontent-%COMP%]{display:flex;align-items:center;justify-content:space-between;box-sizing:border-box;height:56px;padding:0 16px;border-bottom:1px solid var(--avl-overflow-menu-divider-color,var(--avl-divider-color))}#footer[_ngcontent-%COMP%]{display:flex;align-items:center;box-sizing:border-box;height:56px;border-top:1px solid var(--avl-overflow-menu-divider-color,var(--avl-divider-color))}#headerText[_ngcontent-%COMP%]{font-size:20px;color:var(--avl-primary-text-color);overflow:hidden;white-space:nowrap;text-overflow:ellipsis}"] });
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], AvlNotificationMenuComponent.prototype, "open", void 0);
__decorate([
    Input(),
    __metadata("design:type", HTMLElement)
], AvlNotificationMenuComponent.prototype, "trigger", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array),
    __metadata("design:paramtypes", [Array])
], AvlNotificationMenuComponent.prototype, "notifications", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlNotificationMenuComponent.prototype, "openChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlNotificationMenuComponent.prototype, "openDetails", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlNotificationMenuComponent.prototype, "notificationStateChanged", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlNotificationMenuComponent.prototype, "markAllAsRead", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlNotificationMenuComponent.prototype, "markAllAsUnread", void 0);
__decorate([
    ViewChild('list'),
    __metadata("design:type", ElementRef)
], AvlNotificationMenuComponent.prototype, "list", void 0);
AvlNotificationMenuComponent = __decorate([ __metadata("design:paramtypes", [])
], AvlNotificationMenuComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNotificationMenuComponent, [{
        type: Component,
        args: [{
                selector: 'avl-notification-menu',
                template: "<avl-popup\r\n  [noCloseOnInsideClick]=\"true\"\r\n  [open]=\"open\"\r\n  [trigger]=\"trigger\"\r\n  (open-changed)=\"onOpenChanged($event)\"\r\n>\r\n  <div id=\"header\">\r\n    <div id=\"headerText\">\r\n      <span>{{ 'VizBox.NotificationMenu.Notifications' | translate }}</span>\r\n      <span *ngIf=\"numberOfUnreadNotifications\"> ({{ numberOfUnreadNotifications }}) </span>\r\n    </div>\r\n    <avl-overflow-menu (itemClicked)=\"onItemClicked($event)\" [items]=\"overflowMenuItems\">\r\n      <avl-icon-button\r\n        icon=\"more-vert\"\r\n        slot=\"trigger\"\r\n        (click)=\"onOverflowMenuClicked()\"\r\n      ></avl-icon-button>\r\n    </avl-overflow-menu>\r\n  </div>\r\n  <avl-list\r\n    #list\r\n    [rows]=\"notificationList\"\r\n    (row-changed)=\"onNotificationStateChanged($event)\"\r\n  ></avl-list>\r\n  <div id=\"footer\">\r\n    <avl-button (click)=\"onDetailsClicked()\">{{ 'VizBox.NotificationMenu.Details' | translate }}</avl-button>\r\n  </div>\r\n</avl-popup>\r\n",
                styles: ["avl-popup{width:40%}avl-list{width:100%;height:250px;padding-top:8px;--avl-item-padding-left:16px;--avl-item-padding-right:16px;--notification-item-height:48px}#header{display:flex;align-items:center;justify-content:space-between;box-sizing:border-box;height:56px;padding:0 16px;border-bottom:1px solid var(--avl-overflow-menu-divider-color,var(--avl-divider-color))}#footer{display:flex;align-items:center;box-sizing:border-box;height:56px;border-top:1px solid var(--avl-overflow-menu-divider-color,var(--avl-divider-color))}#headerText{font-size:20px;color:var(--avl-primary-text-color);overflow:hidden;white-space:nowrap;text-overflow:ellipsis}"]
            }]
    }], function () { return []; }, { open: [{
            type: Input
        }], openChange: [{
            type: Output
        }], openDetails: [{
            type: Output
        }], notificationStateChanged: [{
            type: Output
        }], markAllAsRead: [{
            type: Output
        }], markAllAsUnread: [{
            type: Output
        }], notifications: [{
            type: Input
        }], trigger: [{
            type: Input
        }], list: [{
            type: ViewChild,
            args: ['list']
        }] }); })();
export { AvlNotificationMenuComponent };
