// @avl-ng-controls/dialog/esm5/lib/avl-about-dialog/avl-about-dialog.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../avl-dialog/avl-dialog.component';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@avl-controls/ng-binders';
import * as ɵngcc4 from '../avl-dialog/avl-dialog-content/avl-dialog-content.component';

function AvlAboutDialogComponent_ng_template_0_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 21);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var infoLabel_r6 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate2(" ", infoLabel_r6.label, ": ", infoLabel_r6.value, " ");
} }
function AvlAboutDialogComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    var _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "div", 5);
    ɵngcc0.ɵɵelementStart(2, "div", 6);
    ɵngcc0.ɵɵelement(3, "img", 7);
    ɵngcc0.ɵɵelement(4, "img", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 9);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 10);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 11);
    ɵngcc0.ɵɵtemplate(10, AvlAboutDialogComponent_ng_template_0_div_10_Template, 2, 2, "div", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 13);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "div", 14);
    ɵngcc0.ɵɵelementStart(14, "a", 15);
    ɵngcc0.ɵɵtext(15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(16, "div", 16);
    ɵngcc0.ɵɵelementStart(17, "div", 17);
    ɵngcc0.ɵɵelementStart(18, "div", 18);
    ɵngcc0.ɵɵelementStart(19, "div", 19);
    ɵngcc0.ɵɵtext(20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "avl-button", 20);
    ɵngcc0.ɵɵlistener("click", function AvlAboutDialogComponent_ng_template_0_Template_avl_button_click_21_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.closeDialog(); });
    ɵngcc0.ɵɵtext(22);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("src", ctx_r1.productLogo, ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.productName);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.brandPromise);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.versionInfo);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.legalText);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("href", ctx_r1.additionalInfoUrl, ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.additionalInfoLabel);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.copyright);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.okButtonLabel);
} }
function AvlAboutDialogComponent_ng_template_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function AvlAboutDialogComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "avl-dialog-content");
    ɵngcc0.ɵɵtemplate(1, AvlAboutDialogComponent_ng_template_3_ng_container_1_Template, 1, 0, "ng-container", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    var _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0);
} }
function AvlAboutDialogComponent_div_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function AvlAboutDialogComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 23);
    ɵngcc0.ɵɵtemplate(1, AvlAboutDialogComponent_div_5_ng_container_1_Template, 1, 0, "ng-container", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    var _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0);
} }
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, Output, EventEmitter } from '@angular/core';
import '@avl-controls/core/dist/avl-button/avl-button';
var AvlAboutDialogComponent = /** @class */ (function () {
    function AvlAboutDialogComponent() {
        this.okButtonLabel = 'OK';
        this._open = false;
        this.openChange = new EventEmitter();
    }
    Object.defineProperty(AvlAboutDialogComponent.prototype, "open", {
        get: function () {
            return this._open;
        },
        set: function (value) {
            this._open = value;
            this.openChange.emit(this._open);
        },
        enumerable: true,
        configurable: true
    });
    AvlAboutDialogComponent.prototype.closeDialog = function () {
        this.open = false;
    };
    /** @hidden */
    AvlAboutDialogComponent.prototype.openChanged = function (value) {
        this.open = value;
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "productLogo", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "productName", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "brandPromise", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], AvlAboutDialogComponent.prototype, "versionInfo", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "legalText", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "additionalInfoLabel", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "additionalInfoUrl", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AvlAboutDialogComponent.prototype, "copyright", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], AvlAboutDialogComponent.prototype, "okButtonLabel", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], AvlAboutDialogComponent.prototype, "open", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], AvlAboutDialogComponent.prototype, "openChange", void 0);
    AvlAboutDialogComponent = __decorate([ __metadata("design:paramtypes", [])
    ], AvlAboutDialogComponent);
AvlAboutDialogComponent.ɵfac = function AvlAboutDialogComponent_Factory(t) { return new (t || AvlAboutDialogComponent)(); };
AvlAboutDialogComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlAboutDialogComponent, selectors: [["avl-about-dialog"]], inputs: { okButtonLabel: "okButtonLabel", open: "open", productLogo: "productLogo", productName: "productName", brandPromise: "brandPromise", versionInfo: "versionInfo", legalText: "legalText", additionalInfoLabel: "additionalInfoLabel", additionalInfoUrl: "additionalInfoUrl", copyright: "copyright" }, outputs: { openChange: "openChange" }, decls: 6, vars: 2, consts: [["aboutContent", ""], [3, "open", "openChange"], ["content", ""], ["class", "full-screen-about", 4, "ngIf"], [1, "about-content"], [1, "images-holder"], [1, "images-flex-container"], ["alt", "logo", 1, "logo-img", 3, "src"], ["src", "assets/images/avl_logo_112x48.png", "alt", "Avl logo", "id", "avl-logo"], [1, "product-name"], [1, "brand-promise"], [1, "version-info"], ["class", "version-info-item", 4, "ngFor", "ngForOf"], [1, "legal-text"], [1, "additional-info"], ["target", "_blank", "rel", "noopener noreferrer", 3, "href"], [1, "spacer"], [1, "bottom-group"], [1, "bottom-group-flex-container"], [1, "copyright"], ["contained", "", 3, "click"], [1, "version-info-item"], [4, "ngTemplateOutlet"], [1, "full-screen-about"]], template: function AvlAboutDialogComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AvlAboutDialogComponent_ng_template_0_Template, 23, 9, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(2, "avl-dialog", 1);
        ɵngcc0.ɵɵlistener("openChange", function AvlAboutDialogComponent_Template_avl_dialog_openChange_2_listener($event) { return ctx.open = $event; })("openChange", function AvlAboutDialogComponent_Template_avl_dialog_openChange_2_listener($event) { return ctx.openChanged($event); });
        ɵngcc0.ɵɵtemplate(3, AvlAboutDialogComponent_ng_template_3_Template, 2, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, AvlAboutDialogComponent_div_5_Template, 2, 1, "div", 3);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("open", ctx.open);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.open);
    } }, directives: [ɵngcc1.AvlDialogComponent, ɵngcc2.NgIf, ɵngcc2.NgForOf, ɵngcc3.AvlInteractionControlDirective, ɵngcc4.AvlDialogContentComponent, ɵngcc2.NgTemplateOutlet], styles: ["[_nghost-%COMP%]{font-family:var(--avl-font-family);--avl-about-dialog-foreground-color:var(--avl-corporate-foreground-color);--avl-about-dialog-foreground-medium-color:var(--avl-corporate-foreground-medium-color);--avl-about-dialog-link-color:var(--avl-corporate-link-color);--avl-about-dialog-link-hover-color:var(--avl-corporate-link-hover-color);--avl-about-dialog-on-button-background-color:var(--avl-corporate-on-button-background-color);--avl-about-dialog-button-background-color:var(--avl-corporate-button-background-color);--avl-about-dialog-button-background-hover-color:var(--avl-corporate-button-background-hover-color);--avl-about-dialog-button-background-focus-color:var(--avl-corporate-button-background-focus-color);--avl-about-dialog-button-background-pressed-color:var(--avl-corporate-button-background-pressed-color);--dialog-padding:48px;--product-logo-size:148px;--product-name-font-size:34px;--margin-above-button-and-copyright:16px;--margin-above-additional-info:8px;--distance-between-version-info-items:8px}@media (max-width:960px) and (min-width:400px){[_nghost-%COMP%]{--dialog-padding:24px;--margin-above-button-and-copyright:24px}}@media (max-width:400px){[_nghost-%COMP%]{--dialog-padding:16px;--product-logo-size:64px;--product-name-font-size:24px;--distance-to-link-for-additional-info:24px;--margin-above-additional-info:4px;--distance-between-version-info-items:4px}}.full-screen-about[_ngcontent-%COMP%]{display:none;position:fixed;top:0;left:0;bottom:0;right:0;overflow:auto;z-index:3000}@media (max-width:864px){avl-dialog[_ngcontent-%COMP%]{display:none}.full-screen-about[_ngcontent-%COMP%]{display:block}}avl-dialog[_ngcontent-%COMP%]{--avl-dialog-width:768px;--avl-dialog-text-color:var(--avl-about-dialog-foreground-medium-color);--avl-dialog-background-color:rgba(46, 46, 46, 1)}.about-content[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:flex-start;padding:var(--dialog-padding);box-sizing:border-box;width:100%;height:100%;font-size:14px;line-height:19px;background-color:#2e2e2e;overflow:auto}avl-dialog-content[_ngcontent-%COMP%]{padding:0}.logo-img[_ngcontent-%COMP%]{height:var(--product-logo-size);width:auto}.product-name[_ngcontent-%COMP%]{font-size:var(--product-name-font-size);line-height:1;margin-top:24px;color:var(--avl-about-dialog-foreground-color)}.brand-promise[_ngcontent-%COMP%]{font-size:16px;line-height:1;margin-top:8px;color:var(--avl-about-dialog-foreground-color)}.version-info[_ngcontent-%COMP%]{margin-top:24px;color:rgba(255,255,255,.7)}.version-info-item[_ngcontent-%COMP%]{margin-bottom:var(--distance-between-version-info-items)}.version-info-item[_ngcontent-%COMP%]:last-child{margin-bottom:0}.legal-text[_ngcontent-%COMP%]{margin-top:24px;color:var(--avl-about-dialog-foreground-color)}.additional-info[_ngcontent-%COMP%]{margin-top:var(--margin-above-additional-info);color:var(--avl-corporate-link)}.additional-info[_ngcontent-%COMP%]:hover{color:var(--avl-corporate-link-hover)}.copyright[_ngcontent-%COMP%]{color:var(--avl-about-dialog-foreground-color);display:inline-block}avl-button[_ngcontent-%COMP%]{--avl-button-label-color:var(--avl-about-dialog-on-button-background-color);--avl-button-background-default-color:var(--avl-about-dialog-button-background-color);--avl-button-background-hover-color:var(--avl-about-dialog-button-background-hover-color);--avl-button-background-focus-color:var(--avl-about-dialog-button-background-focus-color);--avl-button-background-pressed-color:var(--avl-about-dialog-button-background-pressed-color);font-size:14px;margin:-8px;margin-left:calc(48px - 8px)}#avl-logo[_ngcontent-%COMP%]{margin-right:calc(-1 * var(--dialog-padding));height:48px;width:112px;align-self:flex-start}.images-flex-container[_ngcontent-%COMP%]{display:flex;justify-content:space-between}.spacer[_ngcontent-%COMP%]{flex-grow:1}.bottom-group[_ngcontent-%COMP%]{margin-top:var(--margin-above-button-and-copyright)}.bottom-group-flex-container[_ngcontent-%COMP%]{display:flex;justify-content:space-between;align-items:flex-end}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlAboutDialogComponent, [{
        type: Component,
        args: [{
                selector: 'avl-about-dialog',
                template: "<ng-template #aboutContent>\r\n  <div class=\"about-content\">\r\n    <div class=\"images-holder\">\r\n      <div class=\"images-flex-container\">\r\n        <img src=\"{{productLogo}}\" class=\"logo-img\" alt=\"logo\">\r\n        <img src=\"assets/images/avl_logo_112x48.png\" alt=\"Avl logo\" id=\"avl-logo\">\r\n      </div>\r\n    </div>\r\n    <div class=\"product-name\">{{productName}}</div>\r\n    <div class=\"brand-promise\">{{brandPromise}}</div>\r\n    <div class=\"version-info\">\r\n      <div  class=\"version-info-item\" *ngFor=\"let infoLabel of versionInfo\">\r\n        {{infoLabel.label}}: {{infoLabel.value}}\r\n      </div>\r\n    </div>\r\n    <div class=\"legal-text\">{{legalText}}</div>\r\n    <div class=\"additional-info\">\r\n      <a href=\"{{additionalInfoUrl}}\" target=\"_blank\" rel=\"noopener noreferrer\">{{additionalInfoLabel}}</a>\r\n    </div>\r\n    <div class=\"spacer\"></div>\r\n    <div class=\"bottom-group\">\r\n      <div class=\"bottom-group-flex-container\">\r\n        <div class=\"copyright\">{{copyright}}</div>\r\n        <avl-button (click)=\"closeDialog()\" contained>{{ okButtonLabel }}</avl-button>\r\n      </div>\r\n    </div>\r\n  </div>\r\n  </ng-template> \r\n\r\n<avl-dialog [(open)]=\"open\" (openChange)=\"openChanged($event)\">\r\n  <ng-template #content>\r\n    <avl-dialog-content>\r\n      <ng-container *ngTemplateOutlet=\"aboutContent\"></ng-container>\r\n    </avl-dialog-content>\r\n  </ng-template>\r\n</avl-dialog>\r\n\r\n<div *ngIf=\"open\" class=\"full-screen-about\">\r\n  <ng-container *ngTemplateOutlet=\"aboutContent\"></ng-container>\r\n</div>\r\n",
                styles: [":host(){font-family:var(--avl-font-family);--avl-about-dialog-foreground-color:var(--avl-corporate-foreground-color);--avl-about-dialog-foreground-medium-color:var(--avl-corporate-foreground-medium-color);--avl-about-dialog-link-color:var(--avl-corporate-link-color);--avl-about-dialog-link-hover-color:var(--avl-corporate-link-hover-color);--avl-about-dialog-on-button-background-color:var(--avl-corporate-on-button-background-color);--avl-about-dialog-button-background-color:var(--avl-corporate-button-background-color);--avl-about-dialog-button-background-hover-color:var(--avl-corporate-button-background-hover-color);--avl-about-dialog-button-background-focus-color:var(--avl-corporate-button-background-focus-color);--avl-about-dialog-button-background-pressed-color:var(--avl-corporate-button-background-pressed-color);--dialog-padding:48px;--product-logo-size:148px;--product-name-font-size:34px;--margin-above-button-and-copyright:16px;--margin-above-additional-info:8px;--distance-between-version-info-items:8px}@media (max-width:960px) and (min-width:400px){:host(){--dialog-padding:24px;--margin-above-button-and-copyright:24px}}@media (max-width:400px){:host(){--dialog-padding:16px;--product-logo-size:64px;--product-name-font-size:24px;--distance-to-link-for-additional-info:24px;--margin-above-additional-info:4px;--distance-between-version-info-items:4px}}.full-screen-about{display:none;position:fixed;top:0;left:0;bottom:0;right:0;overflow:auto;z-index:3000}@media (max-width:864px){avl-dialog{display:none}.full-screen-about{display:block}}avl-dialog{--avl-dialog-width:768px;--avl-dialog-text-color:var(--avl-about-dialog-foreground-medium-color);--avl-dialog-background-color:rgba(46, 46, 46, 1)}.about-content{display:flex;flex-direction:column;justify-content:flex-start;padding:var(--dialog-padding);box-sizing:border-box;width:100%;height:100%;font-size:14px;line-height:19px;background-color:#2e2e2e;overflow:auto}avl-dialog-content{padding:0}.logo-img{height:var(--product-logo-size);width:auto}.product-name{font-size:var(--product-name-font-size);line-height:1;margin-top:24px;color:var(--avl-about-dialog-foreground-color)}.brand-promise{font-size:16px;line-height:1;margin-top:8px;color:var(--avl-about-dialog-foreground-color)}.version-info{margin-top:24px;color:rgba(255,255,255,.7)}.version-info-item{margin-bottom:var(--distance-between-version-info-items)}.version-info-item:last-child{margin-bottom:0}.legal-text{margin-top:24px;color:var(--avl-about-dialog-foreground-color)}.additional-info{margin-top:var(--margin-above-additional-info);color:var(--avl-corporate-link)}.additional-info:hover{color:var(--avl-corporate-link-hover)}.copyright{color:var(--avl-about-dialog-foreground-color);display:inline-block}avl-button{--avl-button-label-color:var(--avl-about-dialog-on-button-background-color);--avl-button-background-default-color:var(--avl-about-dialog-button-background-color);--avl-button-background-hover-color:var(--avl-about-dialog-button-background-hover-color);--avl-button-background-focus-color:var(--avl-about-dialog-button-background-focus-color);--avl-button-background-pressed-color:var(--avl-about-dialog-button-background-pressed-color);font-size:14px;margin:-8px;margin-left:calc(48px - 8px)}#avl-logo{margin-right:calc(-1 * var(--dialog-padding));height:48px;width:112px;align-self:flex-start}.images-flex-container{display:flex;justify-content:space-between}.spacer{flex-grow:1}.bottom-group{margin-top:var(--margin-above-button-and-copyright)}.bottom-group-flex-container{display:flex;justify-content:space-between;align-items:flex-end}"]
            }]
    }], function () { return []; }, { okButtonLabel: [{
            type: Input
        }], openChange: [{
            type: Output
        }], open: [{
            type: Input
        }], productLogo: [{
            type: Input
        }], productName: [{
            type: Input
        }], brandPromise: [{
            type: Input
        }], versionInfo: [{
            type: Input
        }], legalText: [{
            type: Input
        }], additionalInfoLabel: [{
            type: Input
        }], additionalInfoUrl: [{
            type: Input
        }], copyright: [{
            type: Input
        }] }); })();
    return AvlAboutDialogComponent;
}());
export { AvlAboutDialogComponent };
