// @avl-ng-controls/dialog/esm2015/lib/avl-dialog/avl-dialog-footer/avl-dialog-footer.component.js
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
export class AvlDialogFooterComponent {
}
AvlDialogFooterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogFooterComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
AvlDialogFooterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.3", type: AvlDialogFooterComponent, selector: "avl-dialog-footer", ngImport: i0, template: "<ng-content></ng-content>", styles: [":host{display:flex;flex-grow:0;flex-shrink:0;color:var(--avl-primary-text-color);padding:6px 16px 16px 8px;flex-wrap:wrap;align-items:center;justify-content:flex-end;overflow:auto}:host(.complex){border-top:1px solid var(--avl-dialog-separator-color);padding:2px 16px}avl-dialog.fullscreen :host{justify-content:flex-start}"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogFooterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'avl-dialog-footer',
                    templateUrl: './avl-dialog-footer.component.html',
                    styleUrls: ['./avl-dialog-footer.component.css']
                }]
        }] });
