// @avl-ng-controls/dialog/esm2015/lib/avl-dialog/avl-dialog-footer/avl-dialog-footer.component.js
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component } from '@angular/core';
import { AvlDialogComponent } from '../avl-dialog.component';
import { AvlBaseDialogSection } from '../avl-dialog-section.base';
let AvlDialogFooterComponent = class AvlDialogFooterComponent extends AvlBaseDialogSection {
    constructor(dialog) {
        super(dialog);
        this.dialog = dialog;
    }
};
AvlDialogFooterComponent.ɵfac = function AvlDialogFooterComponent_Factory(t) { return new (t || AvlDialogFooterComponent)(ɵngcc0.ɵɵdirectiveInject(AvlDialogComponent)); };
AvlDialogFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlDialogFooterComponent, selectors: [["avl-dialog-footer"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function AvlDialogFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, styles: ["[_nghost-%COMP%]{display:flex;flex-grow:0;flex-shrink:0;color:var(--avl-primary-text-color);padding:6px 16px 16px 8px;flex-wrap:wrap;align-items:center;justify-content:flex-end;overflow:auto}.complex[_nghost-%COMP%]{border-top:1px solid var(--avl-dialog-separator-color);padding:2px 16px}avl-dialog.fullscreen   [_nghost-%COMP%]{justify-content:flex-start}"] });
AvlDialogFooterComponent.ctorParameters = () => [
    { type: AvlDialogComponent }
];
AvlDialogFooterComponent = __decorate([ __metadata("design:paramtypes", [AvlDialogComponent])
], AvlDialogFooterComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlDialogFooterComponent, [{
        type: Component,
        args: [{
                selector: 'avl-dialog-footer',
                template: "<ng-content></ng-content>",
                styles: [":host(){display:flex;flex-grow:0;flex-shrink:0;color:var(--avl-primary-text-color);padding:6px 16px 16px 8px;flex-wrap:wrap;align-items:center;justify-content:flex-end;overflow:auto}:host(.complex){border-top:1px solid var(--avl-dialog-separator-color);padding:2px 16px}avl-dialog.fullscreen :host{justify-content:flex-start}"]
            }]
    }], function () { return [{ type: AvlDialogComponent }]; }, null); })();
export { AvlDialogFooterComponent };
