// @avl-ng-controls/dialog/esm2015/lib/dialog.module.js
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { A11yModule } from '@angular/cdk/a11y';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
// components
import { AvlDialogComponent } from './avl-dialog/avl-dialog.component';
import { AvlLoginComponent } from './avl-login/avl-login.component';
import { AvlNgBindersModule } from '@avl-controls/ng-binders';
import { AvlDialogHeaderComponent } from './avl-dialog/avl-dialog-header/avl-dialog-header.component';
import { AvlDialogFooterComponent } from './avl-dialog/avl-dialog-footer/avl-dialog-footer.component';
import { AvlDialogContentComponent } from './avl-dialog/avl-dialog-content/avl-dialog-content.component';
import { AvlAboutDialogComponent } from './avl-about-dialog/avl-about-dialog.component';
import { AvlNavigationModule } from '@avl-ng-controls/navigation';
import { AvlCommonModule } from '@avl-ng-controls/common';
import { AvlUnsavedChangesDialogComponent } from './avl-unsaved-changes-dialog/avl-unsaved-changes-dialog.component';
import * as i0 from "@angular/core";
const componentsAndDirectives = [
    AvlDialogComponent,
    AvlLoginComponent,
    AvlDialogHeaderComponent,
    AvlDialogFooterComponent,
    AvlDialogContentComponent,
    AvlAboutDialogComponent,
    AvlUnsavedChangesDialogComponent
];
const providers = [];
export class AvlDialogModule {
}
AvlDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AvlDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogModule, declarations: [AvlDialogComponent,
        AvlLoginComponent,
        AvlDialogHeaderComponent,
        AvlDialogFooterComponent,
        AvlDialogContentComponent,
        AvlAboutDialogComponent,
        AvlUnsavedChangesDialogComponent], imports: [CommonModule,
        AvlNgBindersModule,
        AvlNavigationModule,
        AvlCommonModule,
        A11yModule,
        FormsModule,
        ReactiveFormsModule], exports: [AvlDialogComponent,
        AvlLoginComponent,
        AvlDialogHeaderComponent,
        AvlDialogFooterComponent,
        AvlDialogContentComponent,
        AvlAboutDialogComponent,
        AvlUnsavedChangesDialogComponent] });
AvlDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogModule, providers: providers, imports: [[
            CommonModule,
            AvlNgBindersModule,
            AvlNavigationModule,
            AvlCommonModule,
            A11yModule,
            FormsModule,
            ReactiveFormsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        AvlNgBindersModule,
                        AvlNavigationModule,
                        AvlCommonModule,
                        A11yModule,
                        FormsModule,
                        ReactiveFormsModule
                    ],
                    declarations: componentsAndDirectives,
                    providers: providers,
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: componentsAndDirectives
                }]
        }] });
