// @avl-ng-controls/auth/esm5/lib/services/avl-auth.service.js
import { BehaviorSubject } from 'rxjs';
var AvlAuthService = /** @class */ (function () {
    function AvlAuthService() {
        /**
         * User data is held in this property
         */
        this.userSubject = new BehaviorSubject(undefined);
        /**
         * Current user as Observable
         */
        this.user$ = this.userSubject.asObservable();
        /**
         * String holding error message in case user couldn't authenticate
         */
        this.errorMessageSubject = new BehaviorSubject(undefined);
        /**
         * Error messsage in case failed authentication as Observable
         */
        this.errorMessage$ = this.errorMessageSubject.asObservable();
    }
    Object.defineProperty(AvlAuthService.prototype, "currentUser", {
        /**
         * Currently authenticated user data
         */
        get: function () {
            return this.userSubject.value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AvlAuthService.prototype, "errorMessage", {
        /**
         * Current error message for fail authentication
         * Undefined when user is logged in
         */
        get: function () {
            return this.errorMessageSubject.value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Use to clear error message after user is informed
     */
    AvlAuthService.prototype.clearErrorMessage = function () {
        this.errorMessageSubject.next(undefined);
    };
    return AvlAuthService;
}());
export { AvlAuthService };
