// @avl-ng-controls/auth/esm5/lib/services/token-storage.js
import { StorageNames } from './storage-names';
var TokenStorage = /** @class */ (function () {
    function TokenStorage(prefix) {
        if (prefix === void 0) { prefix = ''; }
        this.prefix = '';
        this.prefix = prefix;
    }
    TokenStorage.prototype.readUser = function () {
        var roles = JSON.parse(this.getItem(StorageNames.Roles));
        return {
            username: this.getItem(StorageNames.UserName),
            name: this.getItem(StorageNames.Name),
            access_token: this.getItem(StorageNames.AccessToken),
            access_token_expires: this.getNumber(StorageNames.AccessTokenExpiresAt),
            access_token_stored: this.getNumber(StorageNames.AccessTokenStoredAt),
            refresh_token: this.getItem(StorageNames.RefreshToken),
            refresh_token_expires: this.getNumber(StorageNames.RefreshTokenExpiresAt),
            id_token: this.getItem(StorageNames.IdToken),
            roles: roles
        };
    };
    TokenStorage.prototype.storeUser = function (user) {
        this.setItem(StorageNames.UserName, user.username);
        this.setItem(StorageNames.Name, user.name);
        this.setItem(StorageNames.AccessToken, user.access_token);
        this.setItem(StorageNames.AccessTokenExpiresAt, "" + user.access_token_expires);
        this.setItem(StorageNames.AccessTokenStoredAt, "" + user.access_token_stored);
        this.setItem(StorageNames.RefreshToken, user.refresh_token);
        this.setItem(StorageNames.RefreshTokenExpiresAt, "" + user.refresh_token_expires);
        this.setItem(StorageNames.IdToken, user.id_token);
        var rolesStr = JSON.stringify(user.roles);
        sessionStorage.setItem(StorageNames.Roles, "" + rolesStr);
    };
    TokenStorage.prototype.clearUser = function () {
        for (var name_1 in StorageNames) {
            if (StorageNames.hasOwnProperty(name_1)) {
                this.removeItem(StorageNames[name_1]);
            }
        }
    };
    TokenStorage.prototype.getNumber = function (key) {
        return parseInt(this.getItem(key), 10);
    };
    TokenStorage.prototype.getItem = function (key) {
        return sessionStorage.getItem(this.prefix + key);
    };
    TokenStorage.prototype.setItem = function (key, value) {
        sessionStorage.setItem(this.prefix + key, value);
    };
    TokenStorage.prototype.removeItem = function (key) {
        sessionStorage.removeItem(this.prefix + key);
    };
    return TokenStorage;
}());
export { TokenStorage };
