// @avl-ng-controls/navigation/esm5/lib/components/avl-app-bar/avl-app-bar-item.component.js
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, OnInit, ElementRef } from '@angular/core';
var AvlAppBarItemComponent = /** @class */ (function () {
    function AvlAppBarItemComponent(elementRef) {
        this.elementRef = elementRef;
    }
    /**
     * @hidden
     */
    AvlAppBarItemComponent.prototype.ngOnInit = function () {
        this.elementRef.nativeElement.classList.add('mdc-toolbar__icon');
    };
    AvlAppBarItemComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AvlAppBarItemComponent = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlAppBarItemComponent);
AvlAppBarItemComponent.ɵfac = function AvlAppBarItemComponent_Factory(t) { return new (t || AvlAppBarItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlAppBarItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlAppBarItemComponent, selectors: [["avl-app-bar-item"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function AvlAppBarItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlAppBarItemComponent, [{
        type: Component,
        args: [{
                selector: 'avl-app-bar-item',
                template: '<ng-content></ng-content>'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
    return AvlAppBarItemComponent;
}());
export { AvlAppBarItemComponent };
