// @avl-ng-controls/navigation/esm5/lib/components/avl-app-layout/avl-no-padding.directive.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { OnInit, Directive, OnDestroy, ElementRef } from '@angular/core';
var AvlNoPaddingDirective = /** @class */ (function () {
    function AvlNoPaddingDirective(elementRef) {
        this.elementRef = elementRef;
    }
    /** @hidden */
    AvlNoPaddingDirective.prototype.ngOnInit = function () {
        var appLayout = this.findAppLayout();
        if (appLayout) {
            appLayout.setAttribute('noPadding', '');
        }
    };
    /** @hidden */
    AvlNoPaddingDirective.prototype.ngOnDestroy = function () {
        var appLayout = this.findAppLayout();
        if (appLayout) {
            appLayout.removeAttribute('noPadding');
        }
    };
    AvlNoPaddingDirective.prototype.findAppLayout = function () {
        var element = this.elementRef.nativeElement;
        while (element) {
            if (element.tagName.toLowerCase() == 'avl-app-layout') {
                return element;
            }
            element = element.parentElement;
        }
    };
    AvlNoPaddingDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AvlNoPaddingDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlNoPaddingDirective);
AvlNoPaddingDirective.ɵfac = function AvlNoPaddingDirective_Factory(t) { return new (t || AvlNoPaddingDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlNoPaddingDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlNoPaddingDirective, selectors: [["", "avl-no-padding", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNoPaddingDirective, [{
        type: Directive,
        args: [{
                selector: '[avl-no-padding]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
    return AvlNoPaddingDirective;
}());
export { AvlNoPaddingDirective };
