// @avl-ng-controls/navigation/esm5/lib/components/avl-app-layout/avl-app-layout.component.js
import * as ɵngcc0 from '@angular/core';

var _c0 = [[["", "slot", "toolbar"]], [["", "slot", "left-side"]], "*", [["", "slot", "right-side"]]];
var _c1 = ["[slot=toolbar]", "[slot=left-side]", "*", "[slot=right-side]"];
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ViewEncapsulation, ContentChild } from '@angular/core';
import { AvlSideNavComponent } from '../avl-side-nav/avl-side-nav.component';
import { AvlAppBarComponent } from '../avl-app-bar/avl-app-bar.component';
var AvlAppLayoutComponent = /** @class */ (function () {
    function AvlAppLayoutComponent() {
    }
    AvlAppLayoutComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (this.sideNav && this.appBar) {
            this.appBar.burgerClicked.subscribe(function () { return _this.sideNav.open = !_this.sideNav.open; });
        }
    };
    __decorate([
        ContentChild(AvlSideNavComponent),
        __metadata("design:type", AvlSideNavComponent)
    ], AvlAppLayoutComponent.prototype, "sideNav", void 0);
    __decorate([
        ContentChild(AvlAppBarComponent),
        __metadata("design:type", AvlAppBarComponent)
    ], AvlAppLayoutComponent.prototype, "appBar", void 0);
    AvlAppLayoutComponent = __decorate([ __metadata("design:paramtypes", [])
    ], AvlAppLayoutComponent);
AvlAppLayoutComponent.ɵfac = function AvlAppLayoutComponent_Factory(t) { return new (t || AvlAppLayoutComponent)(); };
AvlAppLayoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlAppLayoutComponent, selectors: [["avl-app-layout"]], contentQueries: function AvlAppLayoutComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, AvlSideNavComponent, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, AvlAppBarComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.sideNav = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.appBar = _t.first);
    } }, ngContentSelectors: _c1, decls: 6, vars: 0, consts: [[1, "sideNavAndContent"]], template: function AvlAppLayoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵprojection(2, 1);
        ɵngcc0.ɵɵelementStart(3, "main");
        ɵngcc0.ɵɵprojection(4, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(5, 3);
        ɵngcc0.ɵɵelementEnd();
    } }, styles: [".sideNavAndContent[_ngcontent-%COMP%]{display:flex;flex-direction:row;flex-flow:row nowrap;height:100%;padding-bottom:4px;position:relative}.scrim[_ngcontent-%COMP%]{display:none;visibility:hidden}main[_ngcontent-%COMP%]{--app-bar-height:var(--avl-app-bar-height, 53px);color:var(--avl-primary-text-color);width:100%;padding:var(--avl-app-layout-content-padding,24px);overflow:auto;margin-top:var(--app-bar-height)}[noPadding][_nghost-%COMP%]   main[_ngcontent-%COMP%]{padding:0}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlAppLayoutComponent, [{
        type: Component,
        args: [{
                selector: 'avl-app-layout',
                template: "<ng-content select=\"[slot=toolbar]\"></ng-content>\r\n<div class=\"sideNavAndContent\">\r\n  <ng-content select=\"[slot=left-side]\"></ng-content>\r\n  <main>\r\n    <ng-content></ng-content>\r\n  </main>\r\n  <ng-content select=\"[slot=right-side]\"></ng-content>\r\n</div>\r\n",
                encapsulation: ViewEncapsulation.Emulated,
                styles: [".sideNavAndContent{display:flex;flex-direction:row;flex-flow:row nowrap;height:100%;padding-bottom:4px;position:relative}.scrim{display:none;visibility:hidden}main{--app-bar-height:var(--avl-app-bar-height, 53px);color:var(--avl-primary-text-color);width:100%;padding:var(--avl-app-layout-content-padding,24px);overflow:auto;margin-top:var(--app-bar-height)}:host([noPadding]) main{padding:0}"]
            }]
    }], function () { return []; }, { sideNav: [{
            type: ContentChild,
            args: [AvlSideNavComponent]
        }], appBar: [{
            type: ContentChild,
            args: [AvlAppBarComponent]
        }] }); })();
    return AvlAppLayoutComponent;
}());
export { AvlAppLayoutComponent };
