// @avl-ng-controls/dialog/esm2015/lib/avl-dialog/avl-dialog-content/avl-dialog-content.component.js
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
export class AvlDialogContentComponent {
}
AvlDialogContentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogContentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
AvlDialogContentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.3", type: AvlDialogContentComponent, selector: "avl-dialog-content", ngImport: i0, template: "<ng-content></ng-content>", styles: [":host{flex-grow:1;flex-direction:column;font-size:16px;line-height:21px;color:var(--avl-dialog-text-color);padding-top:0;padding-right:24px;padding-right:var(--padding-right,24px);padding-bottom:28px;padding-bottom:var(--padding-bottom,28px);padding-left:24px;padding-left:var(--padding-left,24px);overflow:auto}:host(.complex){padding-top:8px;padding-top:var(--padding-top,8px)}:host(.warning){padding:13px 16px 0 23px}"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlDialogContentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'avl-dialog-content',
                    templateUrl: './avl-dialog-content.component.html',
                    styleUrls: ['./avl-dialog-content.component.css']
                }]
        }] });
