// @avl-ng-controls/navigation/esm2015/lib/components/avl-list-expander/avl-list-expander.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function AvlListExpanderComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelement(1, "avl-icon", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r0.narrow ? "icon-narrow" : "icon");
} }
function AvlListExpanderComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelement(1, "avl-icon", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.narrow ? "icon-narrow" : "icon");
} }
function AvlListExpanderComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵprojection(1, 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.narrow ? "content-narrow" : "content");
} }
const _c0 = [[["div", "role", "header"]], [["div", "role", "content"]]];
const _c1 = ["div[role=header]", "div[role=content]"];
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, HostListener, Output, EventEmitter } from '@angular/core';
let AvlListExpanderComponent = class AvlListExpanderComponent {
    constructor() {
        /**
         * Boolean that changes modes between wide and narrow, false by default.
         */
        this.narrow = false;
        this._open = false;
        /**
         * Event that emits when the expander is opened or closed.
         * Sends boolean value of new open state.
         */
        this.openChange = new EventEmitter();
    }
    /**
     * Boolean that tells whether the expander is open or closed, expander is closed by default.
     */
    set open(value) {
        this._open = value;
        this.openChange.emit(this._open);
    }
    get open() {
        return this._open;
    }
    toggleClick() {
        this.open = !this.open;
    }
    /** @hidden */
    keyEvent(event) {
        const activeEl = document.activeElement;
        if (event.key === ' ' || event.key === 'Enter') {
            activeEl.click();
            return;
        }
    }
};
AvlListExpanderComponent.ɵfac = function AvlListExpanderComponent_Factory(t) { return new (t || AvlListExpanderComponent)(); };
AvlListExpanderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlListExpanderComponent, selectors: [["avl-list-expander"]], hostBindings: function AvlListExpanderComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown", function AvlListExpanderComponent_keydown_HostBindingHandler($event) { return ctx.keyEvent($event); });
    } }, inputs: { narrow: "narrow", open: "open" }, outputs: { openChange: "openChange" }, ngContentSelectors: _c1, decls: 8, vars: 5, consts: [["id", "expander"], ["tabindex", "0", 3, "ngClass", "click"], [3, "ngClass", 4, "ngIf"], [3, "ngClass"], ["icon", "hardware:keyboard-arrow-down", "id", "starsIcon"], ["icon", "hardware:keyboard-arrow-up", "id", "starsIcon"]], template: function AvlListExpanderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelement(0, "hr");
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵlistener("click", function AvlListExpanderComponent_Template_div_click_2_listener() { return ctx.toggleClick(); });
        ɵngcc0.ɵɵtemplate(3, AvlListExpanderComponent_div_3_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵtemplate(4, AvlListExpanderComponent_div_4_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, AvlListExpanderComponent_div_7_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ctx.narrow ? "title-container-narrow" : "title-container");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.open);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.open);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ctx.narrow ? "title-narrow" : "title");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.open);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgIf], styles: [".title-container[_ngcontent-%COMP%]:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container[_ngcontent-%COMP%]:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container[_ngcontent-%COMP%]:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}.title-container[_ngcontent-%COMP%]{height:48px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.title[_ngcontent-%COMP%]{font-size:17px;color:var(--avl-list-expander-title-color,var(--avl-foreground-color))}.icon[_ngcontent-%COMP%]{margin:12px 12px 12px 24px;height:24px}.content[_ngcontent-%COMP%]{margin:4px 24px}hr[_ngcontent-%COMP%]{display:block;border:0;border-top:1px solid var(--avl-list-expander-divider-color,var(--avl-divider-color));margin:0}avl-icon[_ngcontent-%COMP%]{--avl-icon-custom-color:var(--avl-list-expander-icon-color, var(--avl-hint-color))}.icon-narrow[_ngcontent-%COMP%]{margin:16px 24px 16px 16px;position:absolute;right:0;bottom:0;height:24px}.title-container-narrow[_ngcontent-%COMP%]{position:relative;height:56px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.content-narrow[_ngcontent-%COMP%]{margin:0 24px 24px}.title-narrow[_ngcontent-%COMP%]{margin-left:24px;font-size:17px;margin-right:64px;white-space:nowrap;overflow:hidden}.title-container-narrow[_ngcontent-%COMP%]:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container-narrow[_ngcontent-%COMP%]:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container-narrow[_ngcontent-%COMP%]:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}"] });
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], AvlListExpanderComponent.prototype, "narrow", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], AvlListExpanderComponent.prototype, "open", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlListExpanderComponent.prototype, "openChange", void 0);
__decorate([
    HostListener('keydown', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [KeyboardEvent]),
    __metadata("design:returntype", void 0)
], AvlListExpanderComponent.prototype, "keyEvent", null);
AvlListExpanderComponent = __decorate([ __metadata("design:paramtypes", [])
], AvlListExpanderComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlListExpanderComponent, [{
        type: Component,
        args: [{
                selector: 'avl-list-expander',
                template: "<hr/>\r\n<div id=\"expander\">\r\n    <div  [ngClass]=\"narrow ? 'title-container-narrow' : 'title-container'\" tabindex=\"0\" (click)=\"toggleClick()\">\r\n        <div [ngClass]=\"narrow ? 'icon-narrow' : 'icon'\" *ngIf=\"!open\">\r\n            <avl-icon icon=\"hardware:keyboard-arrow-down\" id=\"starsIcon\"></avl-icon>\r\n        </div>\r\n        <div [ngClass]=\"narrow ? 'icon-narrow' : 'icon'\" *ngIf=\"open\">\r\n            <avl-icon icon=\"hardware:keyboard-arrow-up\" id=\"starsIcon\"></avl-icon>\r\n        </div>\r\n        <div [ngClass]=\"narrow ? 'title-narrow' : 'title'\">\r\n            <ng-content select=\"div[role=header]\"></ng-content>\r\n        </div>\r\n    </div>\r\n\r\n    <div *ngIf=\"open\" [ngClass]=\"narrow ? 'content-narrow' : 'content'\">\r\n        <ng-content select=\"div[role=content]\"></ng-content>\r\n    </div>\r\n</div>\r\n\r\n",
                styles: [".title-container:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}.title-container{height:48px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.title{font-size:17px;color:var(--avl-list-expander-title-color,var(--avl-foreground-color))}.icon{margin:12px 12px 12px 24px;height:24px}.content{margin:4px 24px}hr{display:block;border:0;border-top:1px solid var(--avl-list-expander-divider-color,var(--avl-divider-color));margin:0}avl-icon{--avl-icon-custom-color:var(--avl-list-expander-icon-color, var(--avl-hint-color))}.icon-narrow{margin:16px 24px 16px 16px;position:absolute;right:0;bottom:0;height:24px}.title-container-narrow{position:relative;height:56px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.content-narrow{margin:0 24px 24px}.title-narrow{margin-left:24px;font-size:17px;margin-right:64px;white-space:nowrap;overflow:hidden}.title-container-narrow:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container-narrow:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container-narrow:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}"]
            }]
    }], function () { return []; }, { narrow: [{
            type: Input
        }], openChange: [{
            type: Output
        }], open: [{
            type: Input
        }], keyEvent: [{
            type: HostListener,
            args: ['keydown', ['$event']]
        }] }); })();
export { AvlListExpanderComponent };
