// @avl-ng-controls/navigation/esm2015/lib/components/avl-app-layout/avl-app-layout.component.js
import { Component, ViewEncapsulation, ContentChild } from '@angular/core';
import { AvlSideNavComponent } from '../avl-side-nav/avl-side-nav.component';
import { AvlAppBarComponent } from '../avl-app-bar/avl-app-bar.component';
import * as i0 from "@angular/core";
export class AvlAppLayoutComponent {
    constructor() { }
    ngAfterContentInit() {
        if (this.sideNav && this.appBar) {
            this.appBar.burgerClicked.subscribe(() => this.sideNav.open = !this.sideNav.open);
        }
    }
}
AvlAppLayoutComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlAppLayoutComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
AvlAppLayoutComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.3", type: AvlAppLayoutComponent, selector: "avl-app-layout", queries: [{ propertyName: "sideNav", first: true, predicate: AvlSideNavComponent, descendants: true }, { propertyName: "appBar", first: true, predicate: AvlAppBarComponent, descendants: true }], ngImport: i0, template: "<ng-content select=\"[slot=toolbar]\"></ng-content>\r\n<div class=\"sideNavAndContent\">\r\n  <ng-content select=\"[slot=left-side]\"></ng-content>\r\n  <main>\r\n    <ng-content></ng-content>\r\n  </main>\r\n  <ng-content select=\"[slot=right-side]\"></ng-content>\r\n</div>\r\n", styles: [".sideNavAndContent{display:flex;flex-flow:row nowrap;height:100%;padding-bottom:4px;position:relative}.scrim{display:none;visibility:hidden}main{--app-bar-height:var(--avl-app-bar-height,53px);color:var(--avl-primary-text-color);width:100%;padding:24px;padding:var(--avl-app-layout-content-padding,24px);overflow:auto;margin-top:var(--app-bar-height);position:relative}:host([noPadding]) main{padding:0}"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlAppLayoutComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'avl-app-layout',
                    templateUrl: './avl-app-layout.component.html',
                    styleUrls: [
                        './avl-app-layout.component.css'
                    ],
                    encapsulation: ViewEncapsulation.Emulated
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { sideNav: [{
                type: ContentChild,
                args: [AvlSideNavComponent]
            }], appBar: [{
                type: ContentChild,
                args: [AvlAppBarComponent]
            }] } });
