// @avl-controls/ng-binders/esm2015/lib/interface/avl-single-select-control.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
/** Directive for controls implementing [AvlSingleSelectControl]. */
import * as ɵngcc0 from '@angular/core';
let AvlSingleSelectControlDirective = class AvlSingleSelectControlDirective {
    constructor(elementRef) {
        this.selectedIndexChange = new EventEmitter();
        this.selectedValueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'selected-index-changed', this.selectedIndexChange);
        addEventListenerForEmittingChangedEvents(elementRef, 'selected-value-changed', this.selectedValueChange);
    }
};
AvlSingleSelectControlDirective.ɵfac = function AvlSingleSelectControlDirective_Factory(t) { return new (t || AvlSingleSelectControlDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlSingleSelectControlDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlSingleSelectControlDirective, selectors: [["avl-datatable"], ["avl-dropdown"], ["avl-list"], ["avl-radio-group"], ["avl-app-switcher"], ["avl-chip-list"], ["avl-combo-box"]], outputs: { selectedIndexChange: "selectedIndexChange", selectedValueChange: "selectedValueChange" } });
AvlSingleSelectControlDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlSingleSelectControlDirective.prototype, "selectedIndexChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlSingleSelectControlDirective.prototype, "selectedValueChange", void 0);
AvlSingleSelectControlDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlSingleSelectControlDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSingleSelectControlDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-datatable,avl-dropdown,avl-list,avl-radio-group,avl-app-switcher,avl-chip-list,avl-combo-box'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { selectedIndexChange: [{
            type: Output
        }], selectedValueChange: [{
            type: Output
        }] }); })();
export { AvlSingleSelectControlDirective };
