// @avl-controls/ng-binders/esm2015/lib/interface/avl-control.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
/** Directive for controls implementing [AvlControl]. */
import * as ɵngcc0 from '@angular/core';
let AvlControlDirective = class AvlControlDirective {
    constructor(elementRef) {
        this.disabledChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'disabled-changed', this.disabledChange);
    }
};
AvlControlDirective.ɵfac = function AvlControlDirective_Factory(t) { return new (t || AvlControlDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlControlDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlControlDirective, selectors: [["avl-breadcrumb"], ["avl-context-menu"], ["avl-header-bar"], ["avl-menu-button"], ["avl-user-info"]], outputs: { disabledChange: "disabledChange" } });
AvlControlDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlControlDirective.prototype, "disabledChange", void 0);
AvlControlDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlControlDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlControlDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-breadcrumb,avl-context-menu,avl-header-bar,avl-menu-button,avl-user-info'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { disabledChange: [{
            type: Output
        }] }); })();
export { AvlControlDirective };
