// @avl-ng-controls/dialog/esm5/lib/avl-dialog/avl-dialog-content/avl-dialog-content.component.js
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component } from '@angular/core';
import { AvlDialogComponent } from '../avl-dialog.component';
import { AvlBaseDialogSection } from '../avl-dialog-section.base';
var AvlDialogContentComponent = /** @class */ (function (_super) {
    __extends(AvlDialogContentComponent, _super);
    function AvlDialogContentComponent(dialog) {
        var _this = _super.call(this, dialog) || this;
        _this.dialog = dialog;
        return _this;
    }
    AvlDialogContentComponent.ctorParameters = function () { return [
        { type: AvlDialogComponent }
    ]; };
    AvlDialogContentComponent = __decorate([ __metadata("design:paramtypes", [AvlDialogComponent])
    ], AvlDialogContentComponent);
AvlDialogContentComponent.ɵfac = function AvlDialogContentComponent_Factory(t) { return new (t || AvlDialogContentComponent)(ɵngcc0.ɵɵdirectiveInject(AvlDialogComponent)); };
AvlDialogContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlDialogContentComponent, selectors: [["avl-dialog-content"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function AvlDialogContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, styles: ["[_nghost-%COMP%]{flex-grow:1;flex-direction:column;font-size:16px;line-height:21px;color:var(--avl-dialog-text-color);padding-top:0;padding-right:var(--padding-right,24px);padding-bottom:var(--padding-bottom,28px);padding-left:var(--padding-left,24px);overflow:auto}.complex[_nghost-%COMP%]{padding-top:var(--padding-top,8px)}.warning[_nghost-%COMP%]{padding:13px 16px 0 23px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlDialogContentComponent, [{
        type: Component,
        args: [{
                selector: 'avl-dialog-content',
                template: "<ng-content></ng-content>",
                styles: [":host(){flex-grow:1;flex-direction:column;font-size:16px;line-height:21px;color:var(--avl-dialog-text-color);padding-top:0;padding-right:var(--padding-right,24px);padding-bottom:var(--padding-bottom,28px);padding-left:var(--padding-left,24px);overflow:auto}:host(.complex){padding-top:var(--padding-top,8px)}:host(.warning){padding:13px 16px 0 23px}"]
            }]
    }], function () { return [{ type: AvlDialogComponent }]; }, null); })();
    return AvlDialogContentComponent;
}(AvlBaseDialogSection));
export { AvlDialogContentComponent };
