// @avl-ng-controls/notification/esm5/lib/components/avl-notification-dialog/avl-notification-dialog.consts.js
var _a, _b, _c;
import { ColumnType, LeafColumn } from '@avl-controls/datatable/dist/avl-datatable/column-definitions';
export var DEFAULT_PAGE_SIZE_OPTIONS = [
    { value: 15, label: 'Show 15 per page' },
    { value: 50, label: 'Show 50 per page' },
    { value: Number.MAX_SAFE_INTEGER, label: 'Show all' }
];
export var DEFAULT_FILTERS = {
    type: {
        warning: undefined,
        error: undefined,
        info: undefined,
        ok: undefined
    },
    state: {
        read: undefined,
        unread: undefined
    },
    date: {
        dateFromValue: undefined,
        dateToValue: undefined
    }
};
export var DEFAULT_COLUMN_DEFINITIONS = [
    new LeafColumn('VizBox.NotificationDialog.Type', 'icon', ColumnType.Icon, false, false, 0.5),
    new LeafColumn('VizBox.NotificationDialog.Module', 'module', ColumnType.Text, true, false, 3),
    new LeafColumn('VizBox.NotificationDialog.Message', 'messageText', ColumnType.Text, true, false, 8),
    new LeafColumn('VizBox.NotificationDialog.Id', 'id', ColumnType.Number, false, false, 1),
    {
        caption: 'VizBox.NotificationDialog.Timestamp',
        valueField: 'timestamp',
        type: ColumnType.Date,
        sortable: false,
        filterable: false,
        flexWidth: 3,
        dateTimeOptions: { format: 'date-time-short' }
    },
    new LeafColumn('VizBox.NotificationDialog.State', 'currentIconToggle', ColumnType.Icon, false, false, 1)
];
export var NOTIFICATION_TYPE_MAPPINGS = (_a = {},
    _a[0] = 'error',
    _a[1] = 'warning',
    _a[2] = 'info',
    _a[3] = 'ok',
    _a.false = 'unread',
    _a.true = 'read',
    _a);
export var TYPE_ICON_MAPPINGS = (_b = {},
    _b[0 /* Error */] = 'cancel',
    _b[1 /* Warning */] = 'warning',
    _b[2 /* Info */] = 'info-outline',
    _b[3 /* OK */] = 'check',
    _b);
export var STATE_ICON_MAPPINGS = {
    false: 'mail',
    true: 'communication:mail-outline'
};
export var DEFAULT_ICON_STYLE = 'height:24px; width:24px;';
export var TYPE_ICON_STYLE_MAPPINGS = (_c = {},
    _c[0 /* Error */] = "--avl-icon-custom-color: var(--avl-error-color); " + DEFAULT_ICON_STYLE,
    _c[1 /* Warning */] = "--avl-icon-custom-color: var(--avl-warning-color); " + DEFAULT_ICON_STYLE,
    _c[2 /* Info */] = "--avl-icon-custom-color: var(--avl-icon-default-color); " + DEFAULT_ICON_STYLE,
    _c[3 /* OK */] = "--avl-icon-custom-color: var(--avl-ok-color); " + DEFAULT_ICON_STYLE,
    _c);
