// @avl-ng-controls/dialog/esm2015/lib/avl-about-dialog/avl-about-dialog.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, Output, EventEmitter } from '@angular/core';
import '@avl-controls/core/dist/avl-button/avl-button';
let AvlAboutDialogComponent = class AvlAboutDialogComponent {
    constructor() {
        this.okButtonLabel = 'OK';
        this._open = false;
        this.openChange = new EventEmitter();
    }
    set open(value) {
        this._open = value;
        this.openChange.emit(this._open);
    }
    get open() {
        return this._open;
    }
    closeDialog() {
        this.open = false;
    }
    /** @hidden */
    openChanged(value) {
        this.open = value;
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "productLogo", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "productName", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "brandPromise", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], AvlAboutDialogComponent.prototype, "versionInfo", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "legalText", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "additionalInfoLabel", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "additionalInfoUrl", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlAboutDialogComponent.prototype, "copyright", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], AvlAboutDialogComponent.prototype, "okButtonLabel", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], AvlAboutDialogComponent.prototype, "open", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlAboutDialogComponent.prototype, "openChange", void 0);
AvlAboutDialogComponent = __decorate([
    Component({
        selector: 'avl-about-dialog',
        template: "<ng-template #aboutContent>\r\n  <div class=\"about-content\">\r\n    <div class=\"images-holder\">\r\n      <div class=\"images-flex-container\">\r\n        <img src=\"{{productLogo}}\" class=\"logo-img\" alt=\"logo\">\r\n        <img src=\"assets/images/avl_logo_112x48.png\" alt=\"Avl logo\" id=\"avl-logo\">\r\n      </div>\r\n    </div>\r\n    <div class=\"product-name\">{{productName}}</div>\r\n    <div class=\"brand-promise\">{{brandPromise}}</div>\r\n    <div class=\"version-info\">\r\n      <div  class=\"version-info-item\" *ngFor=\"let infoLabel of versionInfo\">\r\n        {{infoLabel.label}}: {{infoLabel.value}}\r\n      </div>\r\n    </div>\r\n    <div class=\"legal-text\">{{legalText}}</div>\r\n    <div class=\"additional-info\">\r\n      <a href=\"{{additionalInfoUrl}}\" target=\"_blank\" rel=\"noopener noreferrer\">{{additionalInfoLabel}}</a>\r\n    </div>\r\n    <div class=\"spacer\"></div>\r\n    <div class=\"bottom-group\">\r\n      <div class=\"bottom-group-flex-container\">\r\n        <div class=\"copyright\">{{copyright}}</div>\r\n        <avl-button (click)=\"closeDialog()\" contained>{{ okButtonLabel }}</avl-button>\r\n      </div>\r\n    </div>\r\n  </div>\r\n  </ng-template> \r\n\r\n<avl-dialog [(open)]=\"open\" (openChange)=\"openChanged($event)\">\r\n  <ng-template #content>\r\n    <avl-dialog-content>\r\n      <ng-container *ngTemplateOutlet=\"aboutContent\"></ng-container>\r\n    </avl-dialog-content>\r\n  </ng-template>\r\n</avl-dialog>\r\n\r\n<div *ngIf=\"open\" class=\"full-screen-about\">\r\n  <ng-container *ngTemplateOutlet=\"aboutContent\"></ng-container>\r\n</div>\r\n",
        styles: [":host(){font-family:var(--avl-font-family);--avl-about-dialog-foreground-color:var(--avl-corporate-foreground-color);--avl-about-dialog-foreground-medium-color:var(--avl-corporate-foreground-medium-color);--avl-about-dialog-link-color:var(--avl-corporate-link-color);--avl-about-dialog-link-hover-color:var(--avl-corporate-link-hover-color);--avl-about-dialog-on-button-background-color:var(--avl-corporate-on-button-background-color);--avl-about-dialog-button-background-color:var(--avl-corporate-button-background-color);--avl-about-dialog-button-background-hover-color:var(--avl-corporate-button-background-hover-color);--avl-about-dialog-button-background-focus-color:var(--avl-corporate-button-background-focus-color);--avl-about-dialog-button-background-pressed-color:var(--avl-corporate-button-background-pressed-color);--dialog-padding:48px;--product-logo-size:148px;--product-name-font-size:34px;--margin-above-button-and-copyright:16px;--margin-above-additional-info:8px;--distance-between-version-info-items:8px}@media (max-width:960px) and (min-width:400px){:host(){--dialog-padding:24px;--margin-above-button-and-copyright:24px}}@media (max-width:400px){:host(){--dialog-padding:16px;--product-logo-size:64px;--product-name-font-size:24px;--distance-to-link-for-additional-info:24px;--margin-above-additional-info:4px;--distance-between-version-info-items:4px}}.full-screen-about{display:none;position:fixed;top:0;left:0;bottom:0;right:0;overflow:auto;z-index:3000}@media (max-width:864px){avl-dialog{display:none}.full-screen-about{display:block}}avl-dialog{--avl-dialog-width:768px;--avl-dialog-text-color:var(--avl-about-dialog-foreground-medium-color);--avl-dialog-background-color:rgba(46, 46, 46, 1)}.about-content{display:flex;flex-direction:column;justify-content:flex-start;padding:var(--dialog-padding);box-sizing:border-box;width:100%;height:100%;font-size:14px;line-height:19px;background-color:#2e2e2e;overflow:auto}avl-dialog-content{padding:0}.logo-img{height:var(--product-logo-size);width:auto}.product-name{font-size:var(--product-name-font-size);line-height:1;margin-top:24px;color:var(--avl-about-dialog-foreground-color)}.brand-promise{font-size:16px;line-height:1;margin-top:8px;color:var(--avl-about-dialog-foreground-color)}.version-info{margin-top:24px;color:rgba(255,255,255,.7)}.version-info-item{margin-bottom:var(--distance-between-version-info-items)}.version-info-item:last-child{margin-bottom:0}.legal-text{margin-top:24px;color:var(--avl-about-dialog-foreground-color)}.additional-info{margin-top:var(--margin-above-additional-info);color:var(--avl-corporate-link)}.additional-info:hover{color:var(--avl-corporate-link-hover)}.copyright{color:var(--avl-about-dialog-foreground-color);display:inline-block}avl-button{--avl-button-label-color:var(--avl-about-dialog-on-button-background-color);--avl-button-background-default-color:var(--avl-about-dialog-button-background-color);--avl-button-background-hover-color:var(--avl-about-dialog-button-background-hover-color);--avl-button-background-focus-color:var(--avl-about-dialog-button-background-focus-color);--avl-button-background-pressed-color:var(--avl-about-dialog-button-background-pressed-color);font-size:14px;margin:-8px;margin-left:calc(48px - 8px)}#avl-logo{margin-right:calc(-1 * var(--dialog-padding));height:48px;width:112px;align-self:flex-start}.images-flex-container{display:flex;justify-content:space-between}.spacer{flex-grow:1}.bottom-group{margin-top:var(--margin-above-button-and-copyright)}.bottom-group-flex-container{display:flex;justify-content:space-between;align-items:flex-end}"]
    }),
    __metadata("design:paramtypes", [])
], AvlAboutDialogComponent);
export { AvlAboutDialogComponent };
