// @avl-ng-controls/navigation/esm2015/lib/components/avl-solution-overview/avl-solution-overview.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@avl-controls/ng-binders';
import * as ɵngcc3 from '@angular/router';

const _c0 = ["searchbar"];
const _c1 = ["iconButtons"];
const _c2 = ["searchbarContainer"];
const _c3 = ["groupsContainer"];
const _c4 = ["itemElement"];
const _c5 = ["group"];
function AvlSolutionOverviewComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "h2", 17);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.solutionSubtitle);
} }
function AvlSolutionOverviewComponent_avl_search_bar_14_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "avl-search-bar", 18, 19);
    ɵngcc0.ɵɵlistener("query-changed", function AvlSolutionOverviewComponent_avl_search_bar_14_Template_avl_search_bar_query_changed_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onQueryChange($event); })("search", function AvlSolutionOverviewComponent_avl_search_bar_14_Template_avl_search_bar_search_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onSearch($event); });
    ɵngcc0.ɵɵelementEnd();
} }
function AvlSolutionOverviewComponent_ng_container_17_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c6 = function (a0) { return { group: a0 }; };
function AvlSolutionOverviewComponent_ng_container_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 20);
    ɵngcc0.ɵɵelementStart(2, "div", 21, 22);
    ɵngcc0.ɵɵtemplate(4, AvlSolutionOverviewComponent_ng_container_17_ng_container_4_Template, 1, 0, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const group_r14 = ctx.$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    const _r6 = ɵngcc0.ɵɵreference(19);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r5.getMaxColumnsClass());
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r6)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c6, group_r14));
} }
function AvlSolutionOverviewComponent_ng_template_18_h2_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h2", 26);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(group_r17.title);
} }
function AvlSolutionOverviewComponent_ng_template_18_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(group_r17.header);
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c7 = function (a0) { return { "with-description": a0 }; };
const _c8 = function (a0) { return { item: a0 }; };
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 29, 30);
    ɵngcc0.ɵɵtemplate(3, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_ng_container_3_Template, 1, 0, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r24 = ɵngcc0.ɵɵnextContext().$implicit;
    const group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵnextContext();
    const _r8 = ɵngcc0.ɵɵreference(21);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("routerLink", item_r24.route)("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c7, group_r17.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r8)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(6, _c8, item_r24));
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 31, 30);
    ɵngcc0.ɵɵtemplate(3, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_ng_container_3_Template, 1, 0, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r24 = ɵngcc0.ɵɵnextContext().$implicit;
    const group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵnextContext();
    const _r8 = ɵngcc0.ɵɵreference(21);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(6, _c7, group_r17.description));
    ɵngcc0.ɵɵattribute("href", item_r24.link, ɵngcc0.ɵɵsanitizeUrl)("target", item_r24.target)("rel", item_r24.target == "_blank" ? "noopener" : null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r8)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(8, _c8, item_r24));
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_Template, 4, 8, "ng-container", 28);
    ɵngcc0.ɵɵtemplate(2, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_Template, 4, 10, "ng-container", 28);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r24 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r24.route);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r24.link);
} }
function AvlSolutionOverviewComponent_ng_template_18_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(group_r17.footer);
} }
function AvlSolutionOverviewComponent_ng_template_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, AvlSolutionOverviewComponent_ng_template_18_h2_0_Template, 2, 1, "h2", 24);
    ɵngcc0.ɵɵtemplate(1, AvlSolutionOverviewComponent_ng_template_18_div_1_Template, 2, 1, "div", 25);
    ɵngcc0.ɵɵtemplate(2, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_Template, 3, 2, "ng-container", 14);
    ɵngcc0.ɵɵtemplate(3, AvlSolutionOverviewComponent_ng_template_18_div_3_Template, 2, 1, "div", 25);
} if (rf & 2) {
    const group_r17 = ctx.group;
    ɵngcc0.ɵɵproperty("ngIf", group_r17.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", group_r17.header);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", group_r17.items);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", group_r17.footer);
} }
function AvlSolutionOverviewComponent_ng_template_20_avl_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-icon", 39);
} if (rf & 2) {
    const item_r36 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵproperty("icon", item_r36.icon);
} }
function AvlSolutionOverviewComponent_ng_template_20_img_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 40);
} if (rf & 2) {
    const item_r36 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵpropertyInterpolate("src", item_r36.iconUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
const _c9 = function (a0) { return { "no-icon-circle": a0 }; };
function AvlSolutionOverviewComponent_ng_template_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 32);
    ɵngcc0.ɵɵtemplate(1, AvlSolutionOverviewComponent_ng_template_20_avl_icon_1_Template, 1, 1, "avl-icon", 33);
    ɵngcc0.ɵɵtemplate(2, AvlSolutionOverviewComponent_ng_template_20_img_2_Template, 1, 1, "img", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 35);
    ɵngcc0.ɵɵelementStart(4, "span", 36);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 37);
    ɵngcc0.ɵɵelementStart(7, "span", 38);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r36 = ctx.item;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c9, item_r36.noIconCircle));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r36.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r36.iconUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(item_r36.label);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r36.description, " ");
} }
const _c10 = function (a0) { return { "top-margin": a0 }; };
const _c11 = ["*"];
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, HostListener, ElementRef, ViewChildren, QueryList, Input, EventEmitter, Output, ViewChild, AfterViewInit, OnInit, OnChanges, SimpleChanges, OnDestroy, ChangeDetectorRef, } from '@angular/core';
import { BehaviorSubject, asyncScheduler } from 'rxjs';
import { throttleTime, map } from 'rxjs/operators';
import { KeyboardNavigator } from './keyboard-navigator';
import '@avl-controls/core/dist/avl-icon/avl-icon';
import '@avl-controls/core/dist/avl-search-bar/avl-search-bar';
let AvlSolutionOverviewComponent = class AvlSolutionOverviewComponent {
    constructor(cdr) {
        this.cdr = cdr;
        /**
         * Emits search query string on change
         */
        this.searchQueryChange = new EventEmitter();
        /**
         * Emits value of the search query when search triggered
         */
        this.search = new EventEmitter();
        this.gridWidth$ = new BehaviorSubject(-1);
        this.largestItemsIndex = -1;
        this.maxItems = 0;
    }
    get groupsOfItems() {
        let v = [];
        for (const gRef of this.groupElementRefs) {
            const gEl = gRef.nativeElement;
            const items = Array.prototype.slice.call(gEl.querySelectorAll('a.item'));
            v = [...v, items];
        }
        return v;
    }
    get iconButtonElements() {
        if (this.iconButtonElementsCache == null) {
            this.iconButtonElementsCache = Array.prototype.slice.call(this.iconButtonsContainer
                .nativeElement.querySelectorAll('avl-icon-button'));
        }
        return this.iconButtonElementsCache;
    }
    ngOnInit() {
        this.setupReferenceVars();
        this.gridWidthClass$ = this.gridWidth$.asObservable().pipe(throttleTime(50, asyncScheduler, { trailing: true }), map(gridWidth => {
            return `grid-width-${gridWidth}`;
        }));
    }
    ngOnDestroy() {
        if (this.groupElementRefsChangesSubscription) {
            this.groupElementRefsChangesSubscription.unsubscribe();
        }
    }
    ngOnChanges(changes) {
        if (changes.groups && !changes.groups.isFirstChange()) {
            this.setupReferenceVars();
            this.gridWidth$.next(this.getGridWidth());
            this.setTopMargin();
        }
    }
    /** @hidden */
    ngAfterViewInit() {
        this.setTopMargin();
        this.setupFocusOrdering();
        this.gridWidth$.next(this.getGridWidth());
    }
    /** @hidden */
    getMaxColumnsClass() {
        if (this.maxColumns && window.outerWidth === this.screenWidth) {
            return `max-columns-${this.maxColumns}`;
        }
        if (this.maxItems === 0) {
            return;
        }
        let maxColumns;
        switch (this.maxItems) {
            case 1:
                maxColumns = 1;
                break;
            case 2:
            case 4:
                maxColumns = 2;
                break;
            case 3:
            case 5:
            case 6:
            case 9:
                maxColumns = 3;
                break;
            case 7:
            case 8:
            case 11:
            case 12:
                maxColumns = 4;
                break;
            default:
                maxColumns = 5;
        }
        this.maxColumns = maxColumns;
        return `max-columns-${maxColumns}`;
    }
    /** @hidden **/
    onResize() {
        this.gridWidth$.next(this.getGridWidth());
    }
    /** @hidden */
    onKeydownHandler(event) {
        const gOfItems = this.groupsOfItems;
        const numGroups = gOfItems.length;
        if (numGroups === 0) {
            return;
        }
        const activeEl = document.activeElement;
        const isDown = ['Down', 'ArrowDown'].includes(event.key);
        const isUp = ['Up', 'ArrowUp'].includes(event.key);
        const isRight = ['Right', 'ArrowRight'].includes(event.key);
        const isLeft = ['Left', 'ArrowLeft'].includes(event.key);
        const isTab = event.key === 'Tab';
        const searchbar = this.showSearchBar
            ? this.searchbarRef.nativeElement
            : undefined;
        let gNum;
        let iIdx;
        for (const items of gOfItems) {
            if (items.indexOf(activeEl) !== -1) {
                iIdx = items.indexOf(activeEl);
                gNum = gOfItems.indexOf(items);
                // Go to route on 'Space'. 'Enter' already works on <a>.
                if (event.key === ' ') {
                    items[iIdx].click();
                    return;
                }
            }
        }
        if (this.iconButtonElements.includes(activeEl)) {
            const numIconButtons = this.iconButtonElements.length;
            const idx = this.iconButtonElements.indexOf(activeEl);
            if ((isRight || isTab) && idx === numIconButtons - 1) {
                if (searchbar != null) {
                    searchbar.focus();
                }
                else {
                    gOfItems[0][0].focus();
                }
            }
            else if (isLeft && idx === 0) {
                const g = gOfItems[numGroups - 1];
                const gLen = g.length;
                g[gLen - 1].focus();
            }
            else if (isLeft) {
                this.iconButtonElements[idx - 1].focus();
            }
            else if (isRight) {
                this.iconButtonElements[idx + 1].focus();
            }
            event.preventDefault();
        }
        const itemElements = gOfItems[gNum];
        const gridWidth = this.gridWidth$.value;
        if (searchbar != null && iIdx < gridWidth && gNum === 0 && isUp) {
            searchbar.focus();
            event.preventDefault();
        }
        else if ((isRight || isTab) &&
            gNum === gOfItems.length - 1 &&
            iIdx === itemElements.length - 1) {
            const iconButtons = this.iconButtonElements;
            if (iconButtons != null && iconButtons.length > 0) {
                iconButtons[0].focus();
                event.preventDefault();
            }
        }
        else if (searchbar != null && searchbar === activeEl && isDown) {
            gOfItems[0][0].focus();
            event.preventDefault();
        }
        else if (gNum != null) {
            const nextGroup = gNum === gOfItems.length - 1 ? undefined : gOfItems[gNum + 1];
            const prevGroup = gNum === 0 ? undefined : gOfItems[gNum - 1];
            const nGLen = nextGroup == null ? 0 : nextGroup.length;
            const pGLen = prevGroup == null ? 0 : prevGroup.length;
            const keyboardNavigator = new KeyboardNavigator(itemElements.length, gridWidth);
            const itemIndexToMoveTo = keyboardNavigator.navigate(iIdx, event.key, nGLen, pGLen);
            if (itemIndexToMoveTo != null) {
                if ((isUp || isLeft) && itemIndexToMoveTo < 0) {
                    prevGroup[itemIndexToMoveTo + pGLen].focus();
                }
                else if ((isDown || isRight) && iIdx + 1 > itemIndexToMoveTo) {
                    nextGroup[itemIndexToMoveTo].focus();
                }
                else {
                    gOfItems[gNum][itemIndexToMoveTo].focus();
                }
                event.preventDefault();
            }
        }
    }
    /** @hidden */
    onQueryChange(event) {
        this.searchQueryChange.emit(event.detail.value);
    }
    /** @hidden */
    onSearch(event) {
        this.search.emit(event.detail.value);
    }
    setupReferenceVars() {
        let max = 0;
        for (const g of this.groups) {
            const itemSz = g.items.length;
            if (max < itemSz) {
                max = itemSz;
                this.largestItemsIndex = this.groups.indexOf(g);
            }
        }
        this.maxItems = max;
    }
    setTopMargin() {
        this.contentHasScroll = false;
        this.cdr.detectChanges();
        const content = this.groupsContainer.nativeElement;
        this.contentHasScroll = content.scrollHeight > content.clientHeight;
        this.cdr.detectChanges();
    }
    getGridWidth() {
        if (this.largestItemsIndex < 0) {
            this.setupReferenceVars();
        }
        const items = this.groupsOfItems[this.largestItemsIndex];
        const firstItemOffsetTop = items[0].offsetTop;
        let i = 0;
        for (const item of items) {
            if (item.offsetTop > firstItemOffsetTop) {
                return i;
            }
            i++;
        }
        return items.length;
    }
    setupFocusOrdering() {
        const buttonElements = this.buttonRefs.map(elRef => elRef.nativeElement);
        let focusableElements = [];
        if (this.showSearchBar) {
            focusableElements.push(this.searchbarRef.nativeElement);
        }
        focusableElements = [
            ...focusableElements,
            ...buttonElements,
            ...this.iconButtonElements
        ];
        focusableElements.forEach((el) => {
            el.addEventListener('focus', (event) => {
                const t = event.target;
                if (t.tagName.toLowerCase() === 'avl-icon-button') {
                    if (this.showSearchBar && (['body', 'html'].includes(document.activeElement.tagName.toLowerCase()) ||
                        this.lastFocusedEl == null ||
                        (this.lastFocusedEl.tagName.toLowerCase() !== 'avl-icon-button' &&
                            this.lastFocusedEl !== buttonElements[buttonElements.length - 1]))) {
                        this.searchbarRef.nativeElement.focus();
                        event.preventDefault();
                    }
                }
                this.lastFocusedEl = event.target;
            });
        });
    }
};
AvlSolutionOverviewComponent.ɵfac = function AvlSolutionOverviewComponent_Factory(t) { return new (t || AvlSolutionOverviewComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
AvlSolutionOverviewComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlSolutionOverviewComponent, selectors: [["avl-solution-overview"]], viewQuery: function AvlSolutionOverviewComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
        ɵngcc0.ɵɵviewQuery(_c1, true);
        ɵngcc0.ɵɵviewQuery(_c2, true);
        ɵngcc0.ɵɵviewQuery(_c3, true);
        ɵngcc0.ɵɵviewQuery(_c4, true);
        ɵngcc0.ɵɵviewQuery(_c5, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchbarRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.iconButtonsContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchbarContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.groupsContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.buttonRefs = _t);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.groupElementRefs = _t);
    } }, hostBindings: function AvlSolutionOverviewComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("resize", function AvlSolutionOverviewComponent_resize_HostBindingHandler() { return ctx.onResize(); }, false, ɵngcc0.ɵɵresolveWindow)("keydown", function AvlSolutionOverviewComponent_keydown_HostBindingHandler($event) { return ctx.onKeydownHandler($event); });
    } }, inputs: { groups: "groups", solutionName: "solutionName", solutionIconUrl: "solutionIconUrl", solutionSubtitle: "solutionSubtitle", showSearchBar: "showSearchBar" }, outputs: { searchQueryChange: "searchQueryChange", search: "search" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c11, decls: 22, vars: 11, consts: [[1, "container"], [1, "top-bar"], [1, "row"], [1, "solution-logo", 3, "src"], [1, "solution-name"], [1, "action-buttons"], ["iconButtons", ""], ["class", "row", 4, "ngIf"], [1, "outer-container", 3, "ngClass"], [1, "searchbar-container", 3, "ngClass"], ["searchbarContainer", ""], ["class", "search-bar", "always-open", "", 3, "query-changed", "search", 4, "ngIf"], [1, "groups"], ["groupsContainer", ""], [4, "ngFor", "ngForOf"], ["groupTpl", ""], ["buttonContent", ""], [1, "solution-subtitle"], ["always-open", "", 1, "search-bar", 3, "query-changed", "search"], ["searchbar", ""], [1, "group-container"], [1, "item-container", 3, "ngClass"], ["group", ""], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], ["class", "group-name", 4, "ngIf"], ["class", "text-container", 4, "ngIf"], [1, "group-name"], [1, "text-container"], [4, "ngIf"], ["tabindex", "0", 1, "item", 3, "routerLink", "ngClass"], ["itemElement", ""], ["tabindex", "0", 1, "item", 3, "ngClass"], [1, "icon-container", 3, "ngClass"], ["class", "item-icon", 3, "icon", 4, "ngIf"], ["class", "item-icon", 3, "src", 4, "ngIf"], [1, "icon-label"], [1, "label"], [1, "description"], [1, "text"], [1, "item-icon", 3, "icon"], [1, "item-icon", 3, "src"]], template: function AvlSolutionOverviewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "img", 3);
        ɵngcc0.ɵɵelementStart(4, "h1", 4);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5, 6);
        ɵngcc0.ɵɵprojection(8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, AvlSolutionOverviewComponent_div_9_Template, 3, 1, "div", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 8);
        ɵngcc0.ɵɵelementStart(11, "div", 9, 10);
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵtemplate(14, AvlSolutionOverviewComponent_avl_search_bar_14_Template, 2, 0, "avl-search-bar", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "div", 12, 13);
        ɵngcc0.ɵɵtemplate(17, AvlSolutionOverviewComponent_ng_container_17_Template, 5, 5, "ng-container", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(18, AvlSolutionOverviewComponent_ng_template_18_Template, 4, 4, "ng-template", null, 15, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(20, AvlSolutionOverviewComponent_ng_template_20_Template, 9, 7, "ng-template", null, 16, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("src", ctx.solutionIconUrl, ɵngcc0.ɵɵsanitizeUrl);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.solutionName);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.solutionSubtitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c10, ctx.contentHasScroll === false));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpipeBind1(13, 7, ctx.gridWidthClass$));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showSearchBar);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.groups);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc1.NgForOf, ɵngcc2.AvlSearchBarDirective, ɵngcc2.AvlFormControlDirective, ɵngcc2.AvlInteractionControlDirective, ɵngcc1.NgTemplateOutlet, ɵngcc3.RouterLinkWithHref], pipes: [ɵngcc1.AsyncPipe], styles: ["[_nghost-%COMP%]{--avl-solution-overview-background-color:var(--avl-background-1-color);--avl-solution-overview-background-image:auto;--avl-solution-overview-text-color:var(--avl-foreground-color);--avl-solution-overview-title-color:var(--avl-foreground-color);--avl-solution-overview-subtitle-color:var(--avl-foreground-medium-color);--avl-solution-overview-item-background-color:var(\r\n    --avl-container-background-color\r\n  );--avl-solution-overview-item-hover-color:var(\r\n    --avl-container-background-product-hover-color,\r\n  );--avl-solution-overview-item-focus-color:var(\r\n    --avl-container-background-product-focus-color\r\n  );--avl-solution-overview-item-pressed-color:var(\r\n    --avl-container-background-product-pressed-color\r\n  );--avl-solution-overview-item-icon-circle-color:var(--avl-product-color);--avl-solution-overview-item-icon-color:var(--avl-on-primary-color);--avl-solution-overview-item-icon-color-no-circle:var(--avl-hint-color);--avl-solution-overview-item-description-color:var(\r\n    --avl-foreground-medium-color\r\n  );--avl-solution-overview-item-icon-pressed-color:var(\r\n    --avl-solution-overview-item-icon-color\r\n  );display:block;width:100%;--item-width:184px;--item-margin:24px;--total-width:calc(var(--item-width) + var(--item-margin));--total-height:100vh;height:var(--total-height);font-family:var(--avl-font-family)}.container[_ngcontent-%COMP%]{background-image:var(--avl-solution-overview-background-image);background-size:cover;background-color:var(--avl-solution-overview-background-color);height:100%;display:flex;flex-direction:column}.groups[_ngcontent-%COMP%]{margin-bottom:24px;width:99%;overflow-y:auto}.group-container[_ngcontent-%COMP%]{display:flex;width:100%;justify-content:center;position:relative}.outer-container[_ngcontent-%COMP%]{display:flex;align-items:center;flex-direction:column;justify-content:center;overflow:hidden;height:100%}.outer-container.top-margin[_ngcontent-%COMP%]{margin-top:-50px}.searchbar-container[_ngcontent-%COMP%]{display:flex;justify-content:space-between}.search-bar[_ngcontent-%COMP%]{padding-right:20px}.external-link-container[_ngcontent-%COMP%], .item-container[_ngcontent-%COMP%]{display:grid;width:100%;margin:0 12px 2px;padding-top:24px;grid-template-columns:repeat(auto-fill,var(--item-width)) [end];grid-gap:var(--item-margin);justify-content:center}.text-container[_ngcontent-%COMP%]{color:var(--avl-solution-overview-text-color);grid-column:1/end}.max-columns-1[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 1)}.max-columns-2[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 2)}.max-columns-3[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 3)}.max-columns-4[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 4)}.max-columns-5[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 5)}.grid-width-1[_ngcontent-%COMP%]{width:calc(var(--total-width) * 1)}.grid-width-2[_ngcontent-%COMP%]{width:calc(var(--total-width) * 2)}.grid-width-3[_ngcontent-%COMP%]{width:calc(var(--total-width) * 3)}.grid-width-4[_ngcontent-%COMP%]{width:calc(var(--total-width) * 4)}.grid-width-5[_ngcontent-%COMP%]{width:calc(var(--total-width) * 5)}.item[_ngcontent-%COMP%]{position:relative;display:flex;flex-direction:column;align-items:center;padding:12px;width:var(--item-width);height:112px;box-sizing:border-box;background-color:var(--avl-solution-overview-item-background-color);outline:0;box-shadow:var(--avl-box-shadow-1dp);transition:box-shadow .15s cubic-bezier(.4,0,.2,1)}.item.with-description[_ngcontent-%COMP%]{height:204px}.description[_ngcontent-%COMP%]{display:none}.item.with-description[_ngcontent-%COMP%]   .description[_ngcontent-%COMP%]{overflow:hidden;display:block;font-weight:400;text-align:center;font-size:14px;line-height:16px;color:var(--avl-foreground-medium-color)}.item[_ngcontent-%COMP%]:active, .item[_ngcontent-%COMP%]:focus, .item[_ngcontent-%COMP%]:hover{box-shadow:var(--avl-box-shadow-8dp)}.item[_ngcontent-%COMP%]:focus{background-color:var(--avl-solution-overview-item-focus-color)}.item[_ngcontent-%COMP%]:hover{background-color:var(--avl-solution-overview-item-hover-color)}.item[_ngcontent-%COMP%]:active{background-color:var(--avl-solution-overview-item-pressed-color)}.icon-container[_ngcontent-%COMP%]{background:var(--avl-solution-overview-item-icon-circle-color);display:flex;align-items:center;justify-content:center;width:48px;height:48px;min-height:48px;border-radius:50%}.icon-container.no-icon-circle[_ngcontent-%COMP%]{background:0 0}.no-icon-circle[_ngcontent-%COMP%]   .item-icon[_ngcontent-%COMP%]{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color-no-circle)}.item-icon[_ngcontent-%COMP%]{width:24px;height:24px;--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color)}.item[_ngcontent-%COMP%]:active   .item-icon[_ngcontent-%COMP%]{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-pressed-color)}.icon-label[_ngcontent-%COMP%]{display:flex;justify-content:center;flex-flow:column;width:100%;font-size:14px;line-height:16px;margin:8px 0;font-weight:500;text-align:center;height:32px;color:var(--avl-solution-overview-text-color)}.group-name[_ngcontent-%COMP%]{margin:0 0 -8px;font-weight:400;font-size:24px;color:var(--avl-solution-overview-text-color);grid-column:1/end}.solution-name[_ngcontent-%COMP%]{font-weight:400;font-size:24px;margin:0 0 0 32px;height:32px;line-height:32px;flex:auto;color:var(--avl-solution-overview-title-color)}.solution-subtitle[_ngcontent-%COMP%]{font-weight:400;font-size:20px;margin:0 0 0 64px;color:var(--avl-solution-overview-subtitle-color)}.top-bar[_ngcontent-%COMP%]{display:flex;flex-direction:column;padding:32px 32px 0}.top-bar[_ngcontent-%COMP%]   .row[_ngcontent-%COMP%]{display:flex;flex-direction:row;justify-content:space-between;align-items:flex-start;height:32px}.top-bar[_ngcontent-%COMP%]   .row[_ngcontent-%COMP%]:last-child{margin-top:12px;height:24px}.action-buttons[_ngcontent-%COMP%]{margin-top:-8px;justify-self:flex-end;height:32px}.solution-logo[_ngcontent-%COMP%]{top:32px;left:32px;width:32px;height:32px}avl-search-bar[_ngcontent-%COMP%]{padding-top:20px}"] });
AvlSolutionOverviewComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", Array)
], AvlSolutionOverviewComponent.prototype, "groups", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlSolutionOverviewComponent.prototype, "solutionName", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlSolutionOverviewComponent.prototype, "solutionIconUrl", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], AvlSolutionOverviewComponent.prototype, "solutionSubtitle", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], AvlSolutionOverviewComponent.prototype, "showSearchBar", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlSolutionOverviewComponent.prototype, "searchQueryChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AvlSolutionOverviewComponent.prototype, "search", void 0);
__decorate([
    ViewChildren('itemElement'),
    __metadata("design:type", QueryList)
], AvlSolutionOverviewComponent.prototype, "buttonRefs", void 0);
__decorate([
    ViewChildren('group'),
    __metadata("design:type", QueryList)
], AvlSolutionOverviewComponent.prototype, "groupElementRefs", void 0);
__decorate([
    ViewChild('searchbar'),
    __metadata("design:type", ElementRef)
], AvlSolutionOverviewComponent.prototype, "searchbarRef", void 0);
__decorate([
    ViewChild('iconButtons'),
    __metadata("design:type", ElementRef)
], AvlSolutionOverviewComponent.prototype, "iconButtonsContainer", void 0);
__decorate([
    ViewChild('searchbarContainer'),
    __metadata("design:type", ElementRef)
], AvlSolutionOverviewComponent.prototype, "searchbarContainer", void 0);
__decorate([
    ViewChild('groupsContainer'),
    __metadata("design:type", ElementRef)
], AvlSolutionOverviewComponent.prototype, "groupsContainer", void 0);
__decorate([
    HostListener('window:resize'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], AvlSolutionOverviewComponent.prototype, "onResize", null);
__decorate([
    HostListener('keydown', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [KeyboardEvent]),
    __metadata("design:returntype", void 0)
], AvlSolutionOverviewComponent.prototype, "onKeydownHandler", null);
AvlSolutionOverviewComponent = __decorate([ __metadata("design:paramtypes", [ChangeDetectorRef])
], AvlSolutionOverviewComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSolutionOverviewComponent, [{
        type: Component,
        args: [{
                selector: 'avl-solution-overview',
                template: "<div class=\"container\">\r\n  <div class=\"top-bar\">\r\n    <div class=\"row\">\r\n      <img class=\"solution-logo\" src=\"{{solutionIconUrl}}\" />\r\n      <h1 class=\"solution-name\">{{solutionName}}</h1>\r\n      <div class=\"action-buttons\" #iconButtons>\r\n        <ng-content></ng-content>\r\n      </div>\r\n    </div>\r\n    <div class=\"row\" *ngIf=\"solutionSubtitle\">\r\n      <h2 class=\"solution-subtitle\">{{solutionSubtitle}}</h2>\r\n    </div>\r\n  </div>\r\n  <div class=\"outer-container\" [ngClass]=\"{'top-margin': contentHasScroll === false}\">\r\n    <div class=\"searchbar-container\" [ngClass]=\"gridWidthClass$ | async\" #searchbarContainer>\r\n      <avl-search-bar class=\"search-bar\" *ngIf=\"showSearchBar\"\r\n        #searchbar\r\n        always-open\r\n        (query-changed)=\"onQueryChange($event)\"\r\n        (search)=\"onSearch($event)\">\r\n      </avl-search-bar>\r\n    </div>\r\n    <div class=\"groups\" #groupsContainer>\r\n      <ng-container *ngFor=\"let group of groups\">\r\n        <div class=\"group-container\">\r\n          <div class=\"item-container\" #group [ngClass]=\"getMaxColumnsClass()\">\r\n            <ng-container *ngTemplateOutlet=\"groupTpl; context: { group: group }\"></ng-container>\r\n          </div>\r\n        </div>\r\n      </ng-container>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #groupTpl let-group=\"group\">\r\n  <h2 class=\"group-name\" *ngIf=\"group.title\">{{group.title}}</h2>\r\n  <div class=\"text-container\" *ngIf=\"group.header\">{{group.header}}</div>\r\n  <ng-container *ngFor=\"let item of group.items\">\r\n    <ng-container *ngIf=\"item.route\">\r\n      <a class=\"item\"\r\n         #itemElement\r\n         [routerLink]=\"item.route\"\r\n         [ngClass]=\"{'with-description': group.description}\"\r\n         tabindex=\"0\">\r\n        <ng-container *ngTemplateOutlet=\"buttonContent; context: { item: item }\">\r\n        </ng-container>\r\n      </a>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"item.link\">\r\n      <a class=\"item\"\r\n         #itemElement\r\n         [attr.href]=\"item.link\"\r\n         [attr.target]=\"item.target\"\r\n         [attr.rel]=\"item.target == '_blank' ? 'noopener' : null\"\r\n         [ngClass]=\"{'with-description': group.description}\"\r\n         tabindex=\"0\">\r\n        <ng-container *ngTemplateOutlet=\"buttonContent; context: { item: item }\">\r\n        </ng-container>\r\n      </a>\r\n    </ng-container>\r\n  </ng-container>\r\n  <div class=\"text-container\" *ngIf=\"group.footer\">{{group.footer}}</div>\r\n</ng-template>\r\n\r\n<ng-template #buttonContent let-item=\"item\">\r\n  <div class=\"icon-container\" [ngClass]=\"{'no-icon-circle': item.noIconCircle}\">\r\n    <avl-icon\r\n      *ngIf=\"item.icon\"\r\n      class=\"item-icon\"\r\n      [icon]=\"item.icon\">\r\n    </avl-icon>\r\n    <img *ngIf=\"item.iconUrl\" class=\"item-icon\" src=\"{{item.iconUrl}}\" />\r\n  </div>\r\n  <div class=\"icon-label\"><span class=\"label\">{{item.label}}</span></div>\r\n  <div class=\"description\">\r\n    <span class=\"text\">\r\n      {{item.description}}\r\n    </span>\r\n  </div>\r\n</ng-template>\r\n",
                styles: [":host{--avl-solution-overview-background-color:var(--avl-background-1-color);--avl-solution-overview-background-image:auto;--avl-solution-overview-text-color:var(--avl-foreground-color);--avl-solution-overview-title-color:var(--avl-foreground-color);--avl-solution-overview-subtitle-color:var(--avl-foreground-medium-color);--avl-solution-overview-item-background-color:var(\r\n    --avl-container-background-color\r\n  );--avl-solution-overview-item-hover-color:var(\r\n    --avl-container-background-product-hover-color,\r\n  );--avl-solution-overview-item-focus-color:var(\r\n    --avl-container-background-product-focus-color\r\n  );--avl-solution-overview-item-pressed-color:var(\r\n    --avl-container-background-product-pressed-color\r\n  );--avl-solution-overview-item-icon-circle-color:var(--avl-product-color);--avl-solution-overview-item-icon-color:var(--avl-on-primary-color);--avl-solution-overview-item-icon-color-no-circle:var(--avl-hint-color);--avl-solution-overview-item-description-color:var(\r\n    --avl-foreground-medium-color\r\n  );--avl-solution-overview-item-icon-pressed-color:var(\r\n    --avl-solution-overview-item-icon-color\r\n  );display:block;width:100%;--item-width:184px;--item-margin:24px;--total-width:calc(var(--item-width) + var(--item-margin));--total-height:100vh;height:var(--total-height);font-family:var(--avl-font-family)}.container{background-image:var(--avl-solution-overview-background-image);background-size:cover;background-color:var(--avl-solution-overview-background-color);height:100%;display:flex;flex-direction:column}.groups{margin-bottom:24px;width:99%;overflow-y:auto}.group-container{display:flex;width:100%;justify-content:center;position:relative}.outer-container{display:flex;align-items:center;flex-direction:column;justify-content:center;overflow:hidden;height:100%}.outer-container.top-margin{margin-top:-50px}.searchbar-container{display:flex;justify-content:space-between}.search-bar{padding-right:20px}.external-link-container,.item-container{display:grid;width:100%;margin:0 12px 2px;padding-top:24px;grid-template-columns:repeat(auto-fill,var(--item-width)) [end];grid-gap:var(--item-margin);justify-content:center}.text-container{color:var(--avl-solution-overview-text-color);grid-column:1/end}.max-columns-1{max-width:calc(var(--total-width) * 1)}.max-columns-2{max-width:calc(var(--total-width) * 2)}.max-columns-3{max-width:calc(var(--total-width) * 3)}.max-columns-4{max-width:calc(var(--total-width) * 4)}.max-columns-5{max-width:calc(var(--total-width) * 5)}.grid-width-1{width:calc(var(--total-width) * 1)}.grid-width-2{width:calc(var(--total-width) * 2)}.grid-width-3{width:calc(var(--total-width) * 3)}.grid-width-4{width:calc(var(--total-width) * 4)}.grid-width-5{width:calc(var(--total-width) * 5)}.item{position:relative;display:flex;flex-direction:column;align-items:center;padding:12px;width:var(--item-width);height:112px;box-sizing:border-box;background-color:var(--avl-solution-overview-item-background-color);outline:0;box-shadow:var(--avl-box-shadow-1dp);transition:box-shadow .15s cubic-bezier(.4,0,.2,1)}.item.with-description{height:204px}.description{display:none}.item.with-description .description{overflow:hidden;display:block;font-weight:400;text-align:center;font-size:14px;line-height:16px;color:var(--avl-foreground-medium-color)}.item:active,.item:focus,.item:hover{box-shadow:var(--avl-box-shadow-8dp)}.item:focus{background-color:var(--avl-solution-overview-item-focus-color)}.item:hover{background-color:var(--avl-solution-overview-item-hover-color)}.item:active{background-color:var(--avl-solution-overview-item-pressed-color)}.icon-container{background:var(--avl-solution-overview-item-icon-circle-color);display:flex;align-items:center;justify-content:center;width:48px;height:48px;min-height:48px;border-radius:50%}.icon-container.no-icon-circle{background:0 0}.no-icon-circle .item-icon{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color-no-circle)}.item-icon{width:24px;height:24px;--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color)}.item:active .item-icon{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-pressed-color)}.icon-label{display:flex;justify-content:center;flex-flow:column;width:100%;font-size:14px;line-height:16px;margin:8px 0;font-weight:500;text-align:center;height:32px;color:var(--avl-solution-overview-text-color)}.group-name{margin:0 0 -8px;font-weight:400;font-size:24px;color:var(--avl-solution-overview-text-color);grid-column:1/end}.solution-name{font-weight:400;font-size:24px;margin:0 0 0 32px;height:32px;line-height:32px;flex:auto;color:var(--avl-solution-overview-title-color)}.solution-subtitle{font-weight:400;font-size:20px;margin:0 0 0 64px;color:var(--avl-solution-overview-subtitle-color)}.top-bar{display:flex;flex-direction:column;padding:32px 32px 0}.top-bar .row{display:flex;flex-direction:row;justify-content:space-between;align-items:flex-start;height:32px}.top-bar .row:last-child{margin-top:12px;height:24px}.action-buttons{margin-top:-8px;justify-self:flex-end;height:32px}.solution-logo{top:32px;left:32px;width:32px;height:32px}avl-search-bar{padding-top:20px}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { searchQueryChange: [{
            type: Output
        }], search: [{
            type: Output
        }], onResize: [{
            type: HostListener,
            args: ['window:resize']
        }], onKeydownHandler: [{
            type: HostListener,
            args: ['keydown', ['$event']]
        }], groups: [{
            type: Input
        }], solutionName: [{
            type: Input
        }], solutionIconUrl: [{
            type: Input
        }], solutionSubtitle: [{
            type: Input
        }], showSearchBar: [{
            type: Input
        }], buttonRefs: [{
            type: ViewChildren,
            args: ['itemElement']
        }], groupElementRefs: [{
            type: ViewChildren,
            args: ['group']
        }], searchbarRef: [{
            type: ViewChild,
            args: ['searchbar']
        }], iconButtonsContainer: [{
            type: ViewChild,
            args: ['iconButtons']
        }], searchbarContainer: [{
            type: ViewChild,
            args: ['searchbarContainer']
        }], groupsContainer: [{
            type: ViewChild,
            args: ['groupsContainer']
        }] }); })();
export { AvlSolutionOverviewComponent };
