// @avl-controls/ng-binders/esm2015/lib/controls/avl-snackbar.directive.js
import { Directive, EventEmitter, Output, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
export class AvlSnackbarDirective {
    constructor() {
        this.buttonClicked = new EventEmitter();
        this.displayedChange = new EventEmitter();
    }
    buttonClickedHandler(e) {
        this.buttonClicked.emit(e.detail.value);
    }
    displayedChangeHandler(e) {
        this.displayedChange.emit(e.detail.value);
    }
}
AvlSnackbarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlSnackbarDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
AvlSnackbarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.4", type: AvlSnackbarDirective, selector: "avl-snackbar", outputs: { buttonClicked: "buttonClicked", displayedChange: "displayedChange" }, host: { listeners: { "button-clicked": "buttonClickedHandler($event)", "displayed-changed": "displayedChangeHandler($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlSnackbarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'avl-snackbar'
                }]
        }], propDecorators: { buttonClicked: [{
                type: Output
            }], displayedChange: [{
                type: Output
            }], buttonClickedHandler: [{
                type: HostListener,
                args: ['button-clicked', ['$event']]
            }], displayedChangeHandler: [{
                type: HostListener,
                args: ['displayed-changed', ['$event']]
            }] } });
