// @avl-controls/ng-binders/esm2015/lib/controls/avl-snackbar.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, EventEmitter, Output, HostListener } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let AvlSnackbarDirective = class AvlSnackbarDirective {
    constructor() {
        this.buttonClicked = new EventEmitter();
        this.displayedChange = new EventEmitter();
    }
    buttonClickedHandler(e) {
        this.buttonClicked.emit(e.detail.value);
    }
    displayedChangeHandler(e) {
        this.displayedChange.emit(e.detail.value);
    }
};
AvlSnackbarDirective.ɵfac = function AvlSnackbarDirective_Factory(t) { return new (t || AvlSnackbarDirective)(); };
AvlSnackbarDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlSnackbarDirective, selectors: [["avl-snackbar"]], hostBindings: function AvlSnackbarDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("button-clicked", function AvlSnackbarDirective_button_clicked_HostBindingHandler($event) { return ctx.buttonClickedHandler($event); })("displayed-changed", function AvlSnackbarDirective_displayed_changed_HostBindingHandler($event) { return ctx.displayedChangeHandler($event); });
    } }, outputs: { buttonClicked: "buttonClicked", displayedChange: "displayedChange" } });
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlSnackbarDirective.prototype, "buttonClicked", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlSnackbarDirective.prototype, "displayedChange", void 0);
__decorate([
    HostListener('button-clicked', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [CustomEvent]),
    __metadata("design:returntype", void 0)
], AvlSnackbarDirective.prototype, "buttonClickedHandler", null);
__decorate([
    HostListener('displayed-changed', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [CustomEvent]),
    __metadata("design:returntype", void 0)
], AvlSnackbarDirective.prototype, "displayedChangeHandler", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSnackbarDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-snackbar'
            }]
    }], function () { return []; }, { buttonClicked: [{
            type: Output
        }], displayedChange: [{
            type: Output
        }], buttonClickedHandler: [{
            type: HostListener,
            args: ['button-clicked', ['$event']]
        }], displayedChangeHandler: [{
            type: HostListener,
            args: ['displayed-changed', ['$event']]
        }] }); })();
export { AvlSnackbarDirective };
