// @avl-controls/ng-binders/esm2015/lib/controls/avl-upload.directive.js
import { Directive, HostListener, Output, EventEmitter, Inject } from '@angular/core';
import { UploadService, UPLOAD_SERVICE } from '@avl-services/ng-services';
import * as i0 from "@angular/core";
export class AvlUploadDirective {
    constructor(element, uploadService) {
        this.element = element;
        this.uploadService = uploadService;
        this.uploadComplete = new EventEmitter();
        this.uploadAbort = new EventEmitter();
        this.uploadFail = new EventEmitter();
        this.uploadStart = new EventEmitter();
        this.filesChanged = new EventEmitter();
        this.fileRejected = new EventEmitter();
        this.menuItemClicked = new EventEmitter();
        this.downloadFail = new EventEmitter();
        this.downloadAbort = new EventEmitter();
    }
    ngOnInit() {
        const element = this.element.nativeElement;
        element.uploadService = this.uploadService;
    }
    uploadCompleteHandler(e) {
        this.uploadComplete.emit(e.detail.value);
    }
    uploadCanceledHandler(e) {
        this.uploadAbort.emit(e.detail.value);
    }
    uploadStartedHandler(e) {
        this.uploadStart.emit(e.detail.value);
    }
    uploadFailedHandler(e) {
        this.uploadFail.emit(e.detail);
    }
    filesChangedHandler(e) {
        this.filesChanged.emit(e.detail.value);
    }
    fileRejectedHandler(e) {
        this.fileRejected.emit(e.detail.value);
    }
    menuItemClickedHandler(e) {
        this.menuItemClicked.emit(e.detail);
    }
    downloadFailedHandler(e) {
        this.downloadFail.emit(e.detail);
    }
    downloadCanceledHandler(e) {
        this.downloadAbort.emit(e.detail.value);
    }
}
AvlUploadDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlUploadDirective, deps: [{ token: i0.ElementRef }, { token: UPLOAD_SERVICE }], target: i0.ɵɵFactoryTarget.Directive });
AvlUploadDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.4", type: AvlUploadDirective, selector: "avl-upload", outputs: { uploadComplete: "uploadComplete", uploadAbort: "uploadAbort", uploadFail: "uploadFail", uploadStart: "uploadStart", filesChanged: "filesChanged", fileRejected: "fileRejected", menuItemClicked: "menuItemClicked", downloadFail: "downloadFail", downloadAbort: "downloadAbort" }, host: { listeners: { "upload-complete": "uploadCompleteHandler($event)", "upload-abort": "uploadCanceledHandler($event)", "upload-start": "uploadStartedHandler($event)", "upload-fail": "uploadFailedHandler($event)", "files-changed": "filesChangedHandler($event)", "file-rejected": "fileRejectedHandler($event)", "menu-item-clicked": "menuItemClickedHandler($event)", "download-fail": "downloadFailedHandler($event)", "download-abort": "downloadCanceledHandler($event)" } }, providers: [{ provide: UPLOAD_SERVICE, useClass: UploadService }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlUploadDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'avl-upload',
                    providers: [{ provide: UPLOAD_SERVICE, useClass: UploadService }]
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [UPLOAD_SERVICE]
                }] }]; }, propDecorators: { uploadComplete: [{
                type: Output
            }], uploadAbort: [{
                type: Output
            }], uploadFail: [{
                type: Output
            }], uploadStart: [{
                type: Output
            }], filesChanged: [{
                type: Output
            }], fileRejected: [{
                type: Output
            }], menuItemClicked: [{
                type: Output
            }], downloadFail: [{
                type: Output
            }], downloadAbort: [{
                type: Output
            }], uploadCompleteHandler: [{
                type: HostListener,
                args: ['upload-complete', ['$event']]
            }], uploadCanceledHandler: [{
                type: HostListener,
                args: ['upload-abort', ['$event']]
            }], uploadStartedHandler: [{
                type: HostListener,
                args: ['upload-start', ['$event']]
            }], uploadFailedHandler: [{
                type: HostListener,
                args: ['upload-fail', ['$event']]
            }], filesChangedHandler: [{
                type: HostListener,
                args: ['files-changed', ['$event']]
            }], fileRejectedHandler: [{
                type: HostListener,
                args: ['file-rejected', ['$event']]
            }], menuItemClickedHandler: [{
                type: HostListener,
                args: ['menu-item-clicked', ['$event']]
            }], downloadFailedHandler: [{
                type: HostListener,
                args: ['download-fail', ['$event']]
            }], downloadCanceledHandler: [{
                type: HostListener,
                args: ['download-abort', ['$event']]
            }] } });
