// @avl-controls/ng-binders/esm2015/lib/controls/avl-tab-bar.directive.js
import { __awaiter, __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Output, EventEmitter } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let AvlTabBarDirective = class AvlTabBarDirective {
    constructor(el) {
        this.el = el;
        this.selectedIndexChange = new EventEmitter();
        this.selectedValueChange = new EventEmitter();
        this.tabBar = this.el.nativeElement;
        Promise.resolve().then(() => {
            this.setEventListeners(el.nativeElement);
        });
    }
    setEventListeners(el) {
        return __awaiter(this, void 0, void 0, function* () {
            el.addEventListener('selected-value-changed', (event) => {
                Promise.resolve().then(() => {
                    this.selectedValueChange.emit(event.detail.value);
                });
            });
            el.addEventListener('selected-index-changed', (event) => {
                Promise.resolve().then(() => {
                    this.selectedIndexChange.emit(event.detail.value);
                });
            });
        });
    }
    setTab(index, value) {
        const tabs = this.tabBar.tabs;
        for (const prop in value) {
            if (value.hasOwnProperty(prop)) {
                tabs[index][prop] = value[prop];
                this.tabBar.notifyPath(`tabs.${index}.${prop}`);
            }
        }
    }
};
AvlTabBarDirective.ɵfac = function AvlTabBarDirective_Factory(t) { return new (t || AvlTabBarDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlTabBarDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlTabBarDirective, selectors: [["avl-tab-bar"]], outputs: { selectedIndexChange: "selectedIndexChange", selectedValueChange: "selectedValueChange" } });
AvlTabBarDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlTabBarDirective.prototype, "selectedIndexChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlTabBarDirective.prototype, "selectedValueChange", void 0);
AvlTabBarDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlTabBarDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlTabBarDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-tab-bar'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { selectedIndexChange: [{
            type: Output
        }], selectedValueChange: [{
            type: Output
        }] }); })();
export { AvlTabBarDirective };
