// @avl-controls/ng-binders/esm2015/lib/controls/avl-timespan.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
let AvlTimespanDirective = class AvlTimespanDirective {
    constructor(elementRef) {
        this.valueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'value-changed', this.valueChange);
    }
};
AvlTimespanDirective.ɵfac = function AvlTimespanDirective_Factory(t) { return new (t || AvlTimespanDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlTimespanDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlTimespanDirective, selectors: [["avl-timespan"]], outputs: { valueChange: "valueChange" } });
AvlTimespanDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlTimespanDirective.prototype, "valueChange", void 0);
AvlTimespanDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlTimespanDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlTimespanDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-timespan'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { valueChange: [{
            type: Output
        }] }); })();
export { AvlTimespanDirective };
