// @avl-controls/ng-binders/esm2015/lib/controls/avl-textarea.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
let AvlTextareaDirective = class AvlTextareaDirective {
    constructor(elementRef) {
        this.valueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'value-changed', this.valueChange);
    }
};
AvlTextareaDirective.ɵfac = function AvlTextareaDirective_Factory(t) { return new (t || AvlTextareaDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlTextareaDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlTextareaDirective, selectors: [["avl-textarea"]], outputs: { valueChange: "valueChange" } });
AvlTextareaDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlTextareaDirective.prototype, "valueChange", void 0);
AvlTextareaDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlTextareaDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlTextareaDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-textarea'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { valueChange: [{
            type: Output
        }] }); })();
export { AvlTextareaDirective };
