// @avl-controls/ng-binders/esm2015/lib/controls/avl-datatable.directive.js
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as i0 from "@angular/core";
export class AvlDatatableDirective {
    constructor(elementRef) {
        this.buttonClick = new EventEmitter();
        this.overflowMenuItemClick = new EventEmitter();
        this.linkClick = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.selectedPageChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'page-size-changed', this.pageSizeChange);
        addEventListenerForEmittingChangedEvents(elementRef, 'selected-page-changed', this.selectedPageChange);
    }
    buttonClickListener(e) {
        this.buttonClick.emit(e.detail);
    }
    overflowMenuItemClickListener(e) {
        this.overflowMenuItemClick.emit(e.detail);
    }
    linkClickListener(e) {
        this.linkClick.emit(e.detail);
    }
}
AvlDatatableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlDatatableDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
AvlDatatableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.4", type: AvlDatatableDirective, selector: "avl-datatable", outputs: { buttonClick: "buttonClick", overflowMenuItemClick: "overflowMenuItemClick", linkClick: "linkClick", pageSizeChange: "pageSizeChange", selectedPageChange: "selectedPageChange" }, host: { listeners: { "button-click": "buttonClickListener($event)", "overflow-menu-item-click": "overflowMenuItemClickListener($event)", "link-click": "linkClickListener($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlDatatableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'avl-datatable'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { buttonClick: [{
                type: Output
            }], overflowMenuItemClick: [{
                type: Output
            }], linkClick: [{
                type: Output
            }], pageSizeChange: [{
                type: Output
            }], selectedPageChange: [{
                type: Output
            }], buttonClickListener: [{
                type: HostListener,
                args: ['button-click', ['$event']]
            }], overflowMenuItemClickListener: [{
                type: HostListener,
                args: ['overflow-menu-item-click', ['$event']]
            }], linkClickListener: [{
                type: HostListener,
                args: ['link-click', ['$event']]
            }] } });
