// @avl-controls/ng-binders/esm2015/lib/controls/avl-context-menu.directive.js
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class AvlContextMenuDirective {
    constructor() {
        this.itemClicked = new EventEmitter();
    }
    itemClickedListener(e) {
        this.itemClicked.emit(e.detail.value);
    }
}
AvlContextMenuDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlContextMenuDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
AvlContextMenuDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.4", type: AvlContextMenuDirective, selector: "avl-context-menu", outputs: { itemClicked: "itemClicked" }, host: { listeners: { "item-clicked": "itemClickedListener($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlContextMenuDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'avl-context-menu'
                }]
        }], propDecorators: { itemClicked: [{
                type: Output
            }], itemClickedListener: [{
                type: HostListener,
                args: ['item-clicked', ['$event']]
            }] } });
