// @avl-controls/ng-binders/esm2015/lib/controls/avl-breadcrumb.directive.js
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class AvlBreadcrumbDirective {
    constructor() {
        this.itemClicked = new EventEmitter();
    }
    itemClickedListener(e) {
        this.itemClicked.emit(e.detail.item);
    }
}
AvlBreadcrumbDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlBreadcrumbDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
AvlBreadcrumbDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.4", type: AvlBreadcrumbDirective, selector: "avl-breadcrumb", outputs: { itemClicked: "itemClicked" }, host: { listeners: { "item-clicked": "itemClickedListener($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.4", ngImport: i0, type: AvlBreadcrumbDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'avl-breadcrumb'
                }]
        }], propDecorators: { itemClicked: [{
                type: Output
            }], itemClickedListener: [{
                type: HostListener,
                args: ['item-clicked', ['$event']]
            }] } });
