// @avl-controls/ng-binders/esm2015/lib/controls/avl-datatable.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, EventEmitter, HostListener, Output, ElementRef } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
let AvlDatatableDirective = class AvlDatatableDirective {
    constructor(elementRef) {
        this.buttonClick = new EventEmitter();
        this.overflowMenuItemClick = new EventEmitter();
        this.linkClick = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.selectedPageChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'page-size-changed', this.pageSizeChange);
        addEventListenerForEmittingChangedEvents(elementRef, 'selected-page-changed', this.selectedPageChange);
    }
    buttonClickListener(e) {
        this.buttonClick.emit(e.detail);
    }
    overflowMenuItemClickListener(e) {
        this.overflowMenuItemClick.emit(e.detail);
    }
    linkClickListener(e) {
        this.linkClick.emit(e.detail);
    }
};
AvlDatatableDirective.ɵfac = function AvlDatatableDirective_Factory(t) { return new (t || AvlDatatableDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlDatatableDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlDatatableDirective, selectors: [["avl-datatable"]], hostBindings: function AvlDatatableDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("button-click", function AvlDatatableDirective_button_click_HostBindingHandler($event) { return ctx.buttonClickListener($event); })("overflow-menu-item-click", function AvlDatatableDirective_overflow_menu_item_click_HostBindingHandler($event) { return ctx.overflowMenuItemClickListener($event); })("link-click", function AvlDatatableDirective_link_click_HostBindingHandler($event) { return ctx.linkClickListener($event); });
    } }, outputs: { buttonClick: "buttonClick", overflowMenuItemClick: "overflowMenuItemClick", linkClick: "linkClick", pageSizeChange: "pageSizeChange", selectedPageChange: "selectedPageChange" } });
AvlDatatableDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlDatatableDirective.prototype, "buttonClick", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlDatatableDirective.prototype, "overflowMenuItemClick", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlDatatableDirective.prototype, "linkClick", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlDatatableDirective.prototype, "pageSizeChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], AvlDatatableDirective.prototype, "selectedPageChange", void 0);
__decorate([
    HostListener('button-click', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [CustomEvent]),
    __metadata("design:returntype", void 0)
], AvlDatatableDirective.prototype, "buttonClickListener", null);
__decorate([
    HostListener('overflow-menu-item-click', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [CustomEvent]),
    __metadata("design:returntype", void 0)
], AvlDatatableDirective.prototype, "overflowMenuItemClickListener", null);
__decorate([
    HostListener('link-click', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [CustomEvent]),
    __metadata("design:returntype", void 0)
], AvlDatatableDirective.prototype, "linkClickListener", null);
AvlDatatableDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlDatatableDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlDatatableDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-datatable'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { buttonClick: [{
            type: Output
        }], overflowMenuItemClick: [{
            type: Output
        }], linkClick: [{
            type: Output
        }], pageSizeChange: [{
            type: Output
        }], selectedPageChange: [{
            type: Output
        }], buttonClickListener: [{
            type: HostListener,
            args: ['button-click', ['$event']]
        }], overflowMenuItemClickListener: [{
            type: HostListener,
            args: ['overflow-menu-item-click', ['$event']]
        }], linkClickListener: [{
            type: HostListener,
            args: ['link-click', ['$event']]
        }] }); })();
export { AvlDatatableDirective };
