// @avl-ng-controls/auth/esm2015/lib/services/token-storage.js
import { StorageNames } from './storage-names';
export class TokenStorage {
    constructor(prefix = '') {
        this.prefix = '';
        this.prefix = prefix;
    }
    readUser() {
        const roles = JSON.parse(this.getItem(StorageNames.Roles));
        return {
            username: this.getItem(StorageNames.UserName),
            name: this.getItem(StorageNames.Name),
            access_token: this.getItem(StorageNames.AccessToken),
            access_token_expires: this.getNumber(StorageNames.AccessTokenExpiresAt),
            access_token_stored: this.getNumber(StorageNames.AccessTokenStoredAt),
            refresh_token: this.getItem(StorageNames.RefreshToken),
            refresh_token_expires: this.getNumber(StorageNames.RefreshTokenExpiresAt),
            id_token: this.getItem(StorageNames.IdToken),
            roles
        };
    }
    storeUser(user) {
        this.setItem(StorageNames.UserName, user.username);
        this.setItem(StorageNames.Name, user.name);
        this.setItem(StorageNames.AccessToken, user.access_token);
        this.setItem(StorageNames.AccessTokenExpiresAt, `${user.access_token_expires}`);
        this.setItem(StorageNames.AccessTokenStoredAt, `${user.access_token_stored}`);
        this.setItem(StorageNames.RefreshToken, user.refresh_token);
        this.setItem(StorageNames.RefreshTokenExpiresAt, `${user.refresh_token_expires}`);
        this.setItem(StorageNames.IdToken, user.id_token);
        const rolesStr = JSON.stringify(user.roles);
        sessionStorage.setItem(StorageNames.Roles, `${rolesStr}`);
    }
    clearUser() {
        for (const name in StorageNames) {
            if (StorageNames.hasOwnProperty(name)) {
                this.removeItem(StorageNames[name]);
            }
        }
    }
    getNumber(key) {
        return parseInt(this.getItem(key), 10);
    }
    getItem(key) {
        return sessionStorage.getItem(this.prefix + key);
    }
    setItem(key, value) {
        sessionStorage.setItem(this.prefix + key, value);
    }
    removeItem(key) {
        sessionStorage.removeItem(this.prefix + key);
    }
}
