// @avl-controls/ng-binders/esm5/lib/interface/avl-checkable-control.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
/** Directive for controls implementing [AvlCheckableControl]. */
import * as ɵngcc0 from '@angular/core';
var AvlCheckableControlDirective = /** @class */ (function () {
    function AvlCheckableControlDirective(elementRef) {
        this.checkedChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'checked-changed', this.checkedChange);
    }
    AvlCheckableControlDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlCheckableControlDirective.prototype, "checkedChange", void 0);
    AvlCheckableControlDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlCheckableControlDirective);
AvlCheckableControlDirective.ɵfac = function AvlCheckableControlDirective_Factory(t) { return new (t || AvlCheckableControlDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlCheckableControlDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlCheckableControlDirective, selectors: [["avl-checkbox"], ["avl-radio-button"], ["avl-icon-toggle"], ["avl-switch"]], outputs: { checkedChange: "checkedChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlCheckableControlDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-checkbox,avl-radio-button, avl-icon-toggle, avl-switch'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { checkedChange: [{
            type: Output
        }] }); })();
    return AvlCheckableControlDirective;
}());
export { AvlCheckableControlDirective };
