// @avl-controls/ng-binders/esm5/lib/interface/avl-multi-select-control.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
/** Directive for controls implementing [AvlMultiSelectControl]. */
import * as ɵngcc0 from '@angular/core';
var AvlMultiSelectControlDirective = /** @class */ (function () {
    function AvlMultiSelectControlDirective(elementRef) {
        this.selectedIndexesChange = new EventEmitter();
        this.selectedValuesChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'selected-indexes-changed', this.selectedIndexesChange);
        addEventListenerForEmittingChangedEvents(elementRef, 'selected-values-changed', this.selectedValuesChange);
    }
    AvlMultiSelectControlDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlMultiSelectControlDirective.prototype, "selectedIndexesChange", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlMultiSelectControlDirective.prototype, "selectedValuesChange", void 0);
    AvlMultiSelectControlDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlMultiSelectControlDirective);
AvlMultiSelectControlDirective.ɵfac = function AvlMultiSelectControlDirective_Factory(t) { return new (t || AvlMultiSelectControlDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlMultiSelectControlDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlMultiSelectControlDirective, selectors: [["avl-datatable"], ["avl-list"], ["avl-chip-list"], ["avl-chip-input"]], outputs: { selectedIndexesChange: "selectedIndexesChange", selectedValuesChange: "selectedValuesChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlMultiSelectControlDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-datatable,avl-list,avl-chip-list, avl-chip-input'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { selectedIndexesChange: [{
            type: Output
        }], selectedValuesChange: [{
            type: Output
        }] }); })();
    return AvlMultiSelectControlDirective;
}());
export { AvlMultiSelectControlDirective };
