// @avl-ng-controls/notification/esm5/lib/notification.module.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
import { CommonModule } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, NgModule } from '@angular/core';
import { TranslateModule } from "@ngx-translate/core";
import { AvlNgBindersModule } from '@avl-controls/ng-binders';
import { FormsModule } from '@angular/forms';
import { AvlNotificationMenuComponent } from './components/avl-notification-menu/avl-notification-menu.component';
import { AnimateButton } from './components/avl-notification-container/directives/animate-button.directive';
import { AnimateSnackbar } from './components/avl-notification-container/directives/animate-snackbar.directive';
import { AnimateStack } from './components/avl-notification-container/directives/animate-stack.directive';
import { ShowHideSnackbarDirective } from './components/avl-notification-container/directives/show-hide-snackbar.directive';
import { SnackbarListener } from './components/avl-notification-container/directives/snackbar-listener.directive';
import { AvlNotificationContainerComponent } from './components/avl-notification-container/avl-notification-container.component';
import { AvlNotificationDialogComponent } from './components/avl-notification-dialog/avl-notification-dialog.component';
import { AvlDialogModule } from '@avl-ng-controls/dialog';
import { AvlNavigationModule } from '@avl-ng-controls/navigation';
var componentsAndDirectives = [
    AvlNotificationMenuComponent,
    AvlNotificationContainerComponent,
    SnackbarListener,
    AnimateStack,
    AnimateSnackbar,
    ShowHideSnackbarDirective,
    AnimateButton,
    AvlNotificationDialogComponent
];
var AvlNotificationModule = /** @class */ (function () {
    function AvlNotificationModule() {
    }
AvlNotificationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AvlNotificationModule });
AvlNotificationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AvlNotificationModule_Factory(t) { return new (t || AvlNotificationModule)(); }, imports: [[CommonModule, TranslateModule, AvlNgBindersModule, AvlDialogModule, AvlNavigationModule, FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AvlNotificationModule, { declarations: function () { return [AvlNotificationMenuComponent,
        AvlNotificationContainerComponent,
        SnackbarListener,
        AnimateStack,
        AnimateSnackbar,
        ShowHideSnackbarDirective,
        AnimateButton,
        AvlNotificationDialogComponent]; }, imports: function () { return [CommonModule, TranslateModule, AvlNgBindersModule, AvlDialogModule, AvlNavigationModule, FormsModule]; }, exports: function () { return [AvlNotificationMenuComponent,
        AvlNotificationContainerComponent,
        SnackbarListener,
        AnimateStack,
        AnimateSnackbar,
        ShowHideSnackbarDirective,
        AnimateButton,
        AvlNotificationDialogComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNotificationModule, [{
        type: NgModule,
        args: [{
                declarations: __spread(componentsAndDirectives),
                imports: [CommonModule, TranslateModule, AvlNgBindersModule, AvlDialogModule, AvlNavigationModule, FormsModule],
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                exports: __spread(componentsAndDirectives)
            }]
    }], function () { return []; }, null); })();
    return AvlNotificationModule;
}());
export { AvlNotificationModule };
