// @avl-ng-controls/dialog/esm5/lib/dialog.module.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@avl-controls/ng-binders';
import * as ɵngcc3 from '@angular/cdk/a11y';
import * as ɵngcc4 from '@angular/forms';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { A11yModule } from '@angular/cdk/a11y';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
// components
import { AvlDialogComponent } from './avl-dialog/avl-dialog.component';
import { AvlNgBindersModule } from '@avl-controls/ng-binders';
import { AvlDialogHeaderComponent } from './avl-dialog/avl-dialog-header/avl-dialog-header.component';
import { AvlDialogFooterComponent } from './avl-dialog/avl-dialog-footer/avl-dialog-footer.component';
import { AvlDialogContentComponent } from './avl-dialog/avl-dialog-content/avl-dialog-content.component';
import { AvlAboutDialogComponent } from './avl-about-dialog/avl-about-dialog.component';
var componentsAndDirectives = [
    AvlDialogComponent,
    AvlDialogHeaderComponent,
    AvlDialogFooterComponent,
    AvlDialogContentComponent,
    AvlAboutDialogComponent,
];
var providers = [];
var AvlDialogModule = /** @class */ (function () {
    function AvlDialogModule() {
    }
AvlDialogModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AvlDialogModule });
AvlDialogModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AvlDialogModule_Factory(t) { return new (t || AvlDialogModule)(); }, providers: providers, imports: [[
            CommonModule,
            AvlNgBindersModule,
            A11yModule,
            FormsModule,
            ReactiveFormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AvlDialogModule, { declarations: function () { return [AvlDialogComponent,
        AvlDialogHeaderComponent,
        AvlDialogFooterComponent,
        AvlDialogContentComponent,
        AvlAboutDialogComponent]; }, imports: function () { return [CommonModule,
        AvlNgBindersModule,
        A11yModule,
        FormsModule,
        ReactiveFormsModule]; }, exports: function () { return [AvlDialogComponent,
        AvlDialogHeaderComponent,
        AvlDialogFooterComponent,
        AvlDialogContentComponent,
        AvlAboutDialogComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlDialogModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AvlNgBindersModule,
                    A11yModule,
                    FormsModule,
                    ReactiveFormsModule
                ],
                declarations: componentsAndDirectives,
                providers: providers,
                schemas: [
                    CUSTOM_ELEMENTS_SCHEMA
                ],
                exports: componentsAndDirectives
            }]
    }], function () { return []; }, null); })();
ɵngcc0.ɵɵsetComponentScope(AvlDialogComponent, [ɵngcc1.NgClass, ɵngcc1.NgComponentOutlet, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgStyle, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase, ɵngcc1.NgSwitchDefault, ɵngcc1.NgPlural, ɵngcc1.NgPluralCase, ɵngcc2.AvlBreadcrumbDirective, ɵngcc2.AvlOverflowMenuDirective, ɵngcc2.AvlSearchBarDirective, ɵngcc2.AvlTextareaDirective, ɵngcc2.AvlDateDirective, ɵngcc2.AvlSideSheetDirective, ɵngcc2.AvlTextfieldDirective, ɵngcc2.AvlTimespanDirective, ɵngcc2.AvlCheckableControlDirective, ɵngcc2.AvlControlDirective, ɵngcc2.AvlFormControlDirective, ɵngcc2.AvlInteractionControlDirective, ɵngcc2.AvlMultiSelectControlDirective, ɵngcc2.AvlSingleSelectControlDirective, ɵngcc2.AvlDropdownDirective, ɵngcc2.AvlTimeDirective, ɵngcc2.AvlListDirective, ɵngcc2.AvlUploadDirective, ɵngcc2.AvlContextMenuDirective, ɵngcc2.AvlNumberfieldDirective, ɵngcc2.AvlCheckboxDirective, ɵngcc2.AvlDatatableDirective, ɵngcc2.AvlComboBoxDirective, ɵngcc2.AvlTabBarDirective, ɵngcc2.AvlSnackbarDirective, ɵngcc2.ɵa, ɵngcc2.AvlIconButtonDirective, ɵngcc3.CdkAriaLive, ɵngcc3.CdkTrapFocus, ɵngcc3.CdkMonitorFocus, ɵngcc4.ɵangular_packages_forms_forms_y, ɵngcc4.NgSelectOption, ɵngcc4.ɵangular_packages_forms_forms_x, ɵngcc4.DefaultValueAccessor, ɵngcc4.NumberValueAccessor, ɵngcc4.RangeValueAccessor, ɵngcc4.CheckboxControlValueAccessor, ɵngcc4.SelectControlValueAccessor, ɵngcc4.SelectMultipleControlValueAccessor, ɵngcc4.RadioControlValueAccessor, ɵngcc4.NgControlStatus, ɵngcc4.NgControlStatusGroup, ɵngcc4.RequiredValidator, ɵngcc4.MinLengthValidator, ɵngcc4.MaxLengthValidator, ɵngcc4.PatternValidator, ɵngcc4.CheckboxRequiredValidator, ɵngcc4.EmailValidator, ɵngcc4.NgModel, ɵngcc4.NgModelGroup, ɵngcc4.NgForm, ɵngcc4.FormControlDirective, ɵngcc4.FormGroupDirective, ɵngcc4.FormControlName, ɵngcc4.FormGroupName, ɵngcc4.FormArrayName, AvlDialogComponent,
    AvlDialogHeaderComponent,
    AvlDialogFooterComponent,
    AvlDialogContentComponent,
    AvlAboutDialogComponent], [ɵngcc1.AsyncPipe, ɵngcc1.UpperCasePipe, ɵngcc1.LowerCasePipe, ɵngcc1.JsonPipe, ɵngcc1.SlicePipe, ɵngcc1.DecimalPipe, ɵngcc1.PercentPipe, ɵngcc1.TitleCasePipe, ɵngcc1.CurrencyPipe, ɵngcc1.DatePipe, ɵngcc1.I18nPluralPipe, ɵngcc1.I18nSelectPipe, ɵngcc1.KeyValuePipe]);
    return AvlDialogModule;
}());
export { AvlDialogModule };
