// @avl-ng-controls/navigation/esm5/lib/avl-navigation.module.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
// toolbar stuff
import { AvlAppBarComponent } from './components/avl-app-bar/avl-app-bar.component';
import { AvlAppBarItemComponent } from './components/avl-app-bar/avl-app-bar-item.component';
// sidenav stuff
import { AvlSideNavComponent } from './components/avl-side-nav/avl-side-nav.component';
import { AvlAppLayoutComponent } from './components/avl-app-layout/avl-app-layout.component';
import { AvlNoPaddingDirective } from './components/avl-app-layout/avl-no-padding.directive';
import { AvlLongPageComponent } from './components/avl-long-page/avl-long-page.component';
import { AvlParagraphDirective } from './components/avl-long-page/avl-paragraph.directive';
import { AvlSolutionOverviewComponent } from './components/avl-solution-overview/avl-solution-overview.component';
import { AvlNgBindersModule } from '@avl-controls/ng-binders';
import { AvlListExpanderComponent } from './components/avl-list-expander/avl-list-expander.component';
var componentsAndDirectives = [
    AvlAppBarComponent,
    AvlAppBarItemComponent,
    AvlSideNavComponent,
    AvlAppLayoutComponent,
    AvlNoPaddingDirective,
    AvlLongPageComponent,
    AvlParagraphDirective,
    AvlSolutionOverviewComponent,
    AvlListExpanderComponent,
];
var AvlNavigationModule = /** @class */ (function () {
    function AvlNavigationModule() {
    }
AvlNavigationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AvlNavigationModule });
AvlNavigationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AvlNavigationModule_Factory(t) { return new (t || AvlNavigationModule)(); }, imports: [[
            CommonModule,
            RouterModule,
            AvlNgBindersModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AvlNavigationModule, { declarations: function () { return [AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent]; }, imports: function () { return [CommonModule,
        RouterModule,
        AvlNgBindersModule]; }, exports: function () { return [AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNavigationModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    AvlNgBindersModule
                ],
                declarations: componentsAndDirectives,
                schemas: [
                    CUSTOM_ELEMENTS_SCHEMA
                ],
                exports: componentsAndDirectives
            }]
    }], function () { return []; }, null); })();
    return AvlNavigationModule;
}());
export { AvlNavigationModule };
