// @avl-ng-controls/notification/esm2015/lib/components/avl-notification-dialog/avl-notification-dialog.consts.js
import { ColumnType, LeafColumn } from '@avl-controls/datatable/dist/avl-datatable/column-definitions';
export const DEFAULT_PAGE_SIZE_OPTIONS = [
    { value: 15, label: 'Show 15 per page' },
    { value: 50, label: 'Show 50 per page' },
    { value: Number.MAX_SAFE_INTEGER, label: 'Show all' }
];
export const DEFAULT_FILTERS = {
    type: {
        warning: undefined,
        error: undefined,
        info: undefined,
        ok: undefined
    },
    state: {
        read: undefined,
        unread: undefined
    },
    date: {
        dateFrom: undefined,
        dateTo: undefined
    },
    search: undefined
};
export const DEFAULT_COLUMN_DEFINITIONS = [
    new LeafColumn('VizBox.NotificationDialog.Type', 'icon', ColumnType.Icon, false, false, 0.5),
    new LeafColumn('VizBox.NotificationDialog.Module', 'module', ColumnType.Text, true, false, 3),
    new LeafColumn('VizBox.NotificationDialog.Message', 'messageText', ColumnType.Text, true, false, 8),
    new LeafColumn('VizBox.NotificationDialog.Id', 'id', ColumnType.Number, false, false, 1),
    {
        caption: 'VizBox.NotificationDialog.Timestamp',
        valueField: 'timestamp',
        type: ColumnType.Date,
        sortable: false,
        filterable: false,
        flexWidth: 3,
        dateTimeOptions: { format: 'date-time-short' }
    },
    new LeafColumn('VizBox.NotificationDialog.State', 'currentIconToggle', ColumnType.Icon, false, false, 1)
];
export const NOTIFICATION_TYPE_MAPPINGS = new Map([
    [0 /* Error */, 'error'],
    [1 /* Warning */, 'warning'],
    [2 /* Info */, 'info'],
    [3 /* OK */, 'ok'],
    [false, 'unread'],
    [true, 'read']
]);
export const TYPE_ICON_MAPPINGS = {
    [0 /* Error */]: 'cancel',
    [1 /* Warning */]: 'warning',
    [2 /* Info */]: 'info-outline',
    [3 /* OK */]: 'check'
};
export const STATE_ICON_MAPPINGS = {
    false: 'mail',
    true: 'communication:mail-outline'
};
export const DEFAULT_ICON_STYLE = 'height:24px; width:24px;';
export const TYPE_ICON_STYLE_MAPPINGS = {
    [0 /* Error */]: `--avl-icon-custom-color: var(--avl-error-color); ${DEFAULT_ICON_STYLE}`,
    [1 /* Warning */]: `--avl-icon-custom-color: var(--avl-warning-color); ${DEFAULT_ICON_STYLE}`,
    [2 /* Info */]: `--avl-icon-custom-color: var(--avl-icon-default-color); ${DEFAULT_ICON_STYLE}`,
    [3 /* OK */]: `--avl-icon-custom-color: var(--avl-ok-color); ${DEFAULT_ICON_STYLE}`
};
export const TRANSLATION_BASE = 'VizBox.NotificationDialog.';
export const ACTION_BUTTONS = [
    {
        id: 'unread',
        tooltip: TRANSLATION_BASE + "MarkSelectedAsUnread"
    },
    {
        id: 'read',
        tooltip: TRANSLATION_BASE + "MarkSelectedAsRead"
    }
];
