// @avl-ng-controls/navigation/esm2015/lib/components/avl-app-bar/avl-app-bar-item.component.js
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
export class AvlAppBarItemComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
    /**
     * @hidden
     */
    ngOnInit() {
        this.elementRef.nativeElement.classList.add('mdc-toolbar__icon');
    }
}
AvlAppBarItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlAppBarItemComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
AvlAppBarItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.3", type: AvlAppBarItemComponent, selector: "avl-app-bar-item", ngImport: i0, template: '<ng-content></ng-content>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlAppBarItemComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'avl-app-bar-item',
                    template: '<ng-content></ng-content>',
                    styleUrls: []
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
