// @avl-controls/ng-binders/esm5/lib/controls/avl-numberfield.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
var AvlNumberfieldDirective = /** @class */ (function () {
    function AvlNumberfieldDirective(elementRef) {
        this.valueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'value-changed', this.valueChange);
    }
    AvlNumberfieldDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlNumberfieldDirective.prototype, "valueChange", void 0);
    AvlNumberfieldDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlNumberfieldDirective);
AvlNumberfieldDirective.ɵfac = function AvlNumberfieldDirective_Factory(t) { return new (t || AvlNumberfieldDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlNumberfieldDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlNumberfieldDirective, selectors: [["avl-numberfield"]], outputs: { valueChange: "valueChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNumberfieldDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-numberfield'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { valueChange: [{
            type: Output
        }] }); })();
    return AvlNumberfieldDirective;
}());
export { AvlNumberfieldDirective };
