// @avl-controls/ng-binders/esm5/lib/controls/avl-textarea.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
var AvlTextareaDirective = /** @class */ (function () {
    function AvlTextareaDirective(elementRef) {
        this.valueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'value-changed', this.valueChange);
    }
    AvlTextareaDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlTextareaDirective.prototype, "valueChange", void 0);
    AvlTextareaDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlTextareaDirective);
AvlTextareaDirective.ɵfac = function AvlTextareaDirective_Factory(t) { return new (t || AvlTextareaDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlTextareaDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlTextareaDirective, selectors: [["avl-textarea"]], outputs: { valueChange: "valueChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlTextareaDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-textarea'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { valueChange: [{
            type: Output
        }] }); })();
    return AvlTextareaDirective;
}());
export { AvlTextareaDirective };
