// @avl-controls/ng-binders/esm5/lib/controls/avl-side-sheet.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
var AvlSideSheetDirective = /** @class */ (function () {
    function AvlSideSheetDirective(elementRef) {
        this.openChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'open-changed', this.openChange);
    }
    AvlSideSheetDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlSideSheetDirective.prototype, "openChange", void 0);
    AvlSideSheetDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlSideSheetDirective);
AvlSideSheetDirective.ɵfac = function AvlSideSheetDirective_Factory(t) { return new (t || AvlSideSheetDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlSideSheetDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlSideSheetDirective, selectors: [["avl-side-sheet"]], outputs: { openChange: "openChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSideSheetDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-side-sheet'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { openChange: [{
            type: Output
        }] }); })();
    return AvlSideSheetDirective;
}());
export { AvlSideSheetDirective };
