// @avl-controls/ng-binders/esm5/lib/controls/avl-combo-box.directive.js
import { __decorate, __extends, __metadata } from "tslib";
import { Directive, ElementRef, EventEmitter, forwardRef, HostListener, Output, Renderer2, Input } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { AvlInteractionControlDirective } from '../interface/avl-interaction-control.directive';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
var AvlComboBoxDirective = /** @class */ (function (_super) {
    __extends(AvlComboBoxDirective, _super);
    function AvlComboBoxDirective(elementRef, renderer) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        _this.renderer = renderer;
        _this.invalidChange = new EventEmitter();
        _this.valueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'invalid-changed', _this.focusedChange);
        addEventListenerForEmittingChangedEvents(elementRef, 'value-changed', _this.valueChange);
        return _this;
    }
    AvlComboBoxDirective_1 = AvlComboBoxDirective;
    Object.defineProperty(AvlComboBoxDirective.prototype, "readonly", {
        // Angular writes readonly attributes to a readOnly property,
        // because that's how it's called on the native input element.
        set: function (value) {
            this.elementRef.nativeElement.readonly = value;
        },
        enumerable: true,
        configurable: true
    });
    AvlComboBoxDirective.prototype.valueChanged = function (e) {
        if (this.onChangeCallback && this.isAllowNewValues()) {
            this.onChangeCallback(e.detail.value);
        }
    };
    AvlComboBoxDirective.prototype.selectedValueChanged = function (e) {
        if (this.onChangeCallback && !this.isAllowNewValues()) {
            this.onChangeCallback(e.detail.value);
        }
    };
    AvlComboBoxDirective.prototype.writeValue = function (value) {
        var property = this.isAllowNewValues() ? 'value' : 'selectedValue';
        this.renderer.setProperty(this.elementRef.nativeElement, property, value);
    };
    AvlComboBoxDirective.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    // Since avl-combo-box doesn't have validate-on-leave, there's no touching
    AvlComboBoxDirective.prototype.registerOnTouched = function (fn) { };
    AvlComboBoxDirective.prototype.setDisabledState = function (isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    };
    AvlComboBoxDirective.prototype.isAllowNewValues = function () {
        return (this.elementRef.nativeElement && this.elementRef.nativeElement.allowNewValues === true);
    };
    //#region Validator
    AvlComboBoxDirective.prototype.validate = function (control) {
        var formControl = this.elementRef.nativeElement;
        if (formControl.invalid) {
            return { invalid: formControl.errorMessage };
        }
        else {
            return null;
        }
    };
    var AvlComboBoxDirective_1;
    AvlComboBoxDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlComboBoxDirective.prototype, "invalidChange", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlComboBoxDirective.prototype, "valueChange", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], AvlComboBoxDirective.prototype, "readonly", null);
    __decorate([
        HostListener('value-changed', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [CustomEvent]),
        __metadata("design:returntype", void 0)
    ], AvlComboBoxDirective.prototype, "valueChanged", null);
    __decorate([
        HostListener('selected-value-changed', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [CustomEvent]),
        __metadata("design:returntype", void 0)
    ], AvlComboBoxDirective.prototype, "selectedValueChanged", null);
    AvlComboBoxDirective = AvlComboBoxDirective_1 = __decorate([ __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], AvlComboBoxDirective);
AvlComboBoxDirective.ɵfac = function AvlComboBoxDirective_Factory(t) { return new (t || AvlComboBoxDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
AvlComboBoxDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlComboBoxDirective, selectors: [["avl-combo-box"]], hostBindings: function AvlComboBoxDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("value-changed", function AvlComboBoxDirective_value_changed_HostBindingHandler($event) { return ctx.valueChanged($event); })("selected-value-changed", function AvlComboBoxDirective_selected_value_changed_HostBindingHandler($event) { return ctx.selectedValueChanged($event); });
    } }, inputs: { readonly: "readonly" }, outputs: { invalidChange: "invalidChange", valueChange: "valueChange" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(function () { return AvlComboBoxDirective_1; }), multi: true },
            { provide: NG_VALIDATORS, useExisting: forwardRef(function () { return AvlComboBoxDirective_1; }), multi: true }]), ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlComboBoxDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-combo-box',
                providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(function () { return AvlComboBoxDirective_1; }), multi: true },
                    { provide: NG_VALIDATORS, useExisting: forwardRef(function () { return AvlComboBoxDirective_1; }), multi: true }]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { readonly: [{
            type: Input
        }], valueChanged: [{
            type: HostListener,
            args: ['value-changed', ['$event']]
        }], selectedValueChanged: [{
            type: HostListener,
            args: ['selected-value-changed', ['$event']]
        }], invalidChange: [{
            type: Output
        }], valueChange: [{
            type: Output
        }] }); })();
    return AvlComboBoxDirective;
}(AvlInteractionControlDirective));
export { AvlComboBoxDirective };
