// @avl-controls/ng-binders/esm5/lib/controls/avl-breadcrumb.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var AvlBreadcrumbDirective = /** @class */ (function () {
    function AvlBreadcrumbDirective() {
        this.itemClicked = new EventEmitter();
    }
    AvlBreadcrumbDirective.prototype.itemClickedListener = function (e) {
        this.itemClicked.emit(e.detail.item);
    };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlBreadcrumbDirective.prototype, "itemClicked", void 0);
    __decorate([
        HostListener('item-clicked', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [CustomEvent]),
        __metadata("design:returntype", void 0)
    ], AvlBreadcrumbDirective.prototype, "itemClickedListener", null);
AvlBreadcrumbDirective.ɵfac = function AvlBreadcrumbDirective_Factory(t) { return new (t || AvlBreadcrumbDirective)(); };
AvlBreadcrumbDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlBreadcrumbDirective, selectors: [["avl-breadcrumb"]], hostBindings: function AvlBreadcrumbDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("item-clicked", function AvlBreadcrumbDirective_item_clicked_HostBindingHandler($event) { return ctx.itemClickedListener($event); });
    } }, outputs: { itemClicked: "itemClicked" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlBreadcrumbDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-breadcrumb'
            }]
    }], function () { return []; }, { itemClicked: [{
            type: Output
        }], itemClickedListener: [{
            type: HostListener,
            args: ['item-clicked', ['$event']]
        }] }); })();
    return AvlBreadcrumbDirective;
}());
export { AvlBreadcrumbDirective };
