// @avl-controls/ng-binders/esm5/lib/controls/avl-tab-bar.directive.js
import { __awaiter, __decorate, __generator, __metadata } from "tslib";
import { Directive, ElementRef, Output, EventEmitter } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var AvlTabBarDirective = /** @class */ (function () {
    function AvlTabBarDirective(el) {
        var _this = this;
        this.el = el;
        this.selectedIndexChange = new EventEmitter();
        this.selectedValueChange = new EventEmitter();
        this.tabBar = this.el.nativeElement;
        Promise.resolve().then(function () {
            _this.setEventListeners(el.nativeElement);
        });
    }
    AvlTabBarDirective.prototype.setEventListeners = function (el) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                el.addEventListener('selected-value-changed', function (event) {
                    Promise.resolve().then(function () {
                        _this.selectedValueChange.emit(event.detail.value);
                    });
                });
                el.addEventListener('selected-index-changed', function (event) {
                    Promise.resolve().then(function () {
                        _this.selectedIndexChange.emit(event.detail.value);
                    });
                });
                return [2 /*return*/];
            });
        });
    };
    AvlTabBarDirective.prototype.setTab = function (index, value) {
        var tabs = this.tabBar.tabs;
        for (var prop in value) {
            if (value.hasOwnProperty(prop)) {
                tabs[index][prop] = value[prop];
                this.tabBar.notifyPath("tabs." + index + "." + prop);
            }
        }
    };
    AvlTabBarDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlTabBarDirective.prototype, "selectedIndexChange", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlTabBarDirective.prototype, "selectedValueChange", void 0);
    AvlTabBarDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlTabBarDirective);
AvlTabBarDirective.ɵfac = function AvlTabBarDirective_Factory(t) { return new (t || AvlTabBarDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlTabBarDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlTabBarDirective, selectors: [["avl-tab-bar"]], outputs: { selectedIndexChange: "selectedIndexChange", selectedValueChange: "selectedValueChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlTabBarDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-tab-bar'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { selectedIndexChange: [{
            type: Output
        }], selectedValueChange: [{
            type: Output
        }] }); })();
    return AvlTabBarDirective;
}());
export { AvlTabBarDirective };
