// @avl-controls/ng-binders/esm5/lib/controls/avl-time.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, EventEmitter, Output, ElementRef } from '@angular/core';
import { addEventListenerForEmittingChangedEvents } from '../binding-utils';
import * as ɵngcc0 from '@angular/core';
var AvlTimeDirective = /** @class */ (function () {
    function AvlTimeDirective(elementRef) {
        this.valueChange = new EventEmitter();
        addEventListenerForEmittingChangedEvents(elementRef, 'value-changed', this.valueChange);
    }
    AvlTimeDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], AvlTimeDirective.prototype, "valueChange", void 0);
    AvlTimeDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
    ], AvlTimeDirective);
AvlTimeDirective.ɵfac = function AvlTimeDirective_Factory(t) { return new (t || AvlTimeDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlTimeDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlTimeDirective, selectors: [["avl-time"]], outputs: { valueChange: "valueChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlTimeDirective, [{
        type: Directive,
        args: [{
                selector: 'avl-time'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { valueChange: [{
            type: Output
        }] }); })();
    return AvlTimeDirective;
}());
export { AvlTimeDirective };
