// @avl-ng-controls/navigation/esm2015/lib/avl-navigation.module.js
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { AvlCommonModule } from '@avl-ng-controls/common';
// toolbar stuff
import { AvlAppBarComponent } from './components/avl-app-bar/avl-app-bar.component';
import { AvlAppBarItemComponent } from './components/avl-app-bar/avl-app-bar-item.component';
// sidenav stuff
import { AvlSideNavComponent } from './components/avl-side-nav/avl-side-nav.component';
import { AvlAppLayoutComponent } from './components/avl-app-layout/avl-app-layout.component';
import { AvlNoPaddingDirective } from './components/avl-app-layout/avl-no-padding.directive';
import { AvlLongPageComponent } from './components/avl-long-page/avl-long-page.component';
import { AvlParagraphDirective } from './components/avl-long-page/avl-paragraph.directive';
import { AvlSolutionOverviewComponent } from './components/avl-solution-overview/avl-solution-overview.component';
import { AvlNgBindersModule } from '@avl-controls/ng-binders';
import { AvlListExpanderComponent } from './components/avl-list-expander/avl-list-expander.component';
import * as i0 from "@angular/core";
const componentsAndDirectives = [
    AvlAppBarComponent,
    AvlAppBarItemComponent,
    AvlSideNavComponent,
    AvlAppLayoutComponent,
    AvlNoPaddingDirective,
    AvlLongPageComponent,
    AvlParagraphDirective,
    AvlSolutionOverviewComponent,
    AvlListExpanderComponent
];
export class AvlNavigationModule {
}
AvlNavigationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNavigationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AvlNavigationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNavigationModule, declarations: [AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent], imports: [CommonModule,
        RouterModule,
        AvlNgBindersModule,
        AvlCommonModule], exports: [AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent] });
AvlNavigationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNavigationModule, imports: [[
            CommonModule,
            RouterModule,
            AvlNgBindersModule,
            AvlCommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNavigationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        AvlNgBindersModule,
                        AvlCommonModule
                    ],
                    declarations: componentsAndDirectives,
                    schemas: [
                        CUSTOM_ELEMENTS_SCHEMA
                    ],
                    exports: componentsAndDirectives
                }]
        }] });
