// @avl-services/ng-services/esm2015/lib/translation-management.service.js
import { __decorate } from "tslib";
/// <reference types="@avl-controls/core" />
import { Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { NgxTranslateTranslationService } from '@avl-controls/ng-binders';
import '@avl-controls/core/dist/avl-localization-service/avl-localization-service';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "@avl-controls/ng-binders";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-translate/core';
import * as ɵngcc2 from '@avl-controls/ng-binders';
let TranslationManagementService = class TranslationManagementService {
    constructor(translateService, translationService) {
        this.translateService = translateService;
        this.translationService = translationService;
        this.languages = [
            { code: 'en', label: 'English' },
            { code: 'hr', label: 'Croatian' },
        ];
    }
    initializeTranslations() {
        return () => {
            AvlLocalizationService.translationService = this.translationService;
            let lang = localStorage.getItem('lang');
            if (lang == null) {
                lang = this.languages[0].code;
                localStorage.setItem('lang', lang);
            }
            return this.translateService.use(lang).toPromise();
        };
    }
    get language() {
        return localStorage.getItem('lang');
    }
    changeLanguage(lang) {
        const oldLang = localStorage.getItem('lang');
        if (!oldLang || oldLang !== lang) {
            localStorage.setItem('lang', lang);
            location.reload();
        }
    }
};
TranslationManagementService.ɵfac = function TranslationManagementService_Factory(t) { return new (t || TranslationManagementService)(ɵngcc0.ɵɵinject(ɵngcc1.TranslateService), ɵngcc0.ɵɵinject(ɵngcc2.NgxTranslateTranslationService)); };
TranslationManagementService.ctorParameters = () => [
    { type: TranslateService },
    { type: NgxTranslateTranslationService }
];
TranslationManagementService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TranslationManagementService_Factory() { return new TranslationManagementService(i0.ɵɵinject(i1.TranslateService), i0.ɵɵinject(i2.NgxTranslateTranslationService)); }, token: TranslationManagementService, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TranslationManagementService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.TranslateService }, { type: ɵngcc2.NgxTranslateTranslationService }]; }, null); })();
export { TranslationManagementService };
