// @avl-ng-controls/dialog/esm5/lib/avl-dialog/avl-dialog-section.base.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { HostBinding, Directive } from '@angular/core';
import { AvlDialogComponent } from './avl-dialog.component';
var AvlBaseDialogSection = /** @class */ (function () {
    function AvlBaseDialogSection(dialog) {
        this.dialog = dialog;
        this.isComplex = false;
    }
    /** @hidden */
    AvlBaseDialogSection.prototype.closeDialog = function () {
        this.dialog.closeDialog();
    };
    AvlBaseDialogSection.prototype.ngOnInit = function () {
        this.isComplex = this.dialog.complex;
    };
    AvlBaseDialogSection.ctorParameters = function () { return [
        { type: AvlDialogComponent }
    ]; };
    __decorate([
        HostBinding('class.complex'),
        __metadata("design:type", Boolean)
    ], AvlBaseDialogSection.prototype, "isComplex", void 0);
    AvlBaseDialogSection = __decorate([ __metadata("design:paramtypes", [AvlDialogComponent])
    ], AvlBaseDialogSection);
AvlBaseDialogSection.ɵfac = function AvlBaseDialogSection_Factory(t) { return new (t || AvlBaseDialogSection)(ɵngcc0.ɵɵdirectiveInject(AvlDialogComponent)); };
AvlBaseDialogSection.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlBaseDialogSection, hostVars: 2, hostBindings: function AvlBaseDialogSection_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("complex", ctx.isComplex);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlBaseDialogSection, [{
        type: Directive
    }], function () { return [{ type: AvlDialogComponent }]; }, { isComplex: [{
            type: HostBinding,
            args: ['class.complex']
        }] }); })();
    return AvlBaseDialogSection;
}());
export { AvlBaseDialogSection };
