// @avl-ng-controls/navigation/bundles/avl-ng-controls-navigation.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/router'),require('@avl-controls/ng-binders'),exports, require('@angular/core'), require('@angular/common'), require('@angular/router'), require('@avl-controls/core/dist/avl-app-switcher/avl-app-switcher'), require('@avl-controls/core/dist/avl-icon/avl-icon.js'), require('@angular/animations'), require('rxjs/operators'), require('@avl-controls/core/dist/avl-icon/avl-icon'), require('rxjs'), require('@avl-controls/core/dist/avl-search-bar/avl-search-bar'), require('@avl-controls/ng-binders')) :
    typeof define === 'function' && define.amd ? define('@avl-ng-controls/navigation', ['@angular/core','@angular/common','@angular/router','@avl-controls/ng-binders','exports', '@angular/core', '@angular/common', '@angular/router', '@avl-controls/core/dist/avl-app-switcher/avl-app-switcher', '@avl-controls/core/dist/avl-icon/avl-icon.js', '@angular/animations', 'rxjs/operators', '@avl-controls/core/dist/avl-icon/avl-icon', 'rxjs', '@avl-controls/core/dist/avl-search-bar/avl-search-bar', '@avl-controls/ng-binders'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,global.ng.router,global.avlControls.ngBinders,(global['avl-ng-controls'] = global['avl-ng-controls'] || {}, global['avl-ng-controls'].navigation = {}), global.ng.core, global.ng.common, global.ng.router, null, null, global.ng.animations, global.rxjs.operators, null, global.rxjs, null, global.ngBinders));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, router, avlAppSwitcher, avlIcon_js, animations, operators, avlIcon, rxjs, avlSearchBar, ngBinders) { 
function AvlAppBarComponent_avl_app_bar_item_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "avl-app-bar-item", 10);
    ɵngcc0.ɵɵelement(1, "a", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("background-image", "url(" + ctx_r0.appLogo + ")", ɵngcc0.ɵɵdefaultStyleSanitizer)("background-repeat", "no-repeat");
    ɵngcc0.ɵɵpropertyInterpolate("routerLink", ctx_r0.homePageUrl ? ctx_r0.homePageUrl : "/");
} }
function AvlAppBarComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div");
} }
function AvlAppBarComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-app-bar-item", 12);
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("innerHTML", ctx_r3.currentApplication == null ? null : ctx_r3.currentApplication.name, ɵngcc0.ɵɵsanitizeHtml);
} }
function AvlAppBarComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    var _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "avl-app-bar-item", 13);
    ɵngcc0.ɵɵelementStart(1, "avl-app-switcher", 14);
    ɵngcc0.ɵɵlistener("itemsChange", function AvlAppBarComponent_ng_template_9_Template_avl_app_switcher_itemsChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); var ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.onApplicationListChanged($event); })("selectedValueChange", function AvlAppBarComponent_ng_template_9_Template_avl_app_switcher_selectedValueChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); var ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.onApplicationChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("items", ctx_r5.applications)("selectedValue", ctx_r5.currentApplication);
} }
var _c0 = ["*", [["avl-overflow-menu", "slot", "menu"]]];
var _c1 = function (a0) { return { "mdc-toolbar--fixed": a0 }; };
var _c2 = ["*", "avl-overflow-menu[slot=menu]"];
var _c3 = ["*"];
function AvlSideNavComponent_ng_template_5_a_0_avl_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-icon");
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵattribute("icon", item_r3.icon);
} }
function AvlSideNavComponent_ng_template_5_a_0_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 21);
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("src", item_r3.iconUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
function AvlSideNavComponent_ng_template_5_a_0_Template(rf, ctx) { if (rf & 1) {
    var _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 15);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_a_0_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); var ctx_r12 = ɵngcc0.ɵɵnextContext(2); return ctx_r12.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_a_0_avl_icon_2_Template, 1, 1, "avl-icon", 17);
    ɵngcc0.ɵɵtemplate(3, AvlSideNavComponent_ng_template_5_a_0_img_3_Template, 1, 1, "img", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 19);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "div", 20);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("selected", item_r3 === ctx_r4.selectedPage);
    ɵngcc0.ɵɵpropertyInterpolate("routerLink", item_r3.route);
    ɵngcc0.ɵɵproperty("queryParams", ctx_r4.getParams(item_r3.params));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.iconUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r3.label);
} }
function AvlSideNavComponent_ng_template_5_div_1_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 27);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_1_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r18); var ctx_r17 = ɵngcc0.ɵɵnextContext(3); return ctx_r17.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var paragraph_r16 = ctx.$implicit;
    var item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r15 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("selected", paragraph_r16.id === ctx_r15.selectedParagraph);
    ɵngcc0.ɵɵpropertyInterpolate("href", ctx_r15.trimSlash(item_r3.route + ctx_r15.getUrlParams(item_r3.params) + "#" + paragraph_r16.id), ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(paragraph_r16.label);
} }
function AvlSideNavComponent_ng_template_5_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_1_a_2_Template, 3, 4, "a", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 24);
    ɵngcc0.ɵɵelement(4, "div", 25);
    ɵngcc0.ɵɵelement(5, "div", 26);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@expandCollapse", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", item_r3.paragraphs);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r5.getScrubberStyle());
} }
function AvlSideNavComponent_ng_template_5_div_2_avl_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-icon");
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵattribute("icon", item_r3.icon);
} }
function AvlSideNavComponent_ng_template_5_div_2_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 21);
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("src", item_r3.iconUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
function AvlSideNavComponent_ng_template_5_div_2_Template(rf, ctx) { if (rf & 1) {
    var _r27 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_2_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r27); var item_r3 = ɵngcc0.ɵɵnextContext().$implicit; var ctx_r25 = ɵngcc0.ɵɵnextContext(); return ctx_r25.groupClicked(item_r3); });
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_2_avl_icon_2_Template, 1, 1, "avl-icon", 17);
    ɵngcc0.ɵɵtemplate(3, AvlSideNavComponent_ng_template_5_div_2_img_3_Template, 1, 1, "img", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 29);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "avl-icon", 30);
    ɵngcc0.ɵɵelement(7, "div", 20);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("expanded", item_r3.expanded)("selected", !ctx_r6.open && item_r3 === ctx_r6.selectedGroup);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.iconUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r3.label);
} }
function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r35 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 34);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r35); var ctx_r34 = ɵngcc0.ɵɵnextContext(5); return ctx_r34.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var paragraph_r33 = ctx.$implicit;
    var subItem_r30 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r32 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("selected", paragraph_r33.id === ctx_r32.selectedParagraph);
    ɵngcc0.ɵɵpropertyInterpolate3("href", "", subItem_r30.route, "", ctx_r32.getUrlParams(item_r3.params), "#", paragraph_r33.id, "", ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(paragraph_r33.label);
} }
function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_a_2_Template, 3, 6, "a", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 24);
    ɵngcc0.ɵɵelement(4, "div", 25);
    ɵngcc0.ɵɵelement(5, "div", 26);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var subItem_r30 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r31 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("@expandCollapse", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", subItem_r30.paragraphs);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r31.getScrubberStyle());
} }
function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    var _r40 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 32);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_3_ng_template_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r40); var ctx_r39 = ɵngcc0.ɵɵnextContext(3); return ctx_r39.mobileClose(); });
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "div", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, AvlSideNavComponent_ng_template_5_div_3_ng_template_2_div_4_Template, 6, 3, "div", 12);
} if (rf & 2) {
    var subItem_r30 = ctx.$implicit;
    var ctx_r29 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassProp("selected", subItem_r30 === ctx_r29.selectedPage);
    ɵngcc0.ɵɵpropertyInterpolate("routerLink", subItem_r30.route);
    ɵngcc0.ɵɵproperty("queryParams", ctx_r29.getParams(subItem_r30.params));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(subItem_r30.label);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", subItem_r30.paragraphs && subItem_r30.paragraphs.length > 0 && subItem_r30 === ctx_r29.selectedPage);
} }
function AvlSideNavComponent_ng_template_5_div_3_Template(rf, ctx) { if (rf & 1) {
    var _r42 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 31);
    ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_ng_template_5_div_3_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r42); var ctx_r41 = ɵngcc0.ɵɵnextContext(2); return ctx_r41.toggleOpen(); });
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_3_ng_template_2_Template, 5, 6, "ng-template", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("@expandCollapse", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", item_r3.subItems);
} }
function AvlSideNavComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, AvlSideNavComponent_ng_template_5_a_0_Template, 7, 7, "a", 11);
    ɵngcc0.ɵɵtemplate(1, AvlSideNavComponent_ng_template_5_div_1_Template, 6, 3, "div", 12);
    ɵngcc0.ɵɵtemplate(2, AvlSideNavComponent_ng_template_5_div_2_Template, 8, 7, "div", 13);
    ɵngcc0.ɵɵtemplate(3, AvlSideNavComponent_ng_template_5_div_3_Template, 3, 2, "div", 14);
} if (rf & 2) {
    var item_r3 = ctx.$implicit;
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", !item_r3.subItems);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.open && item_r3.paragraphs && item_r3.paragraphs.length > 0 && item_r3 === ctx_r2.selectedPage);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.subItems);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.open && item_r3.subItems && item_r3.subItems.length > 0 && item_r3.expanded);
} }
var _c4 = [[["", "slot", "toolbar"]], [["", "slot", "left-side"]], "*", [["", "slot", "right-side"]]];
var _c5 = ["[slot=toolbar]", "[slot=left-side]", "*", "[slot=right-side]"];
var _c6 = ["searchbar"];
var _c7 = ["iconButtons"];
var _c8 = ["searchbarContainer"];
var _c9 = ["groupsContainer"];
var _c10 = ["itemElement"];
var _c11 = ["group"];
function AvlSolutionOverviewComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "h2", 17);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.solutionSubtitle);
} }
function AvlSolutionOverviewComponent_avl_search_bar_14_Template(rf, ctx) { if (rf & 1) {
    var _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "avl-search-bar", 18, 19);
    ɵngcc0.ɵɵlistener("query-changed", function AvlSolutionOverviewComponent_avl_search_bar_14_Template_avl_search_bar_query_changed_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onQueryChange($event); })("search", function AvlSolutionOverviewComponent_avl_search_bar_14_Template_avl_search_bar_search_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onSearch($event); });
    ɵngcc0.ɵɵelementEnd();
} }
function AvlSolutionOverviewComponent_ng_container_17_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var _c12 = function (a0) { return { group: a0 }; };
function AvlSolutionOverviewComponent_ng_container_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 20);
    ɵngcc0.ɵɵelementStart(2, "div", 21, 22);
    ɵngcc0.ɵɵtemplate(4, AvlSolutionOverviewComponent_ng_container_17_ng_container_4_Template, 1, 0, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var group_r14 = ctx.$implicit;
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    var _r6 = ɵngcc0.ɵɵreference(19);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r5.getMaxColumnsClass());
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r6)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c12, group_r14));
} }
function AvlSolutionOverviewComponent_ng_template_18_h2_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h2", 26);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(group_r17.title);
} }
function AvlSolutionOverviewComponent_ng_template_18_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(group_r17.header);
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var _c13 = function (a0) { return { "with-description": a0 }; };
var _c14 = function (a0) { return { item: a0 }; };
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 29, 30);
    ɵngcc0.ɵɵtemplate(3, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_ng_container_3_Template, 1, 0, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r24 = ɵngcc0.ɵɵnextContext().$implicit;
    var group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵnextContext();
    var _r8 = ɵngcc0.ɵɵreference(21);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("routerLink", item_r24.route)("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c13, group_r17.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r8)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(6, _c14, item_r24));
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 31, 30);
    ɵngcc0.ɵɵtemplate(3, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_ng_container_3_Template, 1, 0, "ng-container", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r24 = ɵngcc0.ɵɵnextContext().$implicit;
    var group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵnextContext();
    var _r8 = ɵngcc0.ɵɵreference(21);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(6, _c13, group_r17.description));
    ɵngcc0.ɵɵattribute("href", item_r24.link, ɵngcc0.ɵɵsanitizeUrl)("target", item_r24.target)("rel", item_r24.target == "_blank" ? "noopener" : null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r8)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(8, _c14, item_r24));
} }
function AvlSolutionOverviewComponent_ng_template_18_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_1_Template, 4, 8, "ng-container", 28);
    ɵngcc0.ɵɵtemplate(2, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_ng_container_2_Template, 4, 10, "ng-container", 28);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r24 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r24.route);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r24.link);
} }
function AvlSolutionOverviewComponent_ng_template_18_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var group_r17 = ɵngcc0.ɵɵnextContext().group;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(group_r17.footer);
} }
function AvlSolutionOverviewComponent_ng_template_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, AvlSolutionOverviewComponent_ng_template_18_h2_0_Template, 2, 1, "h2", 24);
    ɵngcc0.ɵɵtemplate(1, AvlSolutionOverviewComponent_ng_template_18_div_1_Template, 2, 1, "div", 25);
    ɵngcc0.ɵɵtemplate(2, AvlSolutionOverviewComponent_ng_template_18_ng_container_2_Template, 3, 2, "ng-container", 14);
    ɵngcc0.ɵɵtemplate(3, AvlSolutionOverviewComponent_ng_template_18_div_3_Template, 2, 1, "div", 25);
} if (rf & 2) {
    var group_r17 = ctx.group;
    ɵngcc0.ɵɵproperty("ngIf", group_r17.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", group_r17.header);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", group_r17.items);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", group_r17.footer);
} }
function AvlSolutionOverviewComponent_ng_template_20_avl_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "avl-icon", 39);
} if (rf & 2) {
    var item_r36 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵproperty("icon", item_r36.icon);
} }
function AvlSolutionOverviewComponent_ng_template_20_img_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 40);
} if (rf & 2) {
    var item_r36 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵpropertyInterpolate("src", item_r36.iconUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
var _c15 = function (a0) { return { "no-icon-circle": a0 }; };
function AvlSolutionOverviewComponent_ng_template_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 32);
    ɵngcc0.ɵɵtemplate(1, AvlSolutionOverviewComponent_ng_template_20_avl_icon_1_Template, 1, 1, "avl-icon", 33);
    ɵngcc0.ɵɵtemplate(2, AvlSolutionOverviewComponent_ng_template_20_img_2_Template, 1, 1, "img", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 35);
    ɵngcc0.ɵɵelementStart(4, "span", 36);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 37);
    ɵngcc0.ɵɵelementStart(7, "span", 38);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r36 = ctx.item;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c15, item_r36.noIconCircle));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r36.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r36.iconUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(item_r36.label);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r36.description, " ");
} }
var _c16 = function (a0) { return { "top-margin": a0 }; };
function AvlListExpanderComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelement(1, "avl-icon", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r0.narrow ? "icon-narrow" : "icon");
} }
function AvlListExpanderComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelement(1, "avl-icon", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.narrow ? "icon-narrow" : "icon");
} }
function AvlListExpanderComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵprojection(1, 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.narrow ? "content-narrow" : "content");
} }
var _c17 = [[["div", "role", "header"]], [["div", "role", "content"]]];
var _c18 = ["div[role=header]", "div[role=content]"];
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var MenuButtonItems = /** @class */ (function () {
        function MenuButtonItems(headerText, headerIconName, items, footerItems) {
            this.headerText = headerText;
            this.headerIconName = headerIconName;
            this.items = items;
            this.footerItems = footerItems;
        }
        return MenuButtonItems;
    }());
    /**
     * Object representing sub-application that is added to app-bar drop-down
     */
    var ApplicationReference = /** @class */ (function () {
        function ApplicationReference() {
        }
        return ApplicationReference;
    }());
    var AvlAppBarComponent = /** @class */ (function () {
        function AvlAppBarComponent() {
            /**
             * List of sub-applications used
             */
            this.applications = [];
            /**
             * Event emitted when new application is added.
             */
            this.applicationListChange = new core.EventEmitter();
            /**
             * Flag indicating app-bar has fixed positioning or no
             */
            this.fixed = true;
            /**
             * Event emitted when application is changed in app-bar drop-down
             */
            this.currentApplicationChange = new core.EventEmitter();
            /**
             * Event emmited when Burger menu is clicked.
             * Toggles side-nav visibility if app-bar is part of app-layout in responsive mode
             */
            this.burgerClicked = new core.EventEmitter();
            /** @hidden */
            this.appLogoSet = false;
            this.appLogoPreset = false;
        }
        Object.defineProperty(AvlAppBarComponent.prototype, "isSingleApp", {
            /**
             * True if only one element added to applications array
             */
            get: function () {
                return !this.applications || this._appsLength < 2;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AvlAppBarComponent.prototype, "showAppLogo", {
            /**
             * @hidden
             */
            get: function () {
                if (this.currentApplication == null)
                    return false;
                var image = this.currentApplication.image;
                return !(image == null || image === '');
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @hidden
         */
        AvlAppBarComponent.prototype.burgerIconClicked = function () {
            this.burgerClicked.emit();
        };
        /**
         * @hidden
         */
        AvlAppBarComponent.prototype.onApplicationChanged = function (app) {
            var _this = this;
            this.currentApplication = app;
            if (!this.appLogoPreset) {
                setTimeout(function () {
                    _this.appLogoSet = true;
                    _this.appLogo = _this.currentApplication.image;
                });
            }
            this.currentApplicationChange.emit(this.currentApplication);
        };
        /**
         * @hidden
         */
        AvlAppBarComponent.prototype.onApplicationListChanged = function (apps) {
            this.applications = apps;
            this.applicationListChange.emit(apps);
        };
        /** @hidden */
        AvlAppBarComponent.prototype.ngOnInit = function () {
            if (this.appLogo) {
                this.appLogoSet = true;
                this.appLogoPreset = true;
            }
            else {
                this.appLogo = this.currentApplication ? this.currentApplication.image : '';
                this.appLogoSet = true;
            }
            this._appsLength = this.applications ? this.applications.length : 0;
        };
        __decorate([
            core.Input(),
            __metadata("design:type", ApplicationReference)
        ], AvlAppBarComponent.prototype, "currentApplication", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('class.responsive'),
            __metadata("design:type", Boolean)
        ], AvlAppBarComponent.prototype, "responsive", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('class.mobile'),
            __metadata("design:type", Boolean)
        ], AvlAppBarComponent.prototype, "mobile", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], AvlAppBarComponent.prototype, "homePageUrl", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], AvlAppBarComponent.prototype, "appLogo", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Array)
        ], AvlAppBarComponent.prototype, "applications", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], AvlAppBarComponent.prototype, "applicationListChange", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], AvlAppBarComponent.prototype, "fixed", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], AvlAppBarComponent.prototype, "currentApplicationChange", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], AvlAppBarComponent.prototype, "burgerClicked", void 0);
AvlAppBarComponent.ɵfac = function AvlAppBarComponent_Factory(t) { return new (t || AvlAppBarComponent)(); };
AvlAppBarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlAppBarComponent, selectors: [["avl-app-bar"]], hostVars: 4, hostBindings: function AvlAppBarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("responsive", ctx.responsive)("mobile", ctx.mobile);
    } }, inputs: { applications: "applications", fixed: "fixed", currentApplication: "currentApplication", appLogo: "appLogo", responsive: "responsive", mobile: "mobile", homePageUrl: "homePageUrl" }, outputs: { applicationListChange: "applicationListChange", currentApplicationChange: "currentApplicationChange", burgerClicked: "burgerClicked" }, ngContentSelectors: _c2, decls: 14, vars: 7, consts: [["id", "header", 1, "mdc-toolbar", "mdc-toolbar--waterfall", 3, "ngClass"], [1, "mdc-toolbar__row"], [1, "mdc-toolbar__section", "mdc-toolbar__section--align-start", "mdc-toolbar__section--shrink-to-fit", "section-with-switcher"], [1, "burger-container"], ["icon", "menu", "theme", "dark", 1, "burgerMenu", 3, "click"], ["class", "noselect", 4, "ngIf"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["singleApp", ""], ["multiApp", ""], ["id", "additionalContent", 1, "mdc-toolbar__section", "mdc-toolbar__section--align-end"], [1, "noselect"], [1, "avl-app-logo", 3, "routerLink"], [1, "single-app-name", 3, "innerHTML"], [1, "toolbar-item-with-switcher"], [3, "items", "selectedValue", "itemsChange", "selectedValueChange"]], template: function AvlAppBarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "header", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "section", 2);
        ɵngcc0.ɵɵelementStart(3, "avl-app-bar-item", 3);
        ɵngcc0.ɵɵelementStart(4, "avl-icon", 4);
        ɵngcc0.ɵɵlistener("click", function AvlAppBarComponent_Template_avl_icon_click_4_listener() { return ctx.burgerIconClicked(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, AvlAppBarComponent_avl_app_bar_item_5_Template, 2, 5, "avl-app-bar-item", 5);
        ɵngcc0.ɵɵtemplate(6, AvlAppBarComponent_div_6_Template, 1, 0, "div", 6);
        ɵngcc0.ɵɵtemplate(7, AvlAppBarComponent_ng_template_7_Template, 1, 1, "ng-template", null, 7, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(9, AvlAppBarComponent_ng_template_9_Template, 2, 2, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "section", 9);
        ɵngcc0.ɵɵprojection(12);
        ɵngcc0.ɵɵprojection(13, 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r2 = ɵngcc0.ɵɵreference(8);
        var _r4 = ɵngcc0.ɵɵreference(10);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx.fixed));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.appLogoSet && ctx.appLogo != undefined);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSingleApp)("ngIfThen", _r2)("ngIfElse", _r4);
    } }, directives: function () { return [ɵngcc1.NgClass, AvlAppBarItemComponent, ɵngcc1.NgIf, ɵngcc2.RouterLinkWithHref, ɵngcc3.AvlSingleSelectControlDirective]; }, styles: [".mdc-toolbar[_ngcontent-%COMP%]{background-color:#3f51b5;background-color:var(--mdc-theme-primary,#3f51b5);color:#fff;color:var(--mdc-theme-text-primary-on-primary,#fff);display:flex;position:relative;flex-direction:column;justify-content:space-between;box-sizing:border-box;width:100%}.mdc-toolbar__row[_ngcontent-%COMP%]{display:flex;position:relative;align-items:center;box-sizing:border-box;width:100%;height:auto;min-height:64px}.mdc-toolbar__section[_ngcontent-%COMP%]{display:inline-flex;flex:1;align-items:flex-start;justify-content:center;min-width:0;height:100%}.mdc-toolbar__section--align-start[_ngcontent-%COMP%]{justify-content:flex-start;order:-1}.mdc-toolbar__section--align-end[_ngcontent-%COMP%]{justify-content:flex-end;order:1}.mdc-toolbar__title[_ngcontent-%COMP%]{font-family:Roboto,sans-serif;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;font-size:1.25rem;font-weight:500;letter-spacing:.02em;text-decoration:inherit;text-transform:inherit;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;margin-left:24px;margin-right:0;align-self:center;padding:16px 0;line-height:1.5rem;z-index:1}.mdc-toolbar__title[dir=rtl][_ngcontent-%COMP%], [dir=rtl][_ngcontent-%COMP%]   .mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:0;margin-right:24px}.mdc-toolbar__icon[_ngcontent-%COMP%]{display:flex;align-items:center;justify-content:center;border:none;background-color:inherit;text-decoration:none;color:#fff;color:var(--mdc-theme-text-primary-on-primary,#fff);cursor:pointer}.mdc-toolbar__icon[_ngcontent-%COMP%]:last-of-type[dir=rtl], [dir=rtl][_ngcontent-%COMP%]   .mdc-toolbar__icon[_ngcontent-%COMP%]:last-of-type{padding-left:24px;padding-right:12px}.mdc-toolbar__menu-icon[_ngcontent-%COMP%]{display:flex;align-items:center;justify-content:center;padding:16px 24px;border:none;background-color:inherit;text-decoration:none;color:#fff;color:var(--mdc-theme-text-primary-on-primary,#fff);cursor:pointer}.mdc-toolbar__menu-icon[_ngcontent-%COMP%] + .mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:8px;margin-right:0}.mdc-toolbar__menu-icon[_ngcontent-%COMP%] + .mdc-toolbar__title[dir=rtl][_ngcontent-%COMP%], [dir=rtl][_ngcontent-%COMP%]   .mdc-toolbar__menu-icon[_ngcontent-%COMP%] + .mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:0;margin-right:8px}.mdc-toolbar--fixed[_ngcontent-%COMP%]{box-shadow:0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12);position:fixed;top:0;left:0;z-index:4;box-shadow:none}.mdc-toolbar--flexible[_ngcontent-%COMP%]{--mdc-toolbar-ratio-to-extend-flexible:4}.mdc-toolbar--flexible[_ngcontent-%COMP%]   .mdc-toolbar__row[_ngcontent-%COMP%]:first-child{height:256px;height:calc(64px * var(--mdc-toolbar-ratio-to-extend-flexible,4))}.mdc-toolbar--flexible[_ngcontent-%COMP%]   .mdc-toolbar__row[_ngcontent-%COMP%]:first-child::after{position:absolute;content:\"\"}.mdc-toolbar--flexible-default-behavior[_ngcontent-%COMP%]   .mdc-toolbar__title[_ngcontent-%COMP%]{font-family:Roboto,sans-serif;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;font-size:2.125rem;font-weight:400;letter-spacing:normal;text-decoration:inherit;text-transform:inherit;align-self:flex-end;line-height:1.5rem}.mdc-toolbar--flexible-default-behavior[_ngcontent-%COMP%]   .mdc-toolbar__row[_ngcontent-%COMP%]:first-child::after{top:0;left:0;width:100%;height:100%;transition:opacity .2s;opacity:1}.mdc-toolbar--flexible-default-behavior.mdc-toolbar--flexible-space-minimized[_ngcontent-%COMP%]   .mdc-toolbar__row[_ngcontent-%COMP%]:first-child::after{opacity:0}.mdc-toolbar--flexible-default-behavior.mdc-toolbar--flexible-space-minimized[_ngcontent-%COMP%]   .mdc-toolbar__title[_ngcontent-%COMP%]{font-weight:500}.mdc-toolbar--waterfall.mdc-toolbar--fixed[_ngcontent-%COMP%]{box-shadow:0 0 0 0 rgba(0,0,0,.2),0 0 0 0 rgba(0,0,0,.14),0 0 0 0 rgba(0,0,0,.12);transition:box-shadow 280ms cubic-bezier(.4,0,.2,1);will-change:box-shadow}.mdc-toolbar--waterfall.mdc-toolbar--fixed.mdc-toolbar--flexible-space-minimized[_ngcontent-%COMP%]{box-shadow:0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12)}.mdc-toolbar--waterfall.mdc-toolbar--fixed.mdc-toolbar--fixed-lastrow-only.mdc-toolbar--flexible-space-minimized[_ngcontent-%COMP%]{box-shadow:0 0 0 0 rgba(0,0,0,.2),0 0 0 0 rgba(0,0,0,.14),0 0 0 0 rgba(0,0,0,.12)}.mdc-toolbar--waterfall.mdc-toolbar--fixed.mdc-toolbar--fixed-lastrow-only.mdc-toolbar--fixed-at-last-row[_ngcontent-%COMP%]{box-shadow:0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12)}.mdc-toolbar-fixed-adjust[_ngcontent-%COMP%]{margin-top:64px}@media (max-width:959px) and (orientation:landscape){.mdc-toolbar__row[_ngcontent-%COMP%]{min-height:48px}.mdc-toolbar-fixed-adjust[_ngcontent-%COMP%]{margin-top:48px}}@media (max-width:599px){.mdc-toolbar__row[_ngcontent-%COMP%]{min-height:56px}.mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:16px;margin-right:0}.mdc-toolbar__title[dir=rtl][_ngcontent-%COMP%], [dir=rtl][_ngcontent-%COMP%]   .mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:0;margin-right:16px}.mdc-toolbar__icon[_ngcontent-%COMP%]{padding-right:8px;padding-left:8px}.mdc-toolbar__icon[_ngcontent-%COMP%]:last-of-type{padding-left:8px;padding-right:16px}.mdc-toolbar__icon[_ngcontent-%COMP%]:last-of-type[dir=rtl], [dir=rtl][_ngcontent-%COMP%]   .mdc-toolbar__icon[_ngcontent-%COMP%]:last-of-type{padding-left:16px;padding-right:8px}.mdc-toolbar__menu-icon[_ngcontent-%COMP%]{padding-right:16px;padding-left:16px}.mdc-toolbar__menu-icon[_ngcontent-%COMP%] + .mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:16px;margin-right:0}.mdc-toolbar__menu-icon[_ngcontent-%COMP%] + .mdc-toolbar__title[dir=rtl][_ngcontent-%COMP%], [dir=rtl][_ngcontent-%COMP%]   .mdc-toolbar__menu-icon[_ngcontent-%COMP%] + .mdc-toolbar__title[_ngcontent-%COMP%]{margin-left:0;margin-right:16px}.mdc-toolbar--flexible[_ngcontent-%COMP%]   .mdc-toolbar__row[_ngcontent-%COMP%]:first-child{height:224px;height:calc(56px * var(--mdc-toolbar-ratio-to-extend-flexible,4))}.mdc-toolbar-fixed-adjust[_ngcontent-%COMP%]{margin-top:56px}}.mdc-toolbar__section--shrink-to-fit[_ngcontent-%COMP%]{flex:none}[_nghost-%COMP%]{--item-color:var(--avl-app-bar-item-color, var(--avl-on-primary-color));--bar-height:var(--avl-app-bar-height, 53px);--background-color:var(--avl-app-bar-background-color, var(--avl-primary-color));--line-underline-color:var(--avl-app-bar-underline-color, var(--avl-product-color));color:var(--item-color);font-family:var(--avl-font-family);font-size:20px}#additionalContent[_ngcontent-%COMP%]{margin-right:12px;z-index:2}.material-icons[_ngcontent-%COMP%]{font-family:'Material Icons';font-weight:400;font-style:normal;font-size:24px;display:inline-block;line-height:1;text-transform:none;letter-spacing:normal;word-wrap:normal;white-space:nowrap;direction:ltr;-webkit-font-smoothing:antialiased;text-rendering:optimizeLegibility;-moz-osx-font-smoothing:grayscale;font-feature-settings:'liga'}.mdc-toolbar[_ngcontent-%COMP%]{background-color:var(--background-color);background-repeat:no-repeat;background-position:calc(100% - 61px) -30px;background-size:158px;color:var(--item-color);box-shadow:none;z-index:4;border-bottom:3px solid var(--line-underline-color)}#header[_ngcontent-%COMP%]{box-shadow:var(--avl-box-shadow-4dp)}avl-menu-button[_ngcontent-%COMP%]   avl-icon-button[_ngcontent-%COMP%]{height:24px;width:24px;margin:0 16px 0 13px;padding:0}.mdc-toolbar__icon[_ngcontent-%COMP%]{color:var(--item-color);margin-bottom:0;padding:0}.mdc-toolbar__icon[_ngcontent-%COMP%]:active{color:var(--item-color)}.mdc-toolbar__icon[_ngcontent-%COMP%]:last-of-type{padding:0 16px}.search-bar-container[_ngcontent-%COMP%]{margin-left:24px;display:flex;justify-content:space-between;--avl-search-bar-text-color:var(--item-color, #fff);--avl-search-bar-search-color:var(--item-color, #fff)}.mdc-toolbar__row[_ngcontent-%COMP%]{min-height:var(--bar-height,53px);max-height:var(--bar-height,53px)}.mdc-toolbar__section[_ngcontent-%COMP%]{align-items:center;z-index:0}.avl-app-logo[_ngcontent-%COMP%]{width:32px;height:32px;background-size:100%}.burgerMenu[_ngcontent-%COMP%]{height:24px;width:24px}avl-icon-button[_ngcontent-%COMP%]{--avl-icon-button-color:white}.detailsMenuButtonSeparator[_ngcontent-%COMP%]{height:1px;background:#000}.icons-placeholder[_ngcontent-%COMP%]{font-size:20px;margin-bottom:0;margin-right:10px}.header-container[_ngcontent-%COMP%]{display:inline-flex;padding-bottom:5px;padding-top:5px}.header-icon[_ngcontent-%COMP%]{font-size:20px;margin-bottom:0;margin-right:10px;margin-top:5px}.header-text[_ngcontent-%COMP%]{display:inline;margin-left:-5px;font-size:25px;font-weight:700;padding-bottom:5px}.toolbar-item-with-switcher[_ngcontent-%COMP%]{padding-left:12px!important}avl-app-bar-item[_ngcontent-%COMP%]{cursor:initial!important}.section-with-switcher[_ngcontent-%COMP%]{position:relative;z-index:2}.burgerMenu[_ngcontent-%COMP%]{display:none;cursor:pointer}.avl-app-logo[_ngcontent-%COMP%]{margin:12px 0 12px 20px}.single-app-name[_ngcontent-%COMP%]{cursor:default}.mobile[_nghost-%COMP%]   .burgerMenu[_ngcontent-%COMP%]{display:block}@media (max-width:960px){.responsive[_nghost-%COMP%]   .burgerMenu[_ngcontent-%COMP%]{display:block}}@media (min-width:400px){.mobile[_nghost-%COMP%]   .burgerMenu[_ngcontent-%COMP%], .responsive[_nghost-%COMP%]   .burgerMenu[_ngcontent-%COMP%]{margin:16px 0 16px 24px}}@media (max-width:400px){.mobile[_nghost-%COMP%]   .avl-app-logo[_ngcontent-%COMP%], .responsive[_nghost-%COMP%]   .avl-app-logo[_ngcontent-%COMP%]{display:none}.mobile[_nghost-%COMP%]   .burgerMenu[_ngcontent-%COMP%], .responsive[_nghost-%COMP%]   .burgerMenu[_ngcontent-%COMP%]{margin:16px -4px 16px 16px}}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlAppBarComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-app-bar',
                template: "<header id=\"header\" class=\"mdc-toolbar mdc-toolbar--waterfall\" [ngClass]=\"{'mdc-toolbar--fixed': fixed}\">\r\n  <div class=\"mdc-toolbar__row\">\r\n    <section class=\"mdc-toolbar__section mdc-toolbar__section--align-start mdc-toolbar__section--shrink-to-fit section-with-switcher\">\r\n      <avl-app-bar-item class=\"burger-container\">\r\n        <avl-icon class=\"burgerMenu\" icon=\"menu\" theme=\"dark\" (click)=\"burgerIconClicked()\"></avl-icon>\r\n      </avl-app-bar-item>\r\n      <avl-app-bar-item class=\"noselect\" *ngIf=\"appLogoSet && appLogo != undefined\">\r\n        <a class=\"avl-app-logo\"\r\n          [style.background-image]=\"'url(' + appLogo + ')'\"\r\n          [style.background-repeat]=\"'no-repeat'\"\r\n          routerLink=\"{{homePageUrl ? homePageUrl : '/' }}\"></a>\r\n      </avl-app-bar-item>\r\n\r\n      <div *ngIf=\"isSingleApp; then singleApp else multiApp\"></div>\r\n      <ng-template #singleApp>\r\n        <avl-app-bar-item class=\"single-app-name\" [innerHTML]=\"currentApplication?.name\">\r\n        </avl-app-bar-item>\r\n      </ng-template>\r\n      <ng-template #multiApp>\r\n        <avl-app-bar-item class=\"toolbar-item-with-switcher\">\r\n          <avl-app-switcher [items]=\"applications\" (itemsChange)=\"onApplicationListChanged($event)\" [selectedValue]=\"currentApplication\" (selectedValueChange)=\"onApplicationChanged($event)\">\r\n          </avl-app-switcher>\r\n        </avl-app-bar-item>\r\n      </ng-template>\r\n    </section>\r\n    <section id=\"additionalContent\" class=\"mdc-toolbar__section mdc-toolbar__section--align-end \">\r\n      <ng-content></ng-content>\r\n      <ng-content select=\"avl-overflow-menu[slot=menu]\"></ng-content>\r\n    </section>\r\n  </div>\r\n</header>\r\n",
                encapsulation: core.ViewEncapsulation.Emulated,
                styles: [".mdc-toolbar{background-color:#3f51b5;background-color:var(--mdc-theme-primary,#3f51b5);color:#fff;color:var(--mdc-theme-text-primary-on-primary,#fff);display:flex;position:relative;flex-direction:column;justify-content:space-between;box-sizing:border-box;width:100%}.mdc-toolbar__row{display:flex;position:relative;align-items:center;box-sizing:border-box;width:100%;height:auto;min-height:64px}.mdc-toolbar__section{display:inline-flex;flex:1;align-items:flex-start;justify-content:center;min-width:0;height:100%}.mdc-toolbar__section--align-start{justify-content:flex-start;order:-1}.mdc-toolbar__section--align-end{justify-content:flex-end;order:1}.mdc-toolbar__title{font-family:Roboto,sans-serif;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;font-size:1.25rem;font-weight:500;letter-spacing:.02em;text-decoration:inherit;text-transform:inherit;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;margin-left:24px;margin-right:0;align-self:center;padding:16px 0;line-height:1.5rem;z-index:1}.mdc-toolbar__title[dir=rtl],[dir=rtl] .mdc-toolbar__title{margin-left:0;margin-right:24px}.mdc-toolbar__icon{display:flex;align-items:center;justify-content:center;border:none;background-color:inherit;text-decoration:none;color:#fff;color:var(--mdc-theme-text-primary-on-primary,#fff);cursor:pointer}.mdc-toolbar__icon:last-of-type[dir=rtl],[dir=rtl] .mdc-toolbar__icon:last-of-type{padding-left:24px;padding-right:12px}.mdc-toolbar__menu-icon{display:flex;align-items:center;justify-content:center;padding:16px 24px;border:none;background-color:inherit;text-decoration:none;color:#fff;color:var(--mdc-theme-text-primary-on-primary,#fff);cursor:pointer}.mdc-toolbar__menu-icon+.mdc-toolbar__title{margin-left:8px;margin-right:0}.mdc-toolbar__menu-icon+.mdc-toolbar__title[dir=rtl],[dir=rtl] .mdc-toolbar__menu-icon+.mdc-toolbar__title{margin-left:0;margin-right:8px}.mdc-toolbar--fixed{box-shadow:0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12);position:fixed;top:0;left:0;z-index:4;box-shadow:none}.mdc-toolbar--flexible{--mdc-toolbar-ratio-to-extend-flexible:4}.mdc-toolbar--flexible .mdc-toolbar__row:first-child{height:256px;height:calc(64px * var(--mdc-toolbar-ratio-to-extend-flexible,4))}.mdc-toolbar--flexible .mdc-toolbar__row:first-child::after{position:absolute;content:\"\"}.mdc-toolbar--flexible-default-behavior .mdc-toolbar__title{font-family:Roboto,sans-serif;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;font-size:2.125rem;font-weight:400;letter-spacing:normal;text-decoration:inherit;text-transform:inherit;align-self:flex-end;line-height:1.5rem}.mdc-toolbar--flexible-default-behavior .mdc-toolbar__row:first-child::after{top:0;left:0;width:100%;height:100%;transition:opacity .2s;opacity:1}.mdc-toolbar--flexible-default-behavior.mdc-toolbar--flexible-space-minimized .mdc-toolbar__row:first-child::after{opacity:0}.mdc-toolbar--flexible-default-behavior.mdc-toolbar--flexible-space-minimized .mdc-toolbar__title{font-weight:500}.mdc-toolbar--waterfall.mdc-toolbar--fixed{box-shadow:0 0 0 0 rgba(0,0,0,.2),0 0 0 0 rgba(0,0,0,.14),0 0 0 0 rgba(0,0,0,.12);transition:box-shadow 280ms cubic-bezier(.4,0,.2,1);will-change:box-shadow}.mdc-toolbar--waterfall.mdc-toolbar--fixed.mdc-toolbar--flexible-space-minimized{box-shadow:0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12)}.mdc-toolbar--waterfall.mdc-toolbar--fixed.mdc-toolbar--fixed-lastrow-only.mdc-toolbar--flexible-space-minimized{box-shadow:0 0 0 0 rgba(0,0,0,.2),0 0 0 0 rgba(0,0,0,.14),0 0 0 0 rgba(0,0,0,.12)}.mdc-toolbar--waterfall.mdc-toolbar--fixed.mdc-toolbar--fixed-lastrow-only.mdc-toolbar--fixed-at-last-row{box-shadow:0 2px 4px -1px rgba(0,0,0,.2),0 4px 5px 0 rgba(0,0,0,.14),0 1px 10px 0 rgba(0,0,0,.12)}.mdc-toolbar-fixed-adjust{margin-top:64px}@media (max-width:959px) and (orientation:landscape){.mdc-toolbar__row{min-height:48px}.mdc-toolbar-fixed-adjust{margin-top:48px}}@media (max-width:599px){.mdc-toolbar__row{min-height:56px}.mdc-toolbar__title{margin-left:16px;margin-right:0}.mdc-toolbar__title[dir=rtl],[dir=rtl] .mdc-toolbar__title{margin-left:0;margin-right:16px}.mdc-toolbar__icon{padding-right:8px;padding-left:8px}.mdc-toolbar__icon:last-of-type{padding-left:8px;padding-right:16px}.mdc-toolbar__icon:last-of-type[dir=rtl],[dir=rtl] .mdc-toolbar__icon:last-of-type{padding-left:16px;padding-right:8px}.mdc-toolbar__menu-icon{padding-right:16px;padding-left:16px}.mdc-toolbar__menu-icon+.mdc-toolbar__title{margin-left:16px;margin-right:0}.mdc-toolbar__menu-icon+.mdc-toolbar__title[dir=rtl],[dir=rtl] .mdc-toolbar__menu-icon+.mdc-toolbar__title{margin-left:0;margin-right:16px}.mdc-toolbar--flexible .mdc-toolbar__row:first-child{height:224px;height:calc(56px * var(--mdc-toolbar-ratio-to-extend-flexible,4))}.mdc-toolbar-fixed-adjust{margin-top:56px}}.mdc-toolbar__section--shrink-to-fit{flex:none}:host(){--item-color:var(--avl-app-bar-item-color, var(--avl-on-primary-color));--bar-height:var(--avl-app-bar-height, 53px);--background-color:var(--avl-app-bar-background-color, var(--avl-primary-color));--line-underline-color:var(--avl-app-bar-underline-color, var(--avl-product-color));color:var(--item-color);font-family:var(--avl-font-family);font-size:20px}#additionalContent{margin-right:12px;z-index:2}.material-icons{font-family:'Material Icons';font-weight:400;font-style:normal;font-size:24px;display:inline-block;line-height:1;text-transform:none;letter-spacing:normal;word-wrap:normal;white-space:nowrap;direction:ltr;-webkit-font-smoothing:antialiased;text-rendering:optimizeLegibility;-moz-osx-font-smoothing:grayscale;font-feature-settings:'liga'}.mdc-toolbar{background-color:var(--background-color);background-repeat:no-repeat;background-position:calc(100% - 61px) -30px;background-size:158px;color:var(--item-color);box-shadow:none;z-index:4;border-bottom:3px solid var(--line-underline-color)}#header{box-shadow:var(--avl-box-shadow-4dp)}avl-menu-button avl-icon-button{height:24px;width:24px;margin:0 16px 0 13px;padding:0}.mdc-toolbar__icon{color:var(--item-color);margin-bottom:0;padding:0}.mdc-toolbar__icon:active{color:var(--item-color)}.mdc-toolbar__icon:last-of-type{padding:0 16px}.search-bar-container{margin-left:24px;display:flex;justify-content:space-between;--avl-search-bar-text-color:var(--item-color, #fff);--avl-search-bar-search-color:var(--item-color, #fff)}.mdc-toolbar__row{min-height:var(--bar-height,53px);max-height:var(--bar-height,53px)}.mdc-toolbar__section{align-items:center;z-index:0}.avl-app-logo{width:32px;height:32px;background-size:100%}.burgerMenu{height:24px;width:24px}avl-icon-button{--avl-icon-button-color:white}.detailsMenuButtonSeparator{height:1px;background:#000}.icons-placeholder{font-size:20px;margin-bottom:0;margin-right:10px}.header-container{display:inline-flex;padding-bottom:5px;padding-top:5px}.header-icon{font-size:20px;margin-bottom:0;margin-right:10px;margin-top:5px}.header-text{display:inline;margin-left:-5px;font-size:25px;font-weight:700;padding-bottom:5px}.toolbar-item-with-switcher{padding-left:12px!important}avl-app-bar-item{cursor:initial!important}.section-with-switcher{position:relative;z-index:2}.burgerMenu{display:none;cursor:pointer}.avl-app-logo{margin:12px 0 12px 20px}.single-app-name{cursor:default}:host(.mobile) .burgerMenu{display:block}@media (max-width:960px){:host(.responsive) .burgerMenu{display:block}}@media (min-width:400px){:host(.mobile) .burgerMenu,:host(.responsive) .burgerMenu{margin:16px 0 16px 24px}}@media (max-width:400px){:host(.mobile) .avl-app-logo,:host(.responsive) .avl-app-logo{display:none}:host(.mobile) .burgerMenu,:host(.responsive) .burgerMenu{margin:16px -4px 16px 16px}}"]
            }]
    }], function () { return []; }, { applications: [{
            type: core.Input
        }], applicationListChange: [{
            type: core.Output
        }], fixed: [{
            type: core.Input
        }], currentApplicationChange: [{
            type: core.Output
        }], burgerClicked: [{
            type: core.Output
        }], currentApplication: [{
            type: core.Input
        }], appLogo: [{
            type: core.Input
        }], responsive: [{
            type: core.Input
        }, {
            type: core.HostBinding,
            args: ['class.responsive']
        }], mobile: [{
            type: core.Input
        }, {
            type: core.HostBinding,
            args: ['class.mobile']
        }], homePageUrl: [{
            type: core.Input
        }] }); })();
        return AvlAppBarComponent;
    }());

    var __decorate$1 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$1 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var AvlAppBarItemComponent = /** @class */ (function () {
        function AvlAppBarItemComponent(elementRef) {
            this.elementRef = elementRef;
        }
        /**
         * @hidden
         */
        AvlAppBarItemComponent.prototype.ngOnInit = function () {
            this.elementRef.nativeElement.classList.add('mdc-toolbar__icon');
        };
        AvlAppBarItemComponent.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        AvlAppBarItemComponent = __decorate$1([ __metadata$1("design:paramtypes", [core.ElementRef])
        ], AvlAppBarItemComponent);
AvlAppBarItemComponent.ɵfac = function AvlAppBarItemComponent_Factory(t) { return new (t || AvlAppBarItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlAppBarItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlAppBarItemComponent, selectors: [["avl-app-bar-item"]], ngContentSelectors: _c3, decls: 1, vars: 0, template: function AvlAppBarItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlAppBarItemComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-app-bar-item',
                template: '<ng-content></ng-content>'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
        return AvlAppBarItemComponent;
    }());

    var __decorate$2 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$2 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var rowHeightPx = 48;
    var AvlSideNavComponent = /** @class */ (function () {
        function AvlSideNavComponent(router$1, activeRoute) {
            var _this = this;
            this.router = router$1;
            this.activeRoute = activeRoute;
            this.openValue = false;
            /**
             * Emitted when the open property changes.
             */
            this.openChange = new core.EventEmitter();
            activeRoute.queryParams.subscribe(function (p) {
                _this.params = p;
            });
            this.subscription = this.router.events
                .pipe(operators.filter(function (event) { return event instanceof router.ActivationEnd; }))
                .subscribe(function (event) {
                _this.selectItems();
            });
        }
        Object.defineProperty(AvlSideNavComponent.prototype, "open", {
            /**
             * Can be set to open or close the side nav.
             * Returns the current state of the side nav.
             */
            get: function () {
                return this.openValue;
            },
            set: function (val) {
                this.openValue = val;
                this.openChange.next(val);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @hidden
         */
        AvlSideNavComponent.prototype.toggleOpen = function () {
            if (!this.manuallyOpened) {
                this.open = !this.open;
            }
        };
        /**
         * @hidden
         */
        AvlSideNavComponent.prototype.arrowToggleOpen = function () {
            this.manuallyOpened = !this.open;
            this.open = !this.open;
        };
        /**
         * @hidden
         */
        AvlSideNavComponent.prototype.mobileClose = function () {
            if (Math.max(document.documentElement.clientWidth, window.innerWidth) < 960) {
                this.open = false;
            }
        };
        /**
         * @hidden
         */
        AvlSideNavComponent.prototype.groupClicked = function (item) {
            if (!this.open) {
                this.open = true;
                if (!item.expanded) {
                    item.expanded = true;
                }
            }
            else {
                item.expanded = !item.expanded;
            }
        };
        /**
         * @hidden
         */
        AvlSideNavComponent.prototype.getScrubberStyle = function () {
            return {
                height: this.selectedParagraph ? '14px' : '0px',
                top: this.selectedParagraphIndex * rowHeightPx + "px"
            };
        };
        /**
        * @hidden
        */
        AvlSideNavComponent.prototype.trimSlash = function (str) {
            if (str[0] === '/') {
                str = str.substring(1, str.length);
            }
            if (str[str.length - 1] === '/') {
                str = str.substring(0, str.length - 1);
            }
            return str;
        };
        AvlSideNavComponent.prototype.selectItems = function () {
            if (!this.menuItems) {
                return;
            }
            var fragment = this.router.parseUrl(this.router.url).fragment;
            var url = this.router.url.split('#')[0];
            var queue = this.menuItems.slice();
            var currentGroup;
            this.selectedGroup = undefined;
            this.selectedPage = undefined;
            this.selectedParagraph = undefined;
            var fallbackSelectedPage;
            var _loop_1 = function () {
                var item = queue.pop();
                if (isGroup(item)) {
                    currentGroup = item;
                    item.subItems.map(function (subitem) { queue.push(subitem); });
                }
                else {
                    if (fallbackSelectedPage == null && this_1.trimSlash(url).startsWith(this_1.trimSlash(item.route))) {
                        fallbackSelectedPage = item;
                    }
                    if (item.route && this_1.trimSlash(url).endsWith(this_1.trimSlash(item.route + this_1.getUrlParams(item.params)))) {
                        if (this_1.selectedPage && this_1.selectedPage.route && this_1.selectedPage.route.length > item.route.length) {
                            return "continue";
                        }
                        if (item.params && item.params.length > 0) {
                            var object_1 = new Object();
                            item.params.forEach(function (i) {
                                object_1[i.name] = i.value;
                            });
                            if (JSON.stringify(this_1.params) === JSON.stringify(object_1)) {
                                this_1.selectedPage = item;
                            }
                        }
                        else {
                            this_1.selectedPage = item;
                        }
                        if (isTopLevelPage(item)) {
                            this_1.selectedGroup = undefined;
                        }
                        else {
                            this_1.selectedGroup = currentGroup;
                        }
                        if (item.paragraphs && item.paragraphs.length > 0) {
                            var paragraphIndex = item.paragraphs.findIndex(function (p) { return p.id === fragment; });
                            if (fragment && item.paragraphs && paragraphIndex !== -1) {
                                this_1.selectedParagraph = fragment;
                                this_1.selectedParagraphIndex = paragraphIndex;
                            }
                        }
                    }
                }
            };
            var this_1 = this;
            while (queue.length) {
                _loop_1();
            }
            if (this.selectedPage == null) {
                this.selectedPage = fallbackSelectedPage;
            }
        };
        /**
        * @hidden
        */
        AvlSideNavComponent.prototype.ngOnChanges = function (changes) {
            if (changes['menuItems']) {
                this.selectItems();
            }
        };
        /**
        * @hidden
        */
        AvlSideNavComponent.prototype.ngOnDestroy = function () {
            this.subscription.unsubscribe();
        };
        /**
        * @hidden
        */
        AvlSideNavComponent.prototype.getParams = function (params) {
            if (params && params.length > 0) {
                var object_2 = {};
                params.forEach(function (param) {
                    object_2[param.name] = param.value;
                });
                return object_2;
            }
            else {
                return null;
            }
        };
        /**
        * @hidden
        */
        AvlSideNavComponent.prototype.getUrlParams = function (params) {
            if (params && params.length > 0) {
                var url_1 = '?';
                params.forEach(function (param) {
                    url_1 += param.name + '=' + param.value + '&';
                });
                return url_1.slice(0, -1);
            }
            else {
                return '';
            }
        };
        AvlSideNavComponent.ctorParameters = function () { return [
            { type: router.Router },
            { type: router.ActivatedRoute }
        ]; };
        __decorate$2([
            core.Input(),
            core.HostBinding('class.open'),
            __metadata$2("design:type", Boolean),
            __metadata$2("design:paramtypes", [Boolean])
        ], AvlSideNavComponent.prototype, "open", null);
        __decorate$2([
            core.Input(),
            __metadata$2("design:type", Array)
        ], AvlSideNavComponent.prototype, "menuItems", void 0);
        __decorate$2([
            core.Input(),
            core.HostBinding('class.responsive'),
            __metadata$2("design:type", Boolean)
        ], AvlSideNavComponent.prototype, "responsive", void 0);
        __decorate$2([
            core.Input(),
            core.HostBinding('class.mobile'),
            __metadata$2("design:type", Boolean)
        ], AvlSideNavComponent.prototype, "mobile", void 0);
        __decorate$2([
            core.Output(),
            __metadata$2("design:type", core.EventEmitter)
        ], AvlSideNavComponent.prototype, "openChange", void 0);
        AvlSideNavComponent = __decorate$2([ __metadata$2("design:paramtypes", [router.Router,
                router.ActivatedRoute])
        ], AvlSideNavComponent);
AvlSideNavComponent.ɵfac = function AvlSideNavComponent_Factory(t) { return new (t || AvlSideNavComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ActivatedRoute)); };
AvlSideNavComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlSideNavComponent, selectors: [["avl-side-nav"]], hostVars: 6, hostBindings: function AvlSideNavComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("open", ctx.open)("responsive", ctx.responsive)("mobile", ctx.mobile);
    } }, inputs: { open: "open", menuItems: "menuItems", responsive: "responsive", mobile: "mobile" }, outputs: { openChange: "openChange" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 11, vars: 5, consts: [[1, "outer-container", "toolbar-spacing"], ["outerContainer", ""], [1, "item-scroll"], [1, "inner-container"], ["innerContainer", ""], ["ngFor", "", 3, "ngForOf"], [1, "arrow-spacer"], ["id", "Layer_1", "data-name", "Layer 1", "xmlns", "http://www.w3.org/2000/svg", "width", "24", "height", "24", "viewBox", "0 0 24 24", 1, "open-close-arrow", 3, "click"], ["transform", "translate(405 102) rotate(-180)"], ["d", "M-972.955,492l6.774-6,1.581,1.4-5.194,4.6,5.194,4.6-1.581,1.4Zm-5.645,0,6.773-6,1.581,1.4-5.193,4.6,5.193,4.6-1.581,1.4Z", "transform", "translate(1364.6 -402)"], [1, "scrim", 3, "click"], ["tabindex", "0", "class", "item top-level-page", 3, "routerLink", "queryParams", "selected", "click", 4, "ngIf"], ["class", "child-container", 4, "ngIf"], ["tabindex", "0", "class", "item group", 3, "expanded", "selected", "click", 4, "ngIf"], ["class", "child-container", 3, "click", 4, "ngIf"], ["tabindex", "0", 1, "item", "top-level-page", 3, "routerLink", "queryParams", "click"], [1, "image-div"], [4, "ngIf"], [3, "src", 4, "ngIf"], [1, "label-div"], [1, "item-overlay"], [3, "src"], [1, "child-container"], ["class", "item top-level-paragraph", 3, "href", "selected", "click", 4, "ngFor", "ngForOf"], [1, "page-progress-bar"], [1, "page-progress-bar-background"], [1, "page-progress-bar-scrubber", 3, "ngStyle"], [1, "item", "top-level-paragraph", 3, "href", "click"], ["tabindex", "0", 1, "item", "group", 3, "click"], [1, "label-div", "group"], ["icon", "arrow-drop-down", 1, "group-arrow"], [1, "child-container", 3, "click"], [1, "item", "subpage", 3, "routerLink", "queryParams", "click"], ["class", "item subpage-paragraph", 3, "href", "selected", "click", 4, "ngFor", "ngForOf"], [1, "item", "subpage-paragraph", 3, "href", "click"]], template: function AvlSideNavComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3, 4);
        ɵngcc0.ɵɵtemplate(5, AvlSideNavComponent_ng_template_5_Template, 4, 4, "ng-template", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(6, "div", 6);
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(7, "svg", 7);
        ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_Template__svg_svg_click_7_listener() { return ctx.arrowToggleOpen(); });
        ɵngcc0.ɵɵelementStart(8, "g", 8);
        ɵngcc0.ɵɵelement(9, "path", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵnamespaceHTML();
        ɵngcc0.ɵɵelementStart(10, "div", 10);
        ɵngcc0.ɵɵlistener("click", function AvlSideNavComponent_Template_div_click_10_listener() { return ctx.open = false; });
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("closed", !ctx.open);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.menuItems);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵclassProp("closed", !ctx.open);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc2.RouterLinkWithHref, ɵngcc1.NgStyle], styles: ["[_nghost-%COMP%]{font-family:var(--avl-font-family);height:100%;display:block;--text-color:var(--avl-side-nav-text-color, var(--avl-emphasis-high-color));--icons-color:var(--avl-side-nav-icons-color, var(--avl-emphasis-medium-color));--background-color:var(--avl-side-nav-background-color, var(--avl-side-nav-color));--item-icon-width:72px;--side-nav-mini-width:var(--item-icon-width);--side-nav-open-width:var(--avl-side-nav-open-width, 256px);--item-height:48px;box-shadow:var(--avl-box-shadow-12dp);z-index:3}.outer-container[_ngcontent-%COMP%]{position:relative;box-sizing:content-box;height:calc(100% - var(--avl-app-bar-height,56px));background-color:var(--background-color);width:var(--side-nav-open-width);overflow:hidden}.inner-container[_ngcontent-%COMP%]{width:var(--side-nav-open-width)}.toolbar-spacing[_ngcontent-%COMP%]{padding-top:var(--avl-app-bar-height,56px)}.open-close-arrow[_ngcontent-%COMP%]{position:absolute;bottom:12px;right:24px;font-size:24px;fill:var(--icons-color);cursor:pointer;-webkit-user-select:none;-moz-user-select:none;user-select:none;width:24px}.open-close-arrow[_ngcontent-%COMP%]:not(.closed){transform:scale(-1,1)}.outer-container.closed[_ngcontent-%COMP%]{width:var(--side-nav-mini-width);-webkit-animation-name:closing;animation-name:closing;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:195ms}.arrow-spacer[_ngcontent-%COMP%]{height:48px}.item-scroll[_ngcontent-%COMP%]{height:calc(100% - 48px);overflow-y:auto;overflow-x:hidden}.item[_ngcontent-%COMP%]{display:flex;align-items:center;cursor:pointer;color:var(--text-color);height:var(--item-height);position:relative;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;isolation:isolate;outline:0;white-space:pre-wrap;padding-right:24px}.item[_ngcontent-%COMP%]:active{color:var(--text-color)}.top-level-paragraph[_ngcontent-%COMP%]{padding-left:88px}.subpage[_ngcontent-%COMP%]{padding-left:72px}.subpage-paragraph[_ngcontent-%COMP%]{padding-left:88px}.label-div[_ngcontent-%COMP%]{z-index:10;font-family:Roboto;max-width:160px;overflow:hidden;text-overflow:ellipsis;max-height:40px;word-break:break-word;-webkit-box-orient:vertical;display:-webkit-box;-webkit-line-clamp:2}.group.expanded[_ngcontent-%COMP%] > .label-div.group[_ngcontent-%COMP%]{font-family:Roboto-Medium}.label-div.group[_ngcontent-%COMP%]{max-width:calc(100% - 112px)}.image-div[_ngcontent-%COMP%]{color:var(--avl-emphasis-medium-color);display:flex;align-items:center;justify-content:center;width:var(--item-icon-width);height:var(--item-height);margin-right:0;z-index:9}img[_ngcontent-%COMP%]{height:auto;width:auto;max-height:24px;max-width:24px}avl-icon[_ngcontent-%COMP%]{margin-bottom:0;font-size:24px;color:var(--avl-emphasis-medium-color)}.group-arrow[_ngcontent-%COMP%]{position:absolute;right:24px}.group.expanded[_ngcontent-%COMP%] > .group-arrow[_ngcontent-%COMP%]{transform:rotate(180deg)}.item-overlay[_ngcontent-%COMP%]{box-sizing:border-box;position:absolute;height:100%;width:100%;left:0;z-index:0}.item.selected[_ngcontent-%COMP%] > .item-overlay[_ngcontent-%COMP%]{background-color:var(--avl-side-nav-product-color,var(--avl-product-focus-color))}.item[_ngcontent-%COMP%]:hover:not(.selected) > .item-overlay[_ngcontent-%COMP%]{background-color:var(--avl-side-nav-product-color-hover,var(--avl-product-hover-color))}.item.selected[_ngcontent-%COMP%]{font-weight:500}.child-container[_ngcontent-%COMP%]{position:relative;display:flex;overflow:hidden;flex-direction:column-reverse}.page-progress-bar[_ngcontent-%COMP%]{position:absolute;left:72px;top:11px;bottom:11px;width:2px;overflow:hidden}.page-progress-bar-background[_ngcontent-%COMP%]{background-color:var(--text-color);opacity:.2;width:100%;height:100%}.page-progress-bar-scrubber[_ngcontent-%COMP%]{position:absolute;width:100%;background-color:var(--text-color);opacity:1;transition:height .2s,top .2s;margin-top:6px;margin-bottom:2px}@-webkit-keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@-webkit-keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}@keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}.outer-container[_ngcontent-%COMP%]:not(.closed){-webkit-animation-name:opening;animation-name:opening;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:225ms}@media (max-width:960px){.responsive[_nghost-%COMP%]{position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}.responsive[slot=left-side][_nghost-%COMP%]{left:0}.responsive[slot=right-side][_nghost-%COMP%]{right:0}.responsive[_nghost-%COMP%]:not(.open){box-shadow:none}.responsive[_nghost-%COMP%]   .item-scroll[_ngcontent-%COMP%]{height:100%}.responsive[_nghost-%COMP%]   .open-close-arrow[_ngcontent-%COMP%]{display:none}.responsive[_nghost-%COMP%]   .arrow-spacer[_ngcontent-%COMP%]{display:none;visibility:hidden}.responsive[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{transition:250ms;transition-duration:250px}.responsive[slot=left-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;left:0}.responsive.open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}.responsive[slot=left-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{left:var(--side-nav-open-width)}.responsive[slot=right-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{right:var(--side-nav-open-width)}.responsive[slot=right-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;right:0}}.mobile[_nghost-%COMP%]{position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}.mobile[slot=left-side][_nghost-%COMP%]{left:0}.mobile[slot=right-side][_nghost-%COMP%]{right:0}.mobile[_nghost-%COMP%]:not(.open){box-shadow:none}.mobile[_nghost-%COMP%]   .item-scroll[_ngcontent-%COMP%]{height:100%}.mobile[_nghost-%COMP%]   .open-close-arrow[_ngcontent-%COMP%]{display:none}.mobile[_nghost-%COMP%]   .arrow-spacer[_ngcontent-%COMP%]{display:none;visibility:hidden}.mobile[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{transition:250ms;transition-duration:250px}.mobile[slot=left-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;left:0}.mobile.open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}.mobile[slot=left-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{left:var(--side-nav-open-width)}.mobile[slot=right-side].open[_nghost-%COMP%]   .scrim[_ngcontent-%COMP%]{right:var(--side-nav-open-width)}.mobile[slot=right-side][_nghost-%COMP%]:not(.open)   .scrim[_ngcontent-%COMP%]{visibility:hidden;right:0}"], data: { animation: [
            animations.trigger('expandCollapse', [
                animations.transition(':enter', [
                    animations.style({ 'max-height': '0px' }),
                    animations.animate('225ms ease-in', animations.style({ transform: 'translateY(0%)', 'max-height': '200px' }))
                ]),
                animations.transition(':leave', [
                    animations.style({ 'max-height': '200px' }),
                    animations.animate('180ms ease-in', animations.style({ 'max-height': '0px' }))
                ])
            ])
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSideNavComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-side-nav',
                template: "<div #outerContainer [class.closed]=\"!open\" class=\"outer-container toolbar-spacing\">\r\n  <div class=\"item-scroll\">\r\n    <div #innerContainer class=\"inner-container\">\r\n      <ng-template ngFor let-item [ngForOf]=\"menuItems\">\r\n        <!-- Top level page -->\r\n        <a *ngIf=\"!item.subItems\" tabindex=\"0\" class=\"item top-level-page\" routerLink={{item.route}} [queryParams]=\"getParams(item.params)\" [class.selected]=\"item === selectedPage\" (click)=\"mobileClose()\">\r\n          <div class=\"image-div\">\r\n            <avl-icon *ngIf=\"item.icon\" attr.icon=\"{{item.icon}}\"></avl-icon>\r\n            <img *ngIf=\"item.iconUrl\" [src]=\"item.iconUrl\">\r\n          </div>\r\n          <div class=\"label-div\">{{item.label}}</div>\r\n          <div class=\"item-overlay\"></div>\r\n        </a>\r\n        <!-- Paragraphs of a top level page -->\r\n        <div class=\"child-container\" *ngIf=\"open && item.paragraphs && item.paragraphs.length > 0 && item === selectedPage\" [@expandCollapse]>\r\n          <!-- div for preventing reversing of the items by flex-direction: column-reverse -->\r\n          <div>\r\n            <a *ngFor=\"let paragraph of item.paragraphs\" class=\"item top-level-paragraph\"\r\n              href=\"{{trimSlash(item.route + getUrlParams(item.params) + '#' + paragraph.id)}}\" [class.selected]=\"paragraph.id === selectedParagraph\" (click)=\"mobileClose()\">\r\n              <div class=\"label-div\">{{paragraph.label}}</div>\r\n            </a>\r\n          </div>\r\n          <div class=\"page-progress-bar\">\r\n            <div class=\"page-progress-bar-background\"></div>\r\n            <div class=\"page-progress-bar-scrubber\" [ngStyle]=\"getScrubberStyle()\"></div>\r\n          </div>\r\n        </div>\r\n        <!-- Group -->\r\n        <div *ngIf=\"item.subItems\" tabindex=\"0\" class=\"item group\" [class.expanded]=\"item.expanded\" (click)=\"groupClicked(item)\"\r\n          [class.selected]=\"!open && item === selectedGroup\">\r\n          <div class=\"image-div\">\r\n            <avl-icon *ngIf=\"item.icon\" attr.icon=\"{{item.icon}}\"></avl-icon>\r\n            <img *ngIf=\"item.iconUrl\" [src]=\"item.iconUrl\">\r\n          </div>\r\n          <div class=\"label-div group\">{{item.label}}</div>\r\n          <avl-icon class=\"group-arrow\" icon=\"arrow-drop-down\"></avl-icon>\r\n          <div class=\"item-overlay\"></div>\r\n        </div>\r\n        <!-- Subpages -->\r\n        <div class=\"child-container\" *ngIf=\"open && item.subItems && item.subItems.length > 0 && item.expanded\" [@expandCollapse] (click)=\"toggleOpen()\">\r\n          <!-- div for preventing reversing of the items by flex-direction: column-reverse -->\r\n          <div>\r\n            <ng-template ngFor let-subItem [ngForOf]=\"item.subItems\">\r\n              <a class=\"item subpage\" routerLink={{subItem.route}} [queryParams]=\"getParams(subItem.params)\" [class.selected]=\"subItem === selectedPage\" (click)=\"mobileClose()\">\r\n                <div class=\"label-div\">{{subItem.label}}</div>\r\n                <div class=\"item-overlay\"></div>\r\n              </a>\r\n              <!-- Paragraphs of a subpage -->\r\n              <div class=\"child-container\" *ngIf=\"subItem.paragraphs && subItem.paragraphs.length > 0 && subItem === selectedPage\" [@expandCollapse]>\r\n                <!-- div for preventing reversing of the items by flex-direction: column-reverse -->\r\n                <div>\r\n                  <a *ngFor=\"let paragraph of subItem.paragraphs\" class=\"item subpage-paragraph\" href=\"{{subItem.route}}{{getUrlParams(item.params)}}#{{paragraph.id}}\"\r\n                    [class.selected]=\"paragraph.id === selectedParagraph\" (click)=\"mobileClose()\">\r\n                    <div class=\"label-div\">{{paragraph.label}}</div>\r\n                  </a>\r\n                </div>\r\n                <div class=\"page-progress-bar\">\r\n                  <div class=\"page-progress-bar-background\"></div>\r\n                  <div class=\"page-progress-bar-scrubber\" [ngStyle]=\"getScrubberStyle()\"></div>\r\n                </div>\r\n              </div>\r\n            </ng-template>\r\n          </div>\r\n        </div>\r\n      </ng-template>\r\n    </div>\r\n  </div>\r\n  <div class=\"arrow-spacer\"></div>\r\n\r\n  <svg (click)=\"arrowToggleOpen()\" [class.closed]=\"!open\" class=\"open-close-arrow\"  id=\"Layer_1\" data-name=\"Layer 1\" xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\" viewBox=\"0 0 24 24\">\r\n    <g transform=\"translate(405 102) rotate(-180)\">\r\n      <path d=\"M-972.955,492l6.774-6,1.581,1.4-5.194,4.6,5.194,4.6-1.581,1.4Zm-5.645,0,6.773-6,1.581,1.4-5.193,4.6,5.193,4.6-1.581,1.4Z\" transform=\"translate(1364.6 -402)\"/>\r\n    </g>\r\n  </svg>\r\n\r\n</div>\r\n<div class=\"scrim\" (click)=\"open = false;\"></div>",
                animations: [
                    animations.trigger('expandCollapse', [
                        animations.transition(':enter', [
                            animations.style({ 'max-height': '0px' }),
                            animations.animate('225ms ease-in', animations.style({ transform: 'translateY(0%)', 'max-height': '200px' }))
                        ]),
                        animations.transition(':leave', [
                            animations.style({ 'max-height': '200px' }),
                            animations.animate('180ms ease-in', animations.style({ 'max-height': '0px' }))
                        ])
                    ])
                ],
                styles: [":host{font-family:var(--avl-font-family);height:100%;display:block;--text-color:var(--avl-side-nav-text-color, var(--avl-emphasis-high-color));--icons-color:var(--avl-side-nav-icons-color, var(--avl-emphasis-medium-color));--background-color:var(--avl-side-nav-background-color, var(--avl-side-nav-color));--item-icon-width:72px;--side-nav-mini-width:var(--item-icon-width);--side-nav-open-width:var(--avl-side-nav-open-width, 256px);--item-height:48px;box-shadow:var(--avl-box-shadow-12dp);z-index:3}.outer-container{position:relative;box-sizing:content-box;height:calc(100% - var(--avl-app-bar-height,56px));background-color:var(--background-color);width:var(--side-nav-open-width);overflow:hidden}.inner-container{width:var(--side-nav-open-width)}.toolbar-spacing{padding-top:var(--avl-app-bar-height,56px)}.open-close-arrow{position:absolute;bottom:12px;right:24px;font-size:24px;fill:var(--icons-color);cursor:pointer;-webkit-user-select:none;-moz-user-select:none;user-select:none;width:24px}.open-close-arrow:not(.closed){transform:scale(-1,1)}.outer-container.closed{width:var(--side-nav-mini-width);-webkit-animation-name:closing;animation-name:closing;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:195ms}.arrow-spacer{height:48px}.item-scroll{height:calc(100% - 48px);overflow-y:auto;overflow-x:hidden}.item{display:flex;align-items:center;cursor:pointer;color:var(--text-color);height:var(--item-height);position:relative;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;isolation:isolate;outline:0;white-space:pre-wrap;padding-right:24px}.item:active{color:var(--text-color)}.top-level-paragraph{padding-left:88px}.subpage{padding-left:72px}.subpage-paragraph{padding-left:88px}.label-div{z-index:10;font-family:Roboto;max-width:160px;overflow:hidden;text-overflow:ellipsis;max-height:40px;word-break:break-word;-webkit-box-orient:vertical;display:-webkit-box;-webkit-line-clamp:2}.group.expanded>.label-div.group{font-family:Roboto-Medium}.label-div.group{max-width:calc(100% - 112px)}.image-div{color:var(--avl-emphasis-medium-color);display:flex;align-items:center;justify-content:center;width:var(--item-icon-width);height:var(--item-height);margin-right:0;z-index:9}img{height:auto;width:auto;max-height:24px;max-width:24px}avl-icon{margin-bottom:0;font-size:24px;color:var(--avl-emphasis-medium-color)}.group-arrow{position:absolute;right:24px}.group.expanded>.group-arrow{transform:rotate(180deg)}.item-overlay{box-sizing:border-box;position:absolute;height:100%;width:100%;left:0;z-index:0}.item.selected>.item-overlay{background-color:var(--avl-side-nav-product-color,var(--avl-product-focus-color))}.item:hover:not(.selected)>.item-overlay{background-color:var(--avl-side-nav-product-color-hover,var(--avl-product-hover-color))}.item.selected{font-weight:500}.child-container{position:relative;display:flex;overflow:hidden;flex-direction:column-reverse}.page-progress-bar{position:absolute;left:72px;top:11px;bottom:11px;width:2px;overflow:hidden}.page-progress-bar-background{background-color:var(--text-color);opacity:.2;width:100%;height:100%}.page-progress-bar-scrubber{position:absolute;width:100%;background-color:var(--text-color);opacity:1;transition:height .2s,top .2s;margin-top:6px;margin-bottom:2px}@-webkit-keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@keyframes closing{from{width:var(--side-nav-open-width)}to{width:var(--side-nav-mini-width)}}@-webkit-keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}@keyframes opening{from{width:var(--side-nav-mini-width)}to{width:var(--side-nav-open-width)}}.outer-container:not(.closed){-webkit-animation-name:opening;animation-name:opening;-webkit-animation-timing-function:cubic-bezier(0,0,.2,1);animation-timing-function:cubic-bezier(0,0,.2,1);transition-duration:225ms}@media (max-width:960px){:host(.responsive){position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}:host(.responsive[slot=left-side]){left:0}:host(.responsive[slot=right-side]){right:0}:host(.responsive:not(.open)){box-shadow:none}:host(.responsive) .item-scroll{height:100%}:host(.responsive) .open-close-arrow{display:none}:host(.responsive) .arrow-spacer{display:none;visibility:hidden}:host(.responsive) .scrim{transition:250ms;transition-duration:250px}:host(.responsive[slot=left-side]:not(.open)) .scrim{visibility:hidden;left:0}:host(.responsive.open) .scrim{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}:host(.responsive[slot=left-side].open) .scrim{left:var(--side-nav-open-width)}:host(.responsive[slot=right-side].open) .scrim{right:var(--side-nav-open-width)}:host(.responsive[slot=right-side]:not(.open)) .scrim{visibility:hidden;right:0}}:host(.mobile){position:absolute;--side-nav-mini-width:0;transition:box-shadow 250ms}:host(.mobile[slot=left-side]){left:0}:host(.mobile[slot=right-side]){right:0}:host(.mobile:not(.open)){box-shadow:none}:host(.mobile) .item-scroll{height:100%}:host(.mobile) .open-close-arrow{display:none}:host(.mobile) .arrow-spacer{display:none;visibility:hidden}:host(.mobile) .scrim{transition:250ms;transition-duration:250px}:host(.mobile[slot=left-side]:not(.open)) .scrim{visibility:hidden;left:0}:host(.mobile.open) .scrim{background-color:var(--avl-scrim-color);position:fixed;visibility:visible;top:0;width:100%;height:100%}:host(.mobile[slot=left-side].open) .scrim{left:var(--side-nav-open-width)}:host(.mobile[slot=right-side].open) .scrim{right:var(--side-nav-open-width)}:host(.mobile[slot=right-side]:not(.open)) .scrim{visibility:hidden;right:0}"]
            }]
    }], function () { return [{ type: ɵngcc2.Router }, { type: ɵngcc2.ActivatedRoute }]; }, { openChange: [{
            type: core.Output
        }], open: [{
            type: core.Input
        }, {
            type: core.HostBinding,
            args: ['class.open']
        }], menuItems: [{
            type: core.Input
        }], responsive: [{
            type: core.Input
        }, {
            type: core.HostBinding,
            args: ['class.responsive']
        }], mobile: [{
            type: core.Input
        }, {
            type: core.HostBinding,
            args: ['class.mobile']
        }] }); })();
        return AvlSideNavComponent;
    }());
    /**
    * @hidden
    */
    function isGroup(item) {
        return item.subItems !== undefined && item.subItems.length > 0;
    }
    /**
    * @hidden
    */
    function isTopLevelPage(item) {
        return !isGroup(item) && (item.icon !== undefined || item.iconUrl !== undefined);
    }

    var __decorate$3 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$3 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var AvlAppLayoutComponent = /** @class */ (function () {
        function AvlAppLayoutComponent() {
        }
        AvlAppLayoutComponent.prototype.ngAfterContentInit = function () {
            var _this = this;
            if (this.sideNav && this.appBar) {
                this.appBar.burgerClicked.subscribe(function () { return _this.sideNav.open = !_this.sideNav.open; });
            }
        };
        __decorate$3([
            core.ContentChild(AvlSideNavComponent),
            __metadata$3("design:type", AvlSideNavComponent)
        ], AvlAppLayoutComponent.prototype, "sideNav", void 0);
        __decorate$3([
            core.ContentChild(AvlAppBarComponent),
            __metadata$3("design:type", AvlAppBarComponent)
        ], AvlAppLayoutComponent.prototype, "appBar", void 0);
        AvlAppLayoutComponent = __decorate$3([ __metadata$3("design:paramtypes", [])
        ], AvlAppLayoutComponent);
AvlAppLayoutComponent.ɵfac = function AvlAppLayoutComponent_Factory(t) { return new (t || AvlAppLayoutComponent)(); };
AvlAppLayoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlAppLayoutComponent, selectors: [["avl-app-layout"]], contentQueries: function AvlAppLayoutComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, AvlSideNavComponent, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, AvlAppBarComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.sideNav = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.appBar = _t.first);
    } }, ngContentSelectors: _c5, decls: 6, vars: 0, consts: [[1, "sideNavAndContent"]], template: function AvlAppLayoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c4);
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵprojection(2, 1);
        ɵngcc0.ɵɵelementStart(3, "main");
        ɵngcc0.ɵɵprojection(4, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(5, 3);
        ɵngcc0.ɵɵelementEnd();
    } }, styles: [".sideNavAndContent[_ngcontent-%COMP%]{display:flex;flex-direction:row;flex-flow:row nowrap;height:100%;padding-bottom:4px;position:relative}.scrim[_ngcontent-%COMP%]{display:none;visibility:hidden}main[_ngcontent-%COMP%]{--app-bar-height:var(--avl-app-bar-height, 53px);color:var(--avl-primary-text-color);width:100%;padding:var(--avl-app-layout-content-padding,24px);overflow:auto;margin-top:var(--app-bar-height)}[noPadding][_nghost-%COMP%]   main[_ngcontent-%COMP%]{padding:0}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlAppLayoutComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-app-layout',
                template: "<ng-content select=\"[slot=toolbar]\"></ng-content>\r\n<div class=\"sideNavAndContent\">\r\n  <ng-content select=\"[slot=left-side]\"></ng-content>\r\n  <main>\r\n    <ng-content></ng-content>\r\n  </main>\r\n  <ng-content select=\"[slot=right-side]\"></ng-content>\r\n</div>\r\n",
                encapsulation: core.ViewEncapsulation.Emulated,
                styles: [".sideNavAndContent{display:flex;flex-direction:row;flex-flow:row nowrap;height:100%;padding-bottom:4px;position:relative}.scrim{display:none;visibility:hidden}main{--app-bar-height:var(--avl-app-bar-height, 53px);color:var(--avl-primary-text-color);width:100%;padding:var(--avl-app-layout-content-padding,24px);overflow:auto;margin-top:var(--app-bar-height)}:host([noPadding]) main{padding:0}"]
            }]
    }], function () { return []; }, { sideNav: [{
            type: core.ContentChild,
            args: [AvlSideNavComponent]
        }], appBar: [{
            type: core.ContentChild,
            args: [AvlAppBarComponent]
        }] }); })();
        return AvlAppLayoutComponent;
    }());

    var __decorate$4 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$4 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var AvlNoPaddingDirective = /** @class */ (function () {
        function AvlNoPaddingDirective(elementRef) {
            this.elementRef = elementRef;
        }
        /** @hidden */
        AvlNoPaddingDirective.prototype.ngOnInit = function () {
            var appLayout = this.findAppLayout();
            if (appLayout) {
                appLayout.setAttribute('noPadding', '');
            }
        };
        /** @hidden */
        AvlNoPaddingDirective.prototype.ngOnDestroy = function () {
            var appLayout = this.findAppLayout();
            if (appLayout) {
                appLayout.removeAttribute('noPadding');
            }
        };
        AvlNoPaddingDirective.prototype.findAppLayout = function () {
            var element = this.elementRef.nativeElement;
            while (element) {
                if (element.tagName.toLowerCase() == 'avl-app-layout') {
                    return element;
                }
                element = element.parentElement;
            }
        };
        AvlNoPaddingDirective.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        AvlNoPaddingDirective = __decorate$4([ __metadata$4("design:paramtypes", [core.ElementRef])
        ], AvlNoPaddingDirective);
AvlNoPaddingDirective.ɵfac = function AvlNoPaddingDirective_Factory(t) { return new (t || AvlNoPaddingDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlNoPaddingDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlNoPaddingDirective, selectors: [["", "avl-no-padding", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNoPaddingDirective, [{
        type: core.Directive,
        args: [{
                selector: '[avl-no-padding]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
        return AvlNoPaddingDirective;
    }());

    var __decorate$5 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$5 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var AvlParagraphDirective = /** @class */ (function () {
        function AvlParagraphDirective(elementRef) {
            this.elementRef = elementRef;
        }
        AvlParagraphDirective.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        __decorate$5([
            core.Input('avl-paragraph'),
            __metadata$5("design:type", String)
        ], AvlParagraphDirective.prototype, "paragraphName", void 0);
        AvlParagraphDirective = __decorate$5([ __metadata$5("design:paramtypes", [core.ElementRef])
        ], AvlParagraphDirective);
AvlParagraphDirective.ɵfac = function AvlParagraphDirective_Factory(t) { return new (t || AvlParagraphDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlParagraphDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlParagraphDirective, selectors: [["", "avl-paragraph", ""]], inputs: { paragraphName: ["avl-paragraph", "paragraphName"] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlParagraphDirective, [{
        type: core.Directive,
        args: [{
                selector: '[avl-paragraph]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { paragraphName: [{
            type: core.Input,
            args: ['avl-paragraph']
        }] }); })();
        return AvlParagraphDirective;
    }());

    var __decorate$6 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$6 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    /**
    * avl-long-page should only be used inside avl-app-layout. The only scrollable container should
    * be avl-app-layout (already set by default) and avl-long-page and components that are using it must
    * not scroll in order for avl-long-page to work properly.
    */
    var AvlLongPageComponent = /** @class */ (function () {
        /**
        * @hidden
        */
        function AvlLongPageComponent(router, renderer, platformId) {
            var _this = this;
            this.router = router;
            this.renderer = renderer;
            this.platformId = platformId;
            this.scrollListener = function (event) {
                var newParagraph;
                if (_this.scrollContainer.scrollHeight < Math.round(_this.scrollContainer.scrollTop + _this.scrollContainer.getBoundingClientRect().height)) {
                    newParagraph = _this.paragraphs.last;
                }
                else {
                    _this.paragraphs.forEach(function (paragraph) {
                        var p = paragraph.elementRef.nativeElement;
                        var rect = p.getBoundingClientRect();
                        if (rect.top < window.innerHeight * 0.1) {
                            newParagraph = paragraph;
                        }
                    });
                }
                if (newParagraph) {
                    _this.currentFragment = newParagraph.paragraphName;
                    _this.router.navigate([], { queryParamsHandling: 'preserve', fragment: newParagraph.paragraphName });
                }
            };
        }
        AvlLongPageComponent.prototype.handleRouteChange = function () {
            var paragraphName = this.router.parseUrl(this.router.url).fragment;
            var paragraph = this.paragraphs.find(function (p) { return p.paragraphName === paragraphName; });
            if (this.currentFragment === paragraphName) {
                return;
            }
            if (paragraph) {
                setTimeout(function () {
                    return paragraph.elementRef.nativeElement.scrollIntoView({ behavior: 'instant', block: 'start', inline: 'nearest' });
                }, 0);
            }
        };
        /**
        * @hidden
        */
        AvlLongPageComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (common.isPlatformBrowser(this.platformId)) {
                this.scrollContainer = document.querySelector('avl-app-layout main');
                this.scrollSubscription = this.renderer.listen(this.scrollContainer, 'scroll', this.scrollListener);
            }
            this.handleRouteChange();
            this.routerSubscription = this.router.events.subscribe(function (newRoute) {
                if (newRoute instanceof router.NavigationEnd) {
                    _this.handleRouteChange();
                }
            });
            this.paragraphsSubscription = this.paragraphs.changes.subscribe(function (_) {
                /* if scrollIntoView has not occured yet try it with new paragraphs
                   this is needed for components which render content (paragraphs)
                   after view has been initialized
                */
                if (!_this.currentFragment) {
                    _this.handleRouteChange();
                }
            });
        };
        /**
        * @hidden
        */
        AvlLongPageComponent.prototype.ngOnDestroy = function () {
            if (this.routerSubscription) {
                this.routerSubscription.unsubscribe();
            }
            if (this.paragraphsSubscription) {
                this.paragraphsSubscription.unsubscribe();
            }
            if (this.scrollSubscription) {
                this.scrollSubscription();
            }
        };
        AvlLongPageComponent.ctorParameters = function () { return [
            { type: router.Router },
            { type: core.Renderer2 },
            { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] }
        ]; };
        __decorate$6([
            core.ContentChildren(AvlParagraphDirective),
            __metadata$6("design:type", core.QueryList)
        ], AvlLongPageComponent.prototype, "paragraphs", void 0);
        AvlLongPageComponent = __decorate$6([ __param(2, core.Inject(core.PLATFORM_ID)),
            __metadata$6("design:paramtypes", [router.Router, core.Renderer2, Object])
        ], AvlLongPageComponent);
AvlLongPageComponent.ɵfac = function AvlLongPageComponent_Factory(t) { return new (t || AvlLongPageComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(core.PLATFORM_ID)); };
AvlLongPageComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlLongPageComponent, selectors: [["avl-long-page"]], contentQueries: function AvlLongPageComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, AvlParagraphDirective, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.paragraphs = _t);
    } }, ngContentSelectors: _c3, decls: 1, vars: 0, template: function AvlLongPageComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, styles: [":host{font-family:var(--avl-font-family)}"], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlLongPageComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-long-page',
                template: "<ng-content></ng-content>",
                encapsulation: core.ViewEncapsulation.None,
                styles: [":host{font-family:var(--avl-font-family)}"]
            }]
    }], function () { return [{ type: ɵngcc2.Router }, { type: ɵngcc0.Renderer2 }, { type: Object, decorators: [{
                type: core.Inject,
                args: [core.PLATFORM_ID]
            }] }]; }, { paragraphs: [{
            type: core.ContentChildren,
            args: [AvlParagraphDirective]
        }] }); })();
        return AvlLongPageComponent;
    }());

    /** @hidden */
    var KeyboardNavigator = /** @class */ (function () {
        function KeyboardNavigator(numberOfItems, gridWidth) {
            this.numberOfItems = numberOfItems;
            this.gridWidth = gridWidth;
        }
        /** @returns to which item to navigate to from the current item and undefined to not move. */
        KeyboardNavigator.prototype.navigate = function (currentItemIndex, keyPressed, nextGroupSize, prevGroupSize) {
            if (nextGroupSize === void 0) { nextGroupSize = 0; }
            if (prevGroupSize === void 0) { prevGroupSize = 0; }
            nextGroupSize = nextGroupSize || 0;
            prevGroupSize = prevGroupSize || 0;
            var currentItemColumn = currentItemIndex % this.gridWidth;
            var currentItemRow = Math.floor(currentItemIndex / this.gridWidth);
            var maxRows = Math.ceil(this.numberOfItems / this.gridWidth);
            var nextIndex;
            switch (keyPressed) {
                case 'ArrowUp':
                case 'Up':
                    nextIndex = currentItemIndex - this.gridWidth;
                    if (currentItemRow === 0) {
                        if (prevGroupSize === 0) {
                            return undefined;
                        }
                        if (nextIndex < 0) {
                            var prevLastIndex = prevGroupSize - 1;
                            // columns of last element and targeted element of previous group
                            // - 0 based index
                            var lastColumn = prevLastIndex % this.gridWidth;
                            return currentItemColumn <= lastColumn
                                ? currentItemColumn - lastColumn - 1 // negative indexes -> -1 is last element
                                : prevGroupSize <= this.gridWidth
                                    // focus last element of previous group if focus should go on 2 groups before current
                                    ? -1
                                    : -this.gridWidth - currentItemColumn;
                        }
                    }
                    return nextIndex;
                case 'ArrowDown':
                case 'Down': {
                    nextIndex = currentItemIndex + this.gridWidth;
                    var indexExists = nextIndex < this.numberOfItems;
                    if (currentItemRow === maxRows - 1) {
                        if (nextGroupSize === 0) {
                            return undefined;
                        }
                        if (!indexExists) {
                            var nextLastIndex = nextGroupSize - 1;
                            var lastColumn = nextLastIndex % this.gridWidth;
                            return lastColumn >= currentItemColumn
                                ? currentItemColumn
                                : nextGroupSize <= this.gridWidth
                                    ? 0
                                    : this.gridWidth + currentItemColumn;
                        }
                    }
                    else if (!indexExists) {
                        return this.numberOfItems - 1;
                    }
                    return nextIndex;
                }
                case 'ArrowLeft':
                case 'Left':
                    if (currentItemIndex === 0) {
                        return prevGroupSize === 0 ? undefined : -1;
                    }
                    return currentItemIndex - 1;
                case 'ArrowRight':
                case 'Right': {
                    if (currentItemIndex === this.numberOfItems - 1) {
                        return nextGroupSize === 0 ? undefined : 0;
                    }
                    nextIndex = currentItemIndex + 1;
                    return nextIndex;
                }
                default:
                    return undefined;
            }
        };
        return KeyboardNavigator;
    }());

    var __decorate$7 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$7 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __values = (this && this.__values) || function(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    };
    var __read = (this && this.__read) || function (o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    };
    var __spread = (this && this.__spread) || function () {
        for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
        return ar;
    };
    var AvlSolutionOverviewComponent = /** @class */ (function () {
        function AvlSolutionOverviewComponent(cdr) {
            this.cdr = cdr;
            /**
             * Emits search query string on change
             */
            this.searchQueryChange = new core.EventEmitter();
            /**
             * Emits value of the search query when search triggered
             */
            this.search = new core.EventEmitter();
            this.gridWidth$ = new rxjs.BehaviorSubject(-1);
            this.largestItemsIndex = -1;
            this.maxItems = 0;
        }
        Object.defineProperty(AvlSolutionOverviewComponent.prototype, "groupsOfItems", {
            get: function () {
                var e_1, _a;
                var v = [];
                try {
                    for (var _b = __values(this.groupElementRefs), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var gRef = _c.value;
                        var gEl = gRef.nativeElement;
                        var items = Array.prototype.slice.call(gEl.querySelectorAll('a.item'));
                        v = __spread(v, [items]);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return v;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AvlSolutionOverviewComponent.prototype, "iconButtonElements", {
            get: function () {
                if (this.iconButtonElementsCache == null) {
                    this.iconButtonElementsCache = Array.prototype.slice.call(this.iconButtonsContainer
                        .nativeElement.querySelectorAll('avl-icon-button'));
                }
                return this.iconButtonElementsCache;
            },
            enumerable: true,
            configurable: true
        });
        AvlSolutionOverviewComponent.prototype.ngOnInit = function () {
            this.setupReferenceVars();
            this.gridWidthClass$ = this.gridWidth$.asObservable().pipe(operators.throttleTime(50, rxjs.asyncScheduler, { trailing: true }), operators.map(function (gridWidth) {
                return "grid-width-" + gridWidth;
            }));
        };
        AvlSolutionOverviewComponent.prototype.ngOnDestroy = function () {
            if (this.groupElementRefsChangesSubscription) {
                this.groupElementRefsChangesSubscription.unsubscribe();
            }
        };
        AvlSolutionOverviewComponent.prototype.ngOnChanges = function (changes) {
            if (changes.groups && !changes.groups.isFirstChange()) {
                this.setupReferenceVars();
                this.gridWidth$.next(this.getGridWidth());
                this.setTopMargin();
            }
        };
        /** @hidden */
        AvlSolutionOverviewComponent.prototype.ngAfterViewInit = function () {
            this.setTopMargin();
            this.setupFocusOrdering();
            this.gridWidth$.next(this.getGridWidth());
        };
        /** @hidden */
        AvlSolutionOverviewComponent.prototype.getMaxColumnsClass = function () {
            if (this.maxColumns && window.outerWidth === this.screenWidth) {
                return "max-columns-" + this.maxColumns;
            }
            if (this.maxItems === 0) {
                return;
            }
            var maxColumns;
            switch (this.maxItems) {
                case 1:
                    maxColumns = 1;
                    break;
                case 2:
                case 4:
                    maxColumns = 2;
                    break;
                case 3:
                case 5:
                case 6:
                case 9:
                    maxColumns = 3;
                    break;
                case 7:
                case 8:
                case 11:
                case 12:
                    maxColumns = 4;
                    break;
                default:
                    maxColumns = 5;
            }
            this.maxColumns = maxColumns;
            return "max-columns-" + maxColumns;
        };
        /** @hidden **/
        AvlSolutionOverviewComponent.prototype.onResize = function () {
            this.gridWidth$.next(this.getGridWidth());
        };
        /** @hidden */
        AvlSolutionOverviewComponent.prototype.onKeydownHandler = function (event) {
            var e_2, _a;
            var gOfItems = this.groupsOfItems;
            var numGroups = gOfItems.length;
            if (numGroups === 0) {
                return;
            }
            var activeEl = document.activeElement;
            var isDown = ['Down', 'ArrowDown'].includes(event.key);
            var isUp = ['Up', 'ArrowUp'].includes(event.key);
            var isRight = ['Right', 'ArrowRight'].includes(event.key);
            var isLeft = ['Left', 'ArrowLeft'].includes(event.key);
            var isTab = event.key === 'Tab';
            var searchbar = this.showSearchBar
                ? this.searchbarRef.nativeElement
                : undefined;
            var gNum;
            var iIdx;
            try {
                for (var gOfItems_1 = __values(gOfItems), gOfItems_1_1 = gOfItems_1.next(); !gOfItems_1_1.done; gOfItems_1_1 = gOfItems_1.next()) {
                    var items = gOfItems_1_1.value;
                    if (items.indexOf(activeEl) !== -1) {
                        iIdx = items.indexOf(activeEl);
                        gNum = gOfItems.indexOf(items);
                        // Go to route on 'Space'. 'Enter' already works on <a>.
                        if (event.key === ' ') {
                            items[iIdx].click();
                            return;
                        }
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (gOfItems_1_1 && !gOfItems_1_1.done && (_a = gOfItems_1.return)) _a.call(gOfItems_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            if (this.iconButtonElements.includes(activeEl)) {
                var numIconButtons = this.iconButtonElements.length;
                var idx = this.iconButtonElements.indexOf(activeEl);
                if ((isRight || isTab) && idx === numIconButtons - 1) {
                    if (searchbar != null) {
                        searchbar.focus();
                    }
                    else {
                        gOfItems[0][0].focus();
                    }
                }
                else if (isLeft && idx === 0) {
                    var g = gOfItems[numGroups - 1];
                    var gLen = g.length;
                    g[gLen - 1].focus();
                }
                else if (isLeft) {
                    this.iconButtonElements[idx - 1].focus();
                }
                else if (isRight) {
                    this.iconButtonElements[idx + 1].focus();
                }
                event.preventDefault();
            }
            var itemElements = gOfItems[gNum];
            var gridWidth = this.gridWidth$.value;
            if (searchbar != null && iIdx < gridWidth && gNum === 0 && isUp) {
                searchbar.focus();
                event.preventDefault();
            }
            else if ((isRight || isTab) &&
                gNum === gOfItems.length - 1 &&
                iIdx === itemElements.length - 1) {
                var iconButtons = this.iconButtonElements;
                if (iconButtons != null && iconButtons.length > 0) {
                    iconButtons[0].focus();
                    event.preventDefault();
                }
            }
            else if (searchbar != null && searchbar === activeEl && isDown) {
                gOfItems[0][0].focus();
                event.preventDefault();
            }
            else if (gNum != null) {
                var nextGroup = gNum === gOfItems.length - 1 ? undefined : gOfItems[gNum + 1];
                var prevGroup = gNum === 0 ? undefined : gOfItems[gNum - 1];
                var nGLen = nextGroup == null ? 0 : nextGroup.length;
                var pGLen = prevGroup == null ? 0 : prevGroup.length;
                var keyboardNavigator = new KeyboardNavigator(itemElements.length, gridWidth);
                var itemIndexToMoveTo = keyboardNavigator.navigate(iIdx, event.key, nGLen, pGLen);
                if (itemIndexToMoveTo != null) {
                    if ((isUp || isLeft) && itemIndexToMoveTo < 0) {
                        prevGroup[itemIndexToMoveTo + pGLen].focus();
                    }
                    else if ((isDown || isRight) && iIdx + 1 > itemIndexToMoveTo) {
                        nextGroup[itemIndexToMoveTo].focus();
                    }
                    else {
                        gOfItems[gNum][itemIndexToMoveTo].focus();
                    }
                    event.preventDefault();
                }
            }
        };
        /** @hidden */
        AvlSolutionOverviewComponent.prototype.onQueryChange = function (event) {
            this.searchQueryChange.emit(event.detail.value);
        };
        /** @hidden */
        AvlSolutionOverviewComponent.prototype.onSearch = function (event) {
            this.search.emit(event.detail.value);
        };
        AvlSolutionOverviewComponent.prototype.setupReferenceVars = function () {
            var e_3, _a;
            var max = 0;
            try {
                for (var _b = __values(this.groups), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var g = _c.value;
                    var itemSz = g.items.length;
                    if (max < itemSz) {
                        max = itemSz;
                        this.largestItemsIndex = this.groups.indexOf(g);
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
            this.maxItems = max;
        };
        AvlSolutionOverviewComponent.prototype.setTopMargin = function () {
            this.contentHasScroll = false;
            this.cdr.detectChanges();
            var content = this.groupsContainer.nativeElement;
            this.contentHasScroll = content.scrollHeight > content.clientHeight;
            this.cdr.detectChanges();
        };
        AvlSolutionOverviewComponent.prototype.getGridWidth = function () {
            var e_4, _a;
            if (this.largestItemsIndex < 0) {
                this.setupReferenceVars();
            }
            var items = this.groupsOfItems[this.largestItemsIndex];
            var firstItemOffsetTop = items[0].offsetTop;
            var i = 0;
            try {
                for (var items_1 = __values(items), items_1_1 = items_1.next(); !items_1_1.done; items_1_1 = items_1.next()) {
                    var item = items_1_1.value;
                    if (item.offsetTop > firstItemOffsetTop) {
                        return i;
                    }
                    i++;
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (items_1_1 && !items_1_1.done && (_a = items_1.return)) _a.call(items_1);
                }
                finally { if (e_4) throw e_4.error; }
            }
            return items.length;
        };
        AvlSolutionOverviewComponent.prototype.setupFocusOrdering = function () {
            var _this = this;
            var buttonElements = this.buttonRefs.map(function (elRef) { return elRef.nativeElement; });
            var focusableElements = [];
            if (this.showSearchBar) {
                focusableElements.push(this.searchbarRef.nativeElement);
            }
            focusableElements = __spread(focusableElements, buttonElements, this.iconButtonElements);
            focusableElements.forEach(function (el) {
                el.addEventListener('focus', function (event) {
                    var t = event.target;
                    if (t.tagName.toLowerCase() === 'avl-icon-button') {
                        if (_this.showSearchBar && (['body', 'html'].includes(document.activeElement.tagName.toLowerCase()) ||
                            _this.lastFocusedEl == null ||
                            (_this.lastFocusedEl.tagName.toLowerCase() !== 'avl-icon-button' &&
                                _this.lastFocusedEl !== buttonElements[buttonElements.length - 1]))) {
                            _this.searchbarRef.nativeElement.focus();
                            event.preventDefault();
                        }
                    }
                    _this.lastFocusedEl = event.target;
                });
            });
        };
        AvlSolutionOverviewComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef }
        ]; };
        __decorate$7([
            core.Input(),
            __metadata$7("design:type", Array)
        ], AvlSolutionOverviewComponent.prototype, "groups", void 0);
        __decorate$7([
            core.Input(),
            __metadata$7("design:type", String)
        ], AvlSolutionOverviewComponent.prototype, "solutionName", void 0);
        __decorate$7([
            core.Input(),
            __metadata$7("design:type", String)
        ], AvlSolutionOverviewComponent.prototype, "solutionIconUrl", void 0);
        __decorate$7([
            core.Input(),
            __metadata$7("design:type", String)
        ], AvlSolutionOverviewComponent.prototype, "solutionSubtitle", void 0);
        __decorate$7([
            core.Input(),
            __metadata$7("design:type", Boolean)
        ], AvlSolutionOverviewComponent.prototype, "showSearchBar", void 0);
        __decorate$7([
            core.Output(),
            __metadata$7("design:type", Object)
        ], AvlSolutionOverviewComponent.prototype, "searchQueryChange", void 0);
        __decorate$7([
            core.Output(),
            __metadata$7("design:type", Object)
        ], AvlSolutionOverviewComponent.prototype, "search", void 0);
        __decorate$7([
            core.ViewChildren('itemElement'),
            __metadata$7("design:type", core.QueryList)
        ], AvlSolutionOverviewComponent.prototype, "buttonRefs", void 0);
        __decorate$7([
            core.ViewChildren('group'),
            __metadata$7("design:type", core.QueryList)
        ], AvlSolutionOverviewComponent.prototype, "groupElementRefs", void 0);
        __decorate$7([
            core.ViewChild('searchbar'),
            __metadata$7("design:type", core.ElementRef)
        ], AvlSolutionOverviewComponent.prototype, "searchbarRef", void 0);
        __decorate$7([
            core.ViewChild('iconButtons'),
            __metadata$7("design:type", core.ElementRef)
        ], AvlSolutionOverviewComponent.prototype, "iconButtonsContainer", void 0);
        __decorate$7([
            core.ViewChild('searchbarContainer'),
            __metadata$7("design:type", core.ElementRef)
        ], AvlSolutionOverviewComponent.prototype, "searchbarContainer", void 0);
        __decorate$7([
            core.ViewChild('groupsContainer'),
            __metadata$7("design:type", core.ElementRef)
        ], AvlSolutionOverviewComponent.prototype, "groupsContainer", void 0);
        __decorate$7([
            core.HostListener('window:resize'),
            __metadata$7("design:type", Function),
            __metadata$7("design:paramtypes", []),
            __metadata$7("design:returntype", void 0)
        ], AvlSolutionOverviewComponent.prototype, "onResize", null);
        __decorate$7([
            core.HostListener('keydown', ['$event']),
            __metadata$7("design:type", Function),
            __metadata$7("design:paramtypes", [KeyboardEvent]),
            __metadata$7("design:returntype", void 0)
        ], AvlSolutionOverviewComponent.prototype, "onKeydownHandler", null);
        AvlSolutionOverviewComponent = __decorate$7([ __metadata$7("design:paramtypes", [core.ChangeDetectorRef])
        ], AvlSolutionOverviewComponent);
AvlSolutionOverviewComponent.ɵfac = function AvlSolutionOverviewComponent_Factory(t) { return new (t || AvlSolutionOverviewComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
AvlSolutionOverviewComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlSolutionOverviewComponent, selectors: [["avl-solution-overview"]], viewQuery: function AvlSolutionOverviewComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c6, true);
        ɵngcc0.ɵɵviewQuery(_c7, true);
        ɵngcc0.ɵɵviewQuery(_c8, true);
        ɵngcc0.ɵɵviewQuery(_c9, true);
        ɵngcc0.ɵɵviewQuery(_c10, true);
        ɵngcc0.ɵɵviewQuery(_c11, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchbarRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.iconButtonsContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchbarContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.groupsContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.buttonRefs = _t);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.groupElementRefs = _t);
    } }, hostBindings: function AvlSolutionOverviewComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("resize", function AvlSolutionOverviewComponent_resize_HostBindingHandler() { return ctx.onResize(); }, false, ɵngcc0.ɵɵresolveWindow)("keydown", function AvlSolutionOverviewComponent_keydown_HostBindingHandler($event) { return ctx.onKeydownHandler($event); });
    } }, inputs: { groups: "groups", solutionName: "solutionName", solutionIconUrl: "solutionIconUrl", solutionSubtitle: "solutionSubtitle", showSearchBar: "showSearchBar" }, outputs: { searchQueryChange: "searchQueryChange", search: "search" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c3, decls: 22, vars: 11, consts: [[1, "container"], [1, "top-bar"], [1, "row"], [1, "solution-logo", 3, "src"], [1, "solution-name"], [1, "action-buttons"], ["iconButtons", ""], ["class", "row", 4, "ngIf"], [1, "outer-container", 3, "ngClass"], [1, "searchbar-container", 3, "ngClass"], ["searchbarContainer", ""], ["class", "search-bar", "always-open", "", 3, "query-changed", "search", 4, "ngIf"], [1, "groups"], ["groupsContainer", ""], [4, "ngFor", "ngForOf"], ["groupTpl", ""], ["buttonContent", ""], [1, "solution-subtitle"], ["always-open", "", 1, "search-bar", 3, "query-changed", "search"], ["searchbar", ""], [1, "group-container"], [1, "item-container", 3, "ngClass"], ["group", ""], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], ["class", "group-name", 4, "ngIf"], ["class", "text-container", 4, "ngIf"], [1, "group-name"], [1, "text-container"], [4, "ngIf"], ["tabindex", "0", 1, "item", 3, "routerLink", "ngClass"], ["itemElement", ""], ["tabindex", "0", 1, "item", 3, "ngClass"], [1, "icon-container", 3, "ngClass"], ["class", "item-icon", 3, "icon", 4, "ngIf"], ["class", "item-icon", 3, "src", 4, "ngIf"], [1, "icon-label"], [1, "label"], [1, "description"], [1, "text"], [1, "item-icon", 3, "icon"], [1, "item-icon", 3, "src"]], template: function AvlSolutionOverviewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "img", 3);
        ɵngcc0.ɵɵelementStart(4, "h1", 4);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5, 6);
        ɵngcc0.ɵɵprojection(8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, AvlSolutionOverviewComponent_div_9_Template, 3, 1, "div", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 8);
        ɵngcc0.ɵɵelementStart(11, "div", 9, 10);
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵtemplate(14, AvlSolutionOverviewComponent_avl_search_bar_14_Template, 2, 0, "avl-search-bar", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "div", 12, 13);
        ɵngcc0.ɵɵtemplate(17, AvlSolutionOverviewComponent_ng_container_17_Template, 5, 5, "ng-container", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(18, AvlSolutionOverviewComponent_ng_template_18_Template, 4, 4, "ng-template", null, 15, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(20, AvlSolutionOverviewComponent_ng_template_20_Template, 9, 7, "ng-template", null, 16, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("src", ctx.solutionIconUrl, ɵngcc0.ɵɵsanitizeUrl);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.solutionName);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.solutionSubtitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c16, ctx.contentHasScroll === false));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpipeBind1(13, 7, ctx.gridWidthClass$));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showSearchBar);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.groups);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc1.NgForOf, ɵngcc3.AvlSearchBarDirective, ɵngcc3.AvlFormControlDirective, ɵngcc3.AvlInteractionControlDirective, ɵngcc1.NgTemplateOutlet, ɵngcc2.RouterLinkWithHref], pipes: [ɵngcc1.AsyncPipe], styles: ["[_nghost-%COMP%]{--avl-solution-overview-background-color:var(--avl-background-1-color);--avl-solution-overview-background-image:auto;--avl-solution-overview-text-color:var(--avl-foreground-color);--avl-solution-overview-title-color:var(--avl-foreground-color);--avl-solution-overview-subtitle-color:var(--avl-foreground-medium-color);--avl-solution-overview-item-background-color:var(\r\n    --avl-container-background-color\r\n  );--avl-solution-overview-item-hover-color:var(\r\n    --avl-container-background-product-hover-color,\r\n  );--avl-solution-overview-item-focus-color:var(\r\n    --avl-container-background-product-focus-color\r\n  );--avl-solution-overview-item-pressed-color:var(\r\n    --avl-container-background-product-pressed-color\r\n  );--avl-solution-overview-item-icon-circle-color:var(--avl-product-color);--avl-solution-overview-item-icon-color:var(--avl-on-primary-color);--avl-solution-overview-item-icon-color-no-circle:var(--avl-hint-color);--avl-solution-overview-item-description-color:var(\r\n    --avl-foreground-medium-color\r\n  );--avl-solution-overview-item-icon-pressed-color:var(\r\n    --avl-solution-overview-item-icon-color\r\n  );display:block;width:100%;--item-width:184px;--item-margin:24px;--total-width:calc(var(--item-width) + var(--item-margin));--total-height:100vh;height:var(--total-height);font-family:var(--avl-font-family)}.container[_ngcontent-%COMP%]{background-image:var(--avl-solution-overview-background-image);background-size:cover;background-color:var(--avl-solution-overview-background-color);height:100%;display:flex;flex-direction:column}.groups[_ngcontent-%COMP%]{margin-bottom:24px;width:99%;overflow-y:auto}.group-container[_ngcontent-%COMP%]{display:flex;width:100%;justify-content:center;position:relative}.outer-container[_ngcontent-%COMP%]{display:flex;align-items:center;flex-direction:column;justify-content:center;overflow:hidden;height:100%}.outer-container.top-margin[_ngcontent-%COMP%]{margin-top:-50px}.searchbar-container[_ngcontent-%COMP%]{display:flex;justify-content:space-between}.search-bar[_ngcontent-%COMP%]{padding-right:20px}.external-link-container[_ngcontent-%COMP%], .item-container[_ngcontent-%COMP%]{display:grid;width:100%;margin:0 12px 2px;padding-top:24px;grid-template-columns:repeat(auto-fill,var(--item-width)) [end];grid-gap:var(--item-margin);justify-content:center}.text-container[_ngcontent-%COMP%]{color:var(--avl-solution-overview-text-color);grid-column:1/end}.max-columns-1[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 1)}.max-columns-2[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 2)}.max-columns-3[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 3)}.max-columns-4[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 4)}.max-columns-5[_ngcontent-%COMP%]{max-width:calc(var(--total-width) * 5)}.grid-width-1[_ngcontent-%COMP%]{width:calc(var(--total-width) * 1)}.grid-width-2[_ngcontent-%COMP%]{width:calc(var(--total-width) * 2)}.grid-width-3[_ngcontent-%COMP%]{width:calc(var(--total-width) * 3)}.grid-width-4[_ngcontent-%COMP%]{width:calc(var(--total-width) * 4)}.grid-width-5[_ngcontent-%COMP%]{width:calc(var(--total-width) * 5)}.item[_ngcontent-%COMP%]{position:relative;display:flex;flex-direction:column;align-items:center;padding:12px;width:var(--item-width);height:112px;box-sizing:border-box;background-color:var(--avl-solution-overview-item-background-color);outline:0;box-shadow:var(--avl-box-shadow-1dp);transition:box-shadow .15s cubic-bezier(.4,0,.2,1)}.item.with-description[_ngcontent-%COMP%]{height:204px}.description[_ngcontent-%COMP%]{display:none}.item.with-description[_ngcontent-%COMP%]   .description[_ngcontent-%COMP%]{overflow:hidden;display:block;font-weight:400;text-align:center;font-size:14px;line-height:16px;color:var(--avl-foreground-medium-color)}.item[_ngcontent-%COMP%]:active, .item[_ngcontent-%COMP%]:focus, .item[_ngcontent-%COMP%]:hover{box-shadow:var(--avl-box-shadow-8dp)}.item[_ngcontent-%COMP%]:focus{background-color:var(--avl-solution-overview-item-focus-color)}.item[_ngcontent-%COMP%]:hover{background-color:var(--avl-solution-overview-item-hover-color)}.item[_ngcontent-%COMP%]:active{background-color:var(--avl-solution-overview-item-pressed-color)}.icon-container[_ngcontent-%COMP%]{background:var(--avl-solution-overview-item-icon-circle-color);display:flex;align-items:center;justify-content:center;width:48px;height:48px;min-height:48px;border-radius:50%}.icon-container.no-icon-circle[_ngcontent-%COMP%]{background:0 0}.no-icon-circle[_ngcontent-%COMP%]   .item-icon[_ngcontent-%COMP%]{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color-no-circle)}.item-icon[_ngcontent-%COMP%]{width:24px;height:24px;--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color)}.item[_ngcontent-%COMP%]:active   .item-icon[_ngcontent-%COMP%]{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-pressed-color)}.icon-label[_ngcontent-%COMP%]{display:flex;justify-content:center;flex-flow:column;width:100%;font-size:14px;line-height:16px;margin:8px 0;font-weight:500;text-align:center;height:32px;color:var(--avl-solution-overview-text-color)}.group-name[_ngcontent-%COMP%]{margin:0 0 -8px;font-weight:400;font-size:24px;color:var(--avl-solution-overview-text-color);grid-column:1/end}.solution-name[_ngcontent-%COMP%]{font-weight:400;font-size:24px;margin:0 0 0 32px;height:32px;line-height:32px;flex:auto;color:var(--avl-solution-overview-title-color)}.solution-subtitle[_ngcontent-%COMP%]{font-weight:400;font-size:20px;margin:0 0 0 64px;color:var(--avl-solution-overview-subtitle-color)}.top-bar[_ngcontent-%COMP%]{display:flex;flex-direction:column;padding:32px 32px 0}.top-bar[_ngcontent-%COMP%]   .row[_ngcontent-%COMP%]{display:flex;flex-direction:row;justify-content:space-between;align-items:flex-start;height:32px}.top-bar[_ngcontent-%COMP%]   .row[_ngcontent-%COMP%]:last-child{margin-top:12px;height:24px}.action-buttons[_ngcontent-%COMP%]{margin-top:-8px;justify-self:flex-end;height:32px}.solution-logo[_ngcontent-%COMP%]{top:32px;left:32px;width:32px;height:32px}avl-search-bar[_ngcontent-%COMP%]{padding-top:20px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlSolutionOverviewComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-solution-overview',
                template: "<div class=\"container\">\r\n  <div class=\"top-bar\">\r\n    <div class=\"row\">\r\n      <img class=\"solution-logo\" src=\"{{solutionIconUrl}}\" />\r\n      <h1 class=\"solution-name\">{{solutionName}}</h1>\r\n      <div class=\"action-buttons\" #iconButtons>\r\n        <ng-content></ng-content>\r\n      </div>\r\n    </div>\r\n    <div class=\"row\" *ngIf=\"solutionSubtitle\">\r\n      <h2 class=\"solution-subtitle\">{{solutionSubtitle}}</h2>\r\n    </div>\r\n  </div>\r\n  <div class=\"outer-container\" [ngClass]=\"{'top-margin': contentHasScroll === false}\">\r\n    <div class=\"searchbar-container\" [ngClass]=\"gridWidthClass$ | async\" #searchbarContainer>\r\n      <avl-search-bar class=\"search-bar\" *ngIf=\"showSearchBar\"\r\n        #searchbar\r\n        always-open\r\n        (query-changed)=\"onQueryChange($event)\"\r\n        (search)=\"onSearch($event)\">\r\n      </avl-search-bar>\r\n    </div>\r\n    <div class=\"groups\" #groupsContainer>\r\n      <ng-container *ngFor=\"let group of groups\">\r\n        <div class=\"group-container\">\r\n          <div class=\"item-container\" #group [ngClass]=\"getMaxColumnsClass()\">\r\n            <ng-container *ngTemplateOutlet=\"groupTpl; context: { group: group }\"></ng-container>\r\n          </div>\r\n        </div>\r\n      </ng-container>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #groupTpl let-group=\"group\">\r\n  <h2 class=\"group-name\" *ngIf=\"group.title\">{{group.title}}</h2>\r\n  <div class=\"text-container\" *ngIf=\"group.header\">{{group.header}}</div>\r\n  <ng-container *ngFor=\"let item of group.items\">\r\n    <ng-container *ngIf=\"item.route\">\r\n      <a class=\"item\"\r\n         #itemElement\r\n         [routerLink]=\"item.route\"\r\n         [ngClass]=\"{'with-description': group.description}\"\r\n         tabindex=\"0\">\r\n        <ng-container *ngTemplateOutlet=\"buttonContent; context: { item: item }\">\r\n        </ng-container>\r\n      </a>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"item.link\">\r\n      <a class=\"item\"\r\n         #itemElement\r\n         [attr.href]=\"item.link\"\r\n         [attr.target]=\"item.target\"\r\n         [attr.rel]=\"item.target == '_blank' ? 'noopener' : null\"\r\n         [ngClass]=\"{'with-description': group.description}\"\r\n         tabindex=\"0\">\r\n        <ng-container *ngTemplateOutlet=\"buttonContent; context: { item: item }\">\r\n        </ng-container>\r\n      </a>\r\n    </ng-container>\r\n  </ng-container>\r\n  <div class=\"text-container\" *ngIf=\"group.footer\">{{group.footer}}</div>\r\n</ng-template>\r\n\r\n<ng-template #buttonContent let-item=\"item\">\r\n  <div class=\"icon-container\" [ngClass]=\"{'no-icon-circle': item.noIconCircle}\">\r\n    <avl-icon\r\n      *ngIf=\"item.icon\"\r\n      class=\"item-icon\"\r\n      [icon]=\"item.icon\">\r\n    </avl-icon>\r\n    <img *ngIf=\"item.iconUrl\" class=\"item-icon\" src=\"{{item.iconUrl}}\" />\r\n  </div>\r\n  <div class=\"icon-label\"><span class=\"label\">{{item.label}}</span></div>\r\n  <div class=\"description\">\r\n    <span class=\"text\">\r\n      {{item.description}}\r\n    </span>\r\n  </div>\r\n</ng-template>\r\n",
                styles: [":host{--avl-solution-overview-background-color:var(--avl-background-1-color);--avl-solution-overview-background-image:auto;--avl-solution-overview-text-color:var(--avl-foreground-color);--avl-solution-overview-title-color:var(--avl-foreground-color);--avl-solution-overview-subtitle-color:var(--avl-foreground-medium-color);--avl-solution-overview-item-background-color:var(\r\n    --avl-container-background-color\r\n  );--avl-solution-overview-item-hover-color:var(\r\n    --avl-container-background-product-hover-color,\r\n  );--avl-solution-overview-item-focus-color:var(\r\n    --avl-container-background-product-focus-color\r\n  );--avl-solution-overview-item-pressed-color:var(\r\n    --avl-container-background-product-pressed-color\r\n  );--avl-solution-overview-item-icon-circle-color:var(--avl-product-color);--avl-solution-overview-item-icon-color:var(--avl-on-primary-color);--avl-solution-overview-item-icon-color-no-circle:var(--avl-hint-color);--avl-solution-overview-item-description-color:var(\r\n    --avl-foreground-medium-color\r\n  );--avl-solution-overview-item-icon-pressed-color:var(\r\n    --avl-solution-overview-item-icon-color\r\n  );display:block;width:100%;--item-width:184px;--item-margin:24px;--total-width:calc(var(--item-width) + var(--item-margin));--total-height:100vh;height:var(--total-height);font-family:var(--avl-font-family)}.container{background-image:var(--avl-solution-overview-background-image);background-size:cover;background-color:var(--avl-solution-overview-background-color);height:100%;display:flex;flex-direction:column}.groups{margin-bottom:24px;width:99%;overflow-y:auto}.group-container{display:flex;width:100%;justify-content:center;position:relative}.outer-container{display:flex;align-items:center;flex-direction:column;justify-content:center;overflow:hidden;height:100%}.outer-container.top-margin{margin-top:-50px}.searchbar-container{display:flex;justify-content:space-between}.search-bar{padding-right:20px}.external-link-container,.item-container{display:grid;width:100%;margin:0 12px 2px;padding-top:24px;grid-template-columns:repeat(auto-fill,var(--item-width)) [end];grid-gap:var(--item-margin);justify-content:center}.text-container{color:var(--avl-solution-overview-text-color);grid-column:1/end}.max-columns-1{max-width:calc(var(--total-width) * 1)}.max-columns-2{max-width:calc(var(--total-width) * 2)}.max-columns-3{max-width:calc(var(--total-width) * 3)}.max-columns-4{max-width:calc(var(--total-width) * 4)}.max-columns-5{max-width:calc(var(--total-width) * 5)}.grid-width-1{width:calc(var(--total-width) * 1)}.grid-width-2{width:calc(var(--total-width) * 2)}.grid-width-3{width:calc(var(--total-width) * 3)}.grid-width-4{width:calc(var(--total-width) * 4)}.grid-width-5{width:calc(var(--total-width) * 5)}.item{position:relative;display:flex;flex-direction:column;align-items:center;padding:12px;width:var(--item-width);height:112px;box-sizing:border-box;background-color:var(--avl-solution-overview-item-background-color);outline:0;box-shadow:var(--avl-box-shadow-1dp);transition:box-shadow .15s cubic-bezier(.4,0,.2,1)}.item.with-description{height:204px}.description{display:none}.item.with-description .description{overflow:hidden;display:block;font-weight:400;text-align:center;font-size:14px;line-height:16px;color:var(--avl-foreground-medium-color)}.item:active,.item:focus,.item:hover{box-shadow:var(--avl-box-shadow-8dp)}.item:focus{background-color:var(--avl-solution-overview-item-focus-color)}.item:hover{background-color:var(--avl-solution-overview-item-hover-color)}.item:active{background-color:var(--avl-solution-overview-item-pressed-color)}.icon-container{background:var(--avl-solution-overview-item-icon-circle-color);display:flex;align-items:center;justify-content:center;width:48px;height:48px;min-height:48px;border-radius:50%}.icon-container.no-icon-circle{background:0 0}.no-icon-circle .item-icon{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color-no-circle)}.item-icon{width:24px;height:24px;--avl-icon-custom-color:var(--avl-solution-overview-item-icon-color)}.item:active .item-icon{--avl-icon-custom-color:var(--avl-solution-overview-item-icon-pressed-color)}.icon-label{display:flex;justify-content:center;flex-flow:column;width:100%;font-size:14px;line-height:16px;margin:8px 0;font-weight:500;text-align:center;height:32px;color:var(--avl-solution-overview-text-color)}.group-name{margin:0 0 -8px;font-weight:400;font-size:24px;color:var(--avl-solution-overview-text-color);grid-column:1/end}.solution-name{font-weight:400;font-size:24px;margin:0 0 0 32px;height:32px;line-height:32px;flex:auto;color:var(--avl-solution-overview-title-color)}.solution-subtitle{font-weight:400;font-size:20px;margin:0 0 0 64px;color:var(--avl-solution-overview-subtitle-color)}.top-bar{display:flex;flex-direction:column;padding:32px 32px 0}.top-bar .row{display:flex;flex-direction:row;justify-content:space-between;align-items:flex-start;height:32px}.top-bar .row:last-child{margin-top:12px;height:24px}.action-buttons{margin-top:-8px;justify-self:flex-end;height:32px}.solution-logo{top:32px;left:32px;width:32px;height:32px}avl-search-bar{padding-top:20px}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { searchQueryChange: [{
            type: core.Output
        }], search: [{
            type: core.Output
        }], onResize: [{
            type: core.HostListener,
            args: ['window:resize']
        }], onKeydownHandler: [{
            type: core.HostListener,
            args: ['keydown', ['$event']]
        }], groups: [{
            type: core.Input
        }], solutionName: [{
            type: core.Input
        }], solutionIconUrl: [{
            type: core.Input
        }], solutionSubtitle: [{
            type: core.Input
        }], showSearchBar: [{
            type: core.Input
        }], buttonRefs: [{
            type: core.ViewChildren,
            args: ['itemElement']
        }], groupElementRefs: [{
            type: core.ViewChildren,
            args: ['group']
        }], searchbarRef: [{
            type: core.ViewChild,
            args: ['searchbar']
        }], iconButtonsContainer: [{
            type: core.ViewChild,
            args: ['iconButtons']
        }], searchbarContainer: [{
            type: core.ViewChild,
            args: ['searchbarContainer']
        }], groupsContainer: [{
            type: core.ViewChild,
            args: ['groupsContainer']
        }] }); })();
        return AvlSolutionOverviewComponent;
    }());

    var __decorate$8 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata$8 = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var AvlListExpanderComponent = /** @class */ (function () {
        function AvlListExpanderComponent() {
            /**
             * Boolean that changes modes between wide and narrow, false by default.
             */
            this.narrow = false;
            this._open = false;
            /**
             * Event that emits when the expander is opened or closed.
             * Sends boolean value of new open state.
             */
            this.openChange = new core.EventEmitter();
        }
        Object.defineProperty(AvlListExpanderComponent.prototype, "open", {
            get: function () {
                return this._open;
            },
            /**
             * Boolean that tells whether the expander is open or closed, expander is closed by default.
             */
            set: function (value) {
                this._open = value;
                this.openChange.emit(this._open);
            },
            enumerable: true,
            configurable: true
        });
        AvlListExpanderComponent.prototype.toggleClick = function () {
            this.open = !this.open;
        };
        /** @hidden */
        AvlListExpanderComponent.prototype.keyEvent = function (event) {
            var activeEl = document.activeElement;
            if (event.key === ' ' || event.key === 'Enter') {
                activeEl.click();
                return;
            }
        };
        __decorate$8([
            core.Input(),
            __metadata$8("design:type", Boolean)
        ], AvlListExpanderComponent.prototype, "narrow", void 0);
        __decorate$8([
            core.Input(),
            __metadata$8("design:type", Boolean),
            __metadata$8("design:paramtypes", [Boolean])
        ], AvlListExpanderComponent.prototype, "open", null);
        __decorate$8([
            core.Output(),
            __metadata$8("design:type", Object)
        ], AvlListExpanderComponent.prototype, "openChange", void 0);
        __decorate$8([
            core.HostListener('keydown', ['$event']),
            __metadata$8("design:type", Function),
            __metadata$8("design:paramtypes", [KeyboardEvent]),
            __metadata$8("design:returntype", void 0)
        ], AvlListExpanderComponent.prototype, "keyEvent", null);
        AvlListExpanderComponent = __decorate$8([ __metadata$8("design:paramtypes", [])
        ], AvlListExpanderComponent);
AvlListExpanderComponent.ɵfac = function AvlListExpanderComponent_Factory(t) { return new (t || AvlListExpanderComponent)(); };
AvlListExpanderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AvlListExpanderComponent, selectors: [["avl-list-expander"]], hostBindings: function AvlListExpanderComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown", function AvlListExpanderComponent_keydown_HostBindingHandler($event) { return ctx.keyEvent($event); });
    } }, inputs: { narrow: "narrow", open: "open" }, outputs: { openChange: "openChange" }, ngContentSelectors: _c18, decls: 8, vars: 5, consts: [["id", "expander"], ["tabindex", "0", 3, "ngClass", "click"], [3, "ngClass", 4, "ngIf"], [3, "ngClass"], ["icon", "hardware:keyboard-arrow-down", "id", "starsIcon"], ["icon", "hardware:keyboard-arrow-up", "id", "starsIcon"]], template: function AvlListExpanderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c17);
        ɵngcc0.ɵɵelement(0, "hr");
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵlistener("click", function AvlListExpanderComponent_Template_div_click_2_listener() { return ctx.toggleClick(); });
        ɵngcc0.ɵɵtemplate(3, AvlListExpanderComponent_div_3_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵtemplate(4, AvlListExpanderComponent_div_4_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, AvlListExpanderComponent_div_7_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ctx.narrow ? "title-container-narrow" : "title-container");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.open);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.open);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ctx.narrow ? "title-narrow" : "title");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.open);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgIf], styles: [".title-container[_ngcontent-%COMP%]:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container[_ngcontent-%COMP%]:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container[_ngcontent-%COMP%]:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}.title-container[_ngcontent-%COMP%]{height:48px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.title[_ngcontent-%COMP%]{font-size:17px;color:var(--avl-list-expander-title-color,var(--avl-foreground-color))}.icon[_ngcontent-%COMP%]{margin:12px 12px 12px 24px;height:24px}.content[_ngcontent-%COMP%]{margin:4px 24px}hr[_ngcontent-%COMP%]{display:block;border:0;border-top:1px solid var(--avl-list-expander-divider-color,var(--avl-divider-color));margin:0}avl-icon[_ngcontent-%COMP%]{--avl-icon-custom-color:var(--avl-list-expander-icon-color, var(--avl-hint-color))}.icon-narrow[_ngcontent-%COMP%]{margin:16px 24px 16px 16px;position:absolute;right:0;bottom:0;height:24px}.title-container-narrow[_ngcontent-%COMP%]{position:relative;height:56px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.content-narrow[_ngcontent-%COMP%]{margin:0 24px 24px}.title-narrow[_ngcontent-%COMP%]{margin-left:24px;font-size:17px;margin-right:64px;white-space:nowrap;overflow:hidden}.title-container-narrow[_ngcontent-%COMP%]:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container-narrow[_ngcontent-%COMP%]:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container-narrow[_ngcontent-%COMP%]:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlListExpanderComponent, [{
        type: core.Component,
        args: [{
                selector: 'avl-list-expander',
                template: "<hr/>\r\n<div id=\"expander\">\r\n    <div  [ngClass]=\"narrow ? 'title-container-narrow' : 'title-container'\" tabindex=\"0\" (click)=\"toggleClick()\">\r\n        <div [ngClass]=\"narrow ? 'icon-narrow' : 'icon'\" *ngIf=\"!open\">\r\n            <avl-icon icon=\"hardware:keyboard-arrow-down\" id=\"starsIcon\"></avl-icon>\r\n        </div>\r\n        <div [ngClass]=\"narrow ? 'icon-narrow' : 'icon'\" *ngIf=\"open\">\r\n            <avl-icon icon=\"hardware:keyboard-arrow-up\" id=\"starsIcon\"></avl-icon>\r\n        </div>\r\n        <div [ngClass]=\"narrow ? 'title-narrow' : 'title'\">\r\n            <ng-content select=\"div[role=header]\"></ng-content>\r\n        </div>\r\n    </div>\r\n\r\n    <div *ngIf=\"open\" [ngClass]=\"narrow ? 'content-narrow' : 'content'\">\r\n        <ng-content select=\"div[role=content]\"></ng-content>\r\n    </div>\r\n</div>\r\n\r\n",
                styles: [".title-container:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}.title-container{height:48px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.title{font-size:17px;color:var(--avl-list-expander-title-color,var(--avl-foreground-color))}.icon{margin:12px 12px 12px 24px;height:24px}.content{margin:4px 24px}hr{display:block;border:0;border-top:1px solid var(--avl-list-expander-divider-color,var(--avl-divider-color));margin:0}avl-icon{--avl-icon-custom-color:var(--avl-list-expander-icon-color, var(--avl-hint-color))}.icon-narrow{margin:16px 24px 16px 16px;position:absolute;right:0;bottom:0;height:24px}.title-container-narrow{position:relative;height:56px;display:flex;flex-flow:row;align-items:center;cursor:pointer}.content-narrow{margin:0 24px 24px}.title-narrow{margin-left:24px;font-size:17px;margin-right:64px;white-space:nowrap;overflow:hidden}.title-container-narrow:hover{background-color:var(--avl-list-expander-header-hover-color,var(--avl-ink-default-hover-color))}.title-container-narrow:focus{background-color:var(--avl-list-expander-header-focus-color,var(--avl-ink-default-focus-color));outline:0}.title-container-narrow:active{background-color:var(--avl-list-expander-header-pressed-color,var(--avl-ink-default-pressed-color))}"]
            }]
    }], function () { return []; }, { narrow: [{
            type: core.Input
        }], openChange: [{
            type: core.Output
        }], open: [{
            type: core.Input
        }], keyEvent: [{
            type: core.HostListener,
            args: ['keydown', ['$event']]
        }] }); })();
        return AvlListExpanderComponent;
    }());

    var __decorate$9 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var componentsAndDirectives = [
        AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent,
    ];
    var AvlNavigationModule = /** @class */ (function () {
        function AvlNavigationModule() {
        }
AvlNavigationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AvlNavigationModule });
AvlNavigationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AvlNavigationModule_Factory(t) { return new (t || AvlNavigationModule)(); }, imports: [[
            common.CommonModule,
            router.RouterModule,
            ngBinders.AvlNgBindersModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AvlNavigationModule, { declarations: [AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent], imports: [ɵngcc1.CommonModule, ɵngcc2.RouterModule, ɵngcc3.AvlNgBindersModule], exports: [AvlAppBarComponent,
        AvlAppBarItemComponent,
        AvlSideNavComponent,
        AvlAppLayoutComponent,
        AvlNoPaddingDirective,
        AvlLongPageComponent,
        AvlParagraphDirective,
        AvlSolutionOverviewComponent,
        AvlListExpanderComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlNavigationModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    common.CommonModule,
                    router.RouterModule,
                    ngBinders.AvlNgBindersModule
                ],
                declarations: componentsAndDirectives,
                schemas: [
                    core.CUSTOM_ELEMENTS_SCHEMA
                ],
                exports: componentsAndDirectives
            }]
    }], function () { return []; }, null); })();
        return AvlNavigationModule;
    }());

    var __extends = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var Paragraph = /** @class */ (function () {
        function Paragraph(
        /**
         * Id/fragment of the paragraph header in the page.
         */
        id, 
        /**
         * Label of the paragraph shown in the side nav.
         */
        label) {
            this.id = id;
            this.label = label;
        }
        return Paragraph;
    }());
    var Params = /** @class */ (function () {
        function Params(
        /**
         * Name of the query parameter.
         */
        name, 
        /**
         * Value of the query parameter.
         */
        value) {
            this.name = name;
            this.value = value;
        }
        return Params;
    }());
    var Page = /** @class */ (function () {
        function Page(
        /**
         * Label of the page shown in the side nav.
         */
        label, 
        /**
         * Route to the page.
         */
        route, 
        /**
         * Paragraphs of the long page.
         * It's a regular page if no paragraphs are defined.
         */
        paragraphs, 
        /**
         * Query parameters of the page.
         */
        params, expanded) {
            this.label = label;
            this.route = route;
            this.paragraphs = paragraphs;
            this.params = params;
            this.expanded = expanded;
            this.paragraphs = paragraphs ? paragraphs : [];
            this.params = params ? params : [];
        }
        return Page;
    }());
    var TopLevelPage = /** @class */ (function (_super) {
        __extends(TopLevelPage, _super);
        function TopLevelPage(
        /**
         * Label of the page shown in the side nav.
         */
        label, 
        /**
         * Route to the page.
         */
        route, 
        /**
         * Material icon name for the icon to show next to the label.
         * https://material.io/tools/icons
         * Only one of icon and iconUrl can be used. The other should be undefined.
         */
        icon, 
        /**
         * URL to the icon image to show next to the label.
         * Only one of icon and iconUrl can be used. The other should be undefined.
         */
        iconUrl, 
        /**
         * Paragraphs of the long page.
         * It's a regular page if no paragraphs are defined.
         */
        paragraphs, 
        /**
         * Query parameters of the page.
         */
        params, 
        /**
         * Sub pages of the top level page.
         */
        subItems, expanded) {
            var _this = _super.call(this, label, route, paragraphs, params, expanded) || this;
            _this.icon = icon;
            _this.iconUrl = iconUrl;
            _this.paragraphs = paragraphs;
            _this.params = params;
            _this.subItems = subItems;
            return _this;
        }
        return TopLevelPage;
    }(Page));
    var Group = /** @class */ (function (_super) {
        __extends(Group, _super);
        function Group(
        /**
         * Label of the Group shown in the side nav.
         */
        label, 
        /**
         * Material icon name for the icon to show next to the label.
         * https://material.io/tools/icons
         * Only one of icon and iconUrl can be used. The other should be undefined.
         */
        icon, 
        /**
         * URL to the icon image to show next to the label.
         * Only one of icon and iconUrl can be used. The other should be undefined.
         */
        iconUrl, 
        /**
         * Sub pages of the group.
         */
        subItems, 
        /**
         * Route to the page.
         */
        route, 
        /**
     * Paragraphs of the long page.
     * It's a regular page if no paragraphs are defined.
     */
        paragraphs, params, expanded) {
            var _this = _super.call(this, label, route, paragraphs, params, expanded) || this;
            _this.icon = icon;
            _this.iconUrl = iconUrl;
            _this.subItems = subItems;
            _this.subItems = subItems ? subItems : [];
            return _this;
        }
        return Group;
    }(Page));

    exports.ApplicationReference = ApplicationReference;
    exports.AvlAppBarComponent = AvlAppBarComponent;
    exports.AvlAppBarItemComponent = AvlAppBarItemComponent;
    exports.AvlAppLayoutComponent = AvlAppLayoutComponent;
    exports.AvlListExpanderComponent = AvlListExpanderComponent;
    exports.AvlLongPageComponent = AvlLongPageComponent;
    exports.AvlNavigationModule = AvlNavigationModule;
    exports.AvlNoPaddingDirective = AvlNoPaddingDirective;
    exports.AvlParagraphDirective = AvlParagraphDirective;
    exports.AvlSideNavComponent = AvlSideNavComponent;
    exports.AvlSolutionOverviewComponent = AvlSolutionOverviewComponent;
    exports.Group = Group;
    exports.KeyboardNavigator = KeyboardNavigator;
    exports.MenuButtonItems = MenuButtonItems;
    exports.Page = Page;
    exports.Paragraph = Paragraph;
    exports.Params = Params;
    exports.TopLevelPage = TopLevelPage;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
