// @avl-ng-controls/navigation/esm2015/lib/components/avl-long-page/avl-paragraph.directive.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, Input, ElementRef } from '@angular/core';
let AvlParagraphDirective = class AvlParagraphDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
};
AvlParagraphDirective.ɵfac = function AvlParagraphDirective_Factory(t) { return new (t || AvlParagraphDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AvlParagraphDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AvlParagraphDirective, selectors: [["", "avl-paragraph", ""]], inputs: { paragraphName: ["avl-paragraph", "paragraphName"] } });
AvlParagraphDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input('avl-paragraph'),
    __metadata("design:type", String)
], AvlParagraphDirective.prototype, "paragraphName", void 0);
AvlParagraphDirective = __decorate([ __metadata("design:paramtypes", [ElementRef])
], AvlParagraphDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvlParagraphDirective, [{
        type: Directive,
        args: [{
                selector: '[avl-paragraph]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { paragraphName: [{
            type: Input,
            args: ['avl-paragraph']
        }] }); })();
export { AvlParagraphDirective };
