// @avl-ng-controls/notification/esm2015/lib/notification.module.js
import { CommonModule } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, NgModule } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';
import { AvlNgBindersModule } from '@avl-controls/ng-binders';
import { FormsModule } from '@angular/forms';
import { AvlNotificationMenuComponent } from './components/avl-notification-menu/avl-notification-menu.component';
import { AnimateButton } from './components/avl-notification-container/directives/animate-button.directive';
import { AnimateSnackbar } from './components/avl-notification-container/directives/animate-snackbar.directive';
import { AnimateStack } from './components/avl-notification-container/directives/animate-stack.directive';
import { ShowHideSnackbarDirective } from './components/avl-notification-container/directives/show-hide-snackbar.directive';
import { SnackbarListener } from './components/avl-notification-container/directives/snackbar-listener.directive';
import { AvlNotificationContainerComponent } from './components/avl-notification-container/avl-notification-container.component';
import { AvlNotificationDialogComponent } from './components/avl-notification-dialog/avl-notification-dialog.component';
import { AvlDialogModule } from '@avl-ng-controls/dialog';
import { AvlNavigationModule } from '@avl-ng-controls/navigation';
import { PrepareNotificationsPipe } from './pipes/prepare-notifications.pipe';
import { CapitalizePipe } from './pipes/capitalize.pipe';
import { DateInvalidValidatorDirective } from './components/avl-notification-dialog/date-validator.directive';
import { IterableObjectPipe } from './pipes/iterable-object.pipe';
import * as i0 from "@angular/core";
const componentsAndDirectives = [
    AvlNotificationMenuComponent,
    AvlNotificationContainerComponent,
    SnackbarListener,
    AnimateStack,
    AnimateSnackbar,
    ShowHideSnackbarDirective,
    AnimateButton,
    AvlNotificationDialogComponent,
    PrepareNotificationsPipe,
    CapitalizePipe,
    IterableObjectPipe,
    DateInvalidValidatorDirective
];
export class AvlNotificationModule {
}
AvlNotificationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNotificationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AvlNotificationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNotificationModule, declarations: [AvlNotificationMenuComponent,
        AvlNotificationContainerComponent,
        SnackbarListener,
        AnimateStack,
        AnimateSnackbar,
        ShowHideSnackbarDirective,
        AnimateButton,
        AvlNotificationDialogComponent,
        PrepareNotificationsPipe,
        CapitalizePipe,
        IterableObjectPipe,
        DateInvalidValidatorDirective], imports: [CommonModule,
        TranslateModule,
        AvlNgBindersModule,
        AvlDialogModule,
        AvlNavigationModule,
        FormsModule], exports: [AvlNotificationMenuComponent,
        AvlNotificationContainerComponent,
        AvlNotificationDialogComponent] });
AvlNotificationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNotificationModule, imports: [[
            CommonModule,
            TranslateModule,
            AvlNgBindersModule,
            AvlDialogModule,
            AvlNavigationModule,
            FormsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.3", ngImport: i0, type: AvlNotificationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...componentsAndDirectives],
                    imports: [
                        CommonModule,
                        TranslateModule,
                        AvlNgBindersModule,
                        AvlDialogModule,
                        AvlNavigationModule,
                        FormsModule
                    ],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [
                        AvlNotificationMenuComponent,
                        AvlNotificationContainerComponent,
                        AvlNotificationDialogComponent
                    ]
                }]
        }] });
